{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5992ae5d",
   "metadata": {
    "Rmd_chunk_options": "setup, include = FALSE",
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "knitr::opts_chunk$set(\n",
    "  collapse = TRUE,\n",
    "  comment = \"#>\",\n",
    "  root.dir = './'\n",
    ")\n",
    "#knitr::opts_chunk$set(eval = FAslice2E)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f20502c",
   "metadata": {},
   "source": [
    "This vignette shows how to apply CellChat to identify major signaling changes across different biological conditions by quantitative contrasts and joint manifold learning. We showcase CellChat’s diverse functionalities by applying it to a scRNA-seq data on cells from two biological conditions: noslice1esional (slice1, normal) and lesional (slice2, diseased) human skin from patients with atopic dermatitis. **These two datasets (conditions) have the same cell population compositions after joint clustering. If there are slightly or vastly different cell population compositions between different datasets, please check out another related tutorial. **\n",
    "\n",
    "CellChat employs a top-down approach, i.e., starting with the big picture and then refining it in a greater detail on the signaling mechanisms, to identify signaling changes at different levels, including both general principles of cell-cell communication and dysfunctional cell populations/signaling pathways/ligand-receptors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c4c42a4",
   "metadata": {},
   "source": [
    "## Load the required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5e0382d0",
   "metadata": {
    "Rmd_chunk_options": "message=FALSE,warning=FALSE",
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "python:         /home/chrissy1/.conda/envs/cellchat_env/bin/python\n",
       "libpython:      /home/chrissy1/.conda/envs/cellchat_env/lib/libpython3.11.so\n",
       "pythonhome:     /home/chrissy1/.conda/envs/cellchat_env:/home/chrissy1/.conda/envs/cellchat_env\n",
       "version:        3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0]\n",
       "numpy:          /home/chrissy1/.conda/envs/cellchat_env/lib/python3.11/site-packages/numpy\n",
       "numpy_version:  1.26.4\n",
       "\n",
       "NOTE: Python version was forced by RETICULATE_PYTHON"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: igraph\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "'2.1.2'"
      ],
      "text/latex": [
       "'2.1.2'"
      ],
      "text/markdown": [
       "'2.1.2'"
      ],
      "text/plain": [
       "[1] \"2.1.2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Sys.setenv(RETICULATE_PYTHON = \"/home/chrissy1/.conda/envs/cellchat_env/bin/python\")\n",
    "library(reticulate)\n",
    "py_config()\n",
    "\n",
    "## Load the required libraries\n",
    "library(CellChat)\n",
    "package.version(\"CellChat\")\n",
    "library(patchwork)\n",
    "options(stringsAsFactors = FALSE)\n",
    "library(ggplot2)\n",
    "library(dplyr)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9838c010",
   "metadata": {},
   "source": [
    "# Load CellChat object of each dataset and then merge together"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b8f9c0c",
   "metadata": {},
   "source": [
    "Users need to run CellChat on each dataset seperately and then merge different CellChat objects together.\n",
    "Please do `updateCellChat` if you have CellChat objects that are obtained using the earlier version (< 1.6.0)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7c293779",
   "metadata": {
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cellchat_dir = '/home/chrissy1/spatial/stomics/ovary_froz/redo/seurat/bin50_processed/cellchat/'\n",
    "base.fig.dir <- paste0(cellchat_dir, 'comparison')\n",
    "dir.create(base.fig.dir, showWarnings = FALSE)\n",
    "key = 'Vitri'  # fresh  vitri slow\n",
    "cellchat.vitri= readRDS(paste0(cellchat_dir, key, '_endo_subbed.rds'))\n",
    "key = 'Slow'  # fresh  vitri slow\n",
    "cellchat.slow = readRDS(paste0(cellchat_dir, key, '_endo_subbed.rds'))\n",
    "key = 'Fresh'  # fresh  vitri slow\n",
    "cellchat.fresh = readRDS(paste0(cellchat_dir, key, '_endo_subbed.rds'))\n",
    "\n",
    "\n",
    "celltype_OI = sapply(unique(cellchat.fresh@meta$annot), function(x) {\n",
    "    if (!grepl(\"ST(\", x, fixed = TRUE)) {\n",
    "        return(x)\n",
    "    } else {\n",
    "        return(NA)\n",
    "    }\n",
    "}) \n",
    "celltype_OI = celltype_OI[!is.na(celltype_OI)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8e8b208b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>ST_LIP(INSL3)</dt><dd>'ST_LIP(INSL3)'</dd><dt>B_Plasma(IGKC)</dt><dd>'B_Plasma(IGKC)'</dd><dt>ENDO_SM_1</dt><dd>'ENDO_SM_1'</dd><dt>FIB(COL1A1)</dt><dd>'FIB(COL1A1)'</dd><dt>ENDO_SM_5</dt><dd>'ENDO_SM_5'</dd><dt>ENDO_SM_4</dt><dd>'ENDO_SM_4'</dd><dt>ST_SM(MT1A)</dt><dd>'ST_SM(MT1A)'</dd><dt>O(PDCD5)</dt><dd>'O(PDCD5)'</dd><dt>ST_LIP(GSTA1)</dt><dd>'ST_LIP(GSTA1)'</dd><dt>ENDO_SM_0</dt><dd>'ENDO_SM_0'</dd><dt>ST_SM(ADAMTS4)</dt><dd>'ST_SM(ADAMTS4)'</dd><dt>ENDO_SM_3</dt><dd>'ENDO_SM_3'</dd><dt>ENDO_SM_2</dt><dd>'ENDO_SM_2'</dd><dt>ERY_ENDO(HBB)</dt><dd>'ERY_ENDO(HBB)'</dd><dt>B_Plasma(IGLL5)</dt><dd>'B_Plasma(IGLL5)'</dd><dt>B_Plasma(IGLC2)</dt><dd>'B_Plasma(IGLC2)'</dd><dt>ENDO_SM_6</dt><dd>'ENDO_SM_6'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[ST\\textbackslash{}\\_LIP(INSL3)] 'ST\\_LIP(INSL3)'\n",
       "\\item[B\\textbackslash{}\\_Plasma(IGKC)] 'B\\_Plasma(IGKC)'\n",
       "\\item[ENDO\\textbackslash{}\\_SM\\textbackslash{}\\_1] 'ENDO\\_SM\\_1'\n",
       "\\item[FIB(COL1A1)] 'FIB(COL1A1)'\n",
       "\\item[ENDO\\textbackslash{}\\_SM\\textbackslash{}\\_5] 'ENDO\\_SM\\_5'\n",
       "\\item[ENDO\\textbackslash{}\\_SM\\textbackslash{}\\_4] 'ENDO\\_SM\\_4'\n",
       "\\item[ST\\textbackslash{}\\_SM(MT1A)] 'ST\\_SM(MT1A)'\n",
       "\\item[O(PDCD5)] 'O(PDCD5)'\n",
       "\\item[ST\\textbackslash{}\\_LIP(GSTA1)] 'ST\\_LIP(GSTA1)'\n",
       "\\item[ENDO\\textbackslash{}\\_SM\\textbackslash{}\\_0] 'ENDO\\_SM\\_0'\n",
       "\\item[ST\\textbackslash{}\\_SM(ADAMTS4)] 'ST\\_SM(ADAMTS4)'\n",
       "\\item[ENDO\\textbackslash{}\\_SM\\textbackslash{}\\_3] 'ENDO\\_SM\\_3'\n",
       "\\item[ENDO\\textbackslash{}\\_SM\\textbackslash{}\\_2] 'ENDO\\_SM\\_2'\n",
       "\\item[ERY\\textbackslash{}\\_ENDO(HBB)] 'ERY\\_ENDO(HBB)'\n",
       "\\item[B\\textbackslash{}\\_Plasma(IGLL5)] 'B\\_Plasma(IGLL5)'\n",
       "\\item[B\\textbackslash{}\\_Plasma(IGLC2)] 'B\\_Plasma(IGLC2)'\n",
       "\\item[ENDO\\textbackslash{}\\_SM\\textbackslash{}\\_6] 'ENDO\\_SM\\_6'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "ST_LIP(INSL3)\n",
       ":   'ST_LIP(INSL3)'B_Plasma(IGKC)\n",
       ":   'B_Plasma(IGKC)'ENDO_SM_1\n",
       ":   'ENDO_SM_1'FIB(COL1A1)\n",
       ":   'FIB(COL1A1)'ENDO_SM_5\n",
       ":   'ENDO_SM_5'ENDO_SM_4\n",
       ":   'ENDO_SM_4'ST_SM(MT1A)\n",
       ":   'ST_SM(MT1A)'O(PDCD5)\n",
       ":   'O(PDCD5)'ST_LIP(GSTA1)\n",
       ":   'ST_LIP(GSTA1)'ENDO_SM_0\n",
       ":   'ENDO_SM_0'ST_SM(ADAMTS4)\n",
       ":   'ST_SM(ADAMTS4)'ENDO_SM_3\n",
       ":   'ENDO_SM_3'ENDO_SM_2\n",
       ":   'ENDO_SM_2'ERY_ENDO(HBB)\n",
       ":   'ERY_ENDO(HBB)'B_Plasma(IGLL5)\n",
       ":   'B_Plasma(IGLL5)'B_Plasma(IGLC2)\n",
       ":   'B_Plasma(IGLC2)'ENDO_SM_6\n",
       ":   'ENDO_SM_6'\n",
       "\n"
      ],
      "text/plain": [
       "    ST_LIP(INSL3)    B_Plasma(IGKC)         ENDO_SM_1       FIB(COL1A1) \n",
       "  \"ST_LIP(INSL3)\"  \"B_Plasma(IGKC)\"       \"ENDO_SM_1\"     \"FIB(COL1A1)\" \n",
       "        ENDO_SM_5         ENDO_SM_4       ST_SM(MT1A)          O(PDCD5) \n",
       "      \"ENDO_SM_5\"       \"ENDO_SM_4\"     \"ST_SM(MT1A)\"        \"O(PDCD5)\" \n",
       "    ST_LIP(GSTA1)         ENDO_SM_0    ST_SM(ADAMTS4)         ENDO_SM_3 \n",
       "  \"ST_LIP(GSTA1)\"       \"ENDO_SM_0\"  \"ST_SM(ADAMTS4)\"       \"ENDO_SM_3\" \n",
       "        ENDO_SM_2     ERY_ENDO(HBB)   B_Plasma(IGLL5)   B_Plasma(IGLC2) \n",
       "      \"ENDO_SM_2\"   \"ERY_ENDO(HBB)\" \"B_Plasma(IGLL5)\" \"B_Plasma(IGLC2)\" \n",
       "        ENDO_SM_6 \n",
       "      \"ENDO_SM_6\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "celltype_OI = sapply(as.character(unique(cellchat.fresh@meta$annot)), function(x)\n",
    "    if (!grepl(\"ST(\", x, fixed = TRUE)) {\n",
    "        return(x)\n",
    "    } else {\n",
    "        return(NA)}\n",
    "    , simplify = \"array\") %>% .[!is.na(.)]\n",
    "celltype_OI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check numerical stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "df <- list(vitri=subsetCommunication(cellchat.vitri), slow=subsetCommunication(cellchat.slow), fresh=subsetCommunication(cellchat.fresh))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"union of all samples\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TGFb'</li><li>'BMP'</li><li>'WNT'</li><li>'ncWNT'</li><li>'EGF'</li><li>'NRG'</li><li>'FGF'</li><li>'PDGF'</li><li>'VEGF'</li><li>'IGF'</li><li>'CCL'</li><li>'CXCL'</li><li>'MIF'</li><li>'TWEAK'</li><li>'VISFATIN'</li><li>'ANGPTL'</li><li>'ANGPT'</li><li>'MK'</li><li>'PTN'</li><li>'NT'</li><li>'SEMA3'</li><li>'CALCR'</li><li>'GAS'</li><li>'GRN'</li><li>'PROS'</li><li>'PTPR'</li><li>'IGFBP'</li><li>'CypA'</li><li>'COLLAGEN'</li><li>'FN1'</li><li>'LAMININ'</li><li>'RELN'</li><li>'THBS'</li><li>'TENASCIN'</li><li>'HSPG'</li><li>'27HC'</li><li>'Androstenedione'</li><li>'Androsterone'</li><li>'Cholesterol'</li><li>'Desmosterol'</li><li>'Testosterone'</li><li>'APP'</li><li>'CADM'</li><li>'CD46'</li><li>'CD99'</li><li>'CDH'</li><li>'CDH5'</li><li>'CLDN'</li><li>'CSPG4'</li><li>'EPHA'</li><li>'EPHB'</li><li>'ESAM'</li><li>'JAM'</li><li>'MAG'</li><li>'MPZ'</li><li>'NCAM'</li><li>'NECTIN'</li><li>'NEGR'</li><li>'NOTCH'</li><li>'NRXN'</li><li>'PECAM1'</li><li>'PTPRM'</li><li>'SEMA4'</li><li>'SEMA6'</li><li>'SEMA7'</li><li>'VCAM'</li><li>'VISTA'</li><li>'GAP'</li><li>'PCDH'</li><li>'ADGRA'</li><li>'ADGRE'</li><li>'ADGRG'</li><li>'ADGRL'</li><li>'FLRT'</li><li>'UNC5'</li><li>'CNTN'</li><li>'Netrin'</li><li>'LIFR'</li><li>'SPP1'</li><li>'SLIT'</li><li>'AGRN'</li><li>'Prostaglandin'</li><li>'RA'</li><li>'CD34'</li><li>'SELE'</li><li>'EDN'</li><li>'DHEA'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TGFb'\n",
       "\\item 'BMP'\n",
       "\\item 'WNT'\n",
       "\\item 'ncWNT'\n",
       "\\item 'EGF'\n",
       "\\item 'NRG'\n",
       "\\item 'FGF'\n",
       "\\item 'PDGF'\n",
       "\\item 'VEGF'\n",
       "\\item 'IGF'\n",
       "\\item 'CCL'\n",
       "\\item 'CXCL'\n",
       "\\item 'MIF'\n",
       "\\item 'TWEAK'\n",
       "\\item 'VISFATIN'\n",
       "\\item 'ANGPTL'\n",
       "\\item 'ANGPT'\n",
       "\\item 'MK'\n",
       "\\item 'PTN'\n",
       "\\item 'NT'\n",
       "\\item 'SEMA3'\n",
       "\\item 'CALCR'\n",
       "\\item 'GAS'\n",
       "\\item 'GRN'\n",
       "\\item 'PROS'\n",
       "\\item 'PTPR'\n",
       "\\item 'IGFBP'\n",
       "\\item 'CypA'\n",
       "\\item 'COLLAGEN'\n",
       "\\item 'FN1'\n",
       "\\item 'LAMININ'\n",
       "\\item 'RELN'\n",
       "\\item 'THBS'\n",
       "\\item 'TENASCIN'\n",
       "\\item 'HSPG'\n",
       "\\item '27HC'\n",
       "\\item 'Androstenedione'\n",
       "\\item 'Androsterone'\n",
       "\\item 'Cholesterol'\n",
       "\\item 'Desmosterol'\n",
       "\\item 'Testosterone'\n",
       "\\item 'APP'\n",
       "\\item 'CADM'\n",
       "\\item 'CD46'\n",
       "\\item 'CD99'\n",
       "\\item 'CDH'\n",
       "\\item 'CDH5'\n",
       "\\item 'CLDN'\n",
       "\\item 'CSPG4'\n",
       "\\item 'EPHA'\n",
       "\\item 'EPHB'\n",
       "\\item 'ESAM'\n",
       "\\item 'JAM'\n",
       "\\item 'MAG'\n",
       "\\item 'MPZ'\n",
       "\\item 'NCAM'\n",
       "\\item 'NECTIN'\n",
       "\\item 'NEGR'\n",
       "\\item 'NOTCH'\n",
       "\\item 'NRXN'\n",
       "\\item 'PECAM1'\n",
       "\\item 'PTPRM'\n",
       "\\item 'SEMA4'\n",
       "\\item 'SEMA6'\n",
       "\\item 'SEMA7'\n",
       "\\item 'VCAM'\n",
       "\\item 'VISTA'\n",
       "\\item 'GAP'\n",
       "\\item 'PCDH'\n",
       "\\item 'ADGRA'\n",
       "\\item 'ADGRE'\n",
       "\\item 'ADGRG'\n",
       "\\item 'ADGRL'\n",
       "\\item 'FLRT'\n",
       "\\item 'UNC5'\n",
       "\\item 'CNTN'\n",
       "\\item 'Netrin'\n",
       "\\item 'LIFR'\n",
       "\\item 'SPP1'\n",
       "\\item 'SLIT'\n",
       "\\item 'AGRN'\n",
       "\\item 'Prostaglandin'\n",
       "\\item 'RA'\n",
       "\\item 'CD34'\n",
       "\\item 'SELE'\n",
       "\\item 'EDN'\n",
       "\\item 'DHEA'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TGFb'\n",
       "2. 'BMP'\n",
       "3. 'WNT'\n",
       "4. 'ncWNT'\n",
       "5. 'EGF'\n",
       "6. 'NRG'\n",
       "7. 'FGF'\n",
       "8. 'PDGF'\n",
       "9. 'VEGF'\n",
       "10. 'IGF'\n",
       "11. 'CCL'\n",
       "12. 'CXCL'\n",
       "13. 'MIF'\n",
       "14. 'TWEAK'\n",
       "15. 'VISFATIN'\n",
       "16. 'ANGPTL'\n",
       "17. 'ANGPT'\n",
       "18. 'MK'\n",
       "19. 'PTN'\n",
       "20. 'NT'\n",
       "21. 'SEMA3'\n",
       "22. 'CALCR'\n",
       "23. 'GAS'\n",
       "24. 'GRN'\n",
       "25. 'PROS'\n",
       "26. 'PTPR'\n",
       "27. 'IGFBP'\n",
       "28. 'CypA'\n",
       "29. 'COLLAGEN'\n",
       "30. 'FN1'\n",
       "31. 'LAMININ'\n",
       "32. 'RELN'\n",
       "33. 'THBS'\n",
       "34. 'TENASCIN'\n",
       "35. 'HSPG'\n",
       "36. '27HC'\n",
       "37. 'Androstenedione'\n",
       "38. 'Androsterone'\n",
       "39. 'Cholesterol'\n",
       "40. 'Desmosterol'\n",
       "41. 'Testosterone'\n",
       "42. 'APP'\n",
       "43. 'CADM'\n",
       "44. 'CD46'\n",
       "45. 'CD99'\n",
       "46. 'CDH'\n",
       "47. 'CDH5'\n",
       "48. 'CLDN'\n",
       "49. 'CSPG4'\n",
       "50. 'EPHA'\n",
       "51. 'EPHB'\n",
       "52. 'ESAM'\n",
       "53. 'JAM'\n",
       "54. 'MAG'\n",
       "55. 'MPZ'\n",
       "56. 'NCAM'\n",
       "57. 'NECTIN'\n",
       "58. 'NEGR'\n",
       "59. 'NOTCH'\n",
       "60. 'NRXN'\n",
       "61. 'PECAM1'\n",
       "62. 'PTPRM'\n",
       "63. 'SEMA4'\n",
       "64. 'SEMA6'\n",
       "65. 'SEMA7'\n",
       "66. 'VCAM'\n",
       "67. 'VISTA'\n",
       "68. 'GAP'\n",
       "69. 'PCDH'\n",
       "70. 'ADGRA'\n",
       "71. 'ADGRE'\n",
       "72. 'ADGRG'\n",
       "73. 'ADGRL'\n",
       "74. 'FLRT'\n",
       "75. 'UNC5'\n",
       "76. 'CNTN'\n",
       "77. 'Netrin'\n",
       "78. 'LIFR'\n",
       "79. 'SPP1'\n",
       "80. 'SLIT'\n",
       "81. 'AGRN'\n",
       "82. 'Prostaglandin'\n",
       "83. 'RA'\n",
       "84. 'CD34'\n",
       "85. 'SELE'\n",
       "86. 'EDN'\n",
       "87. 'DHEA'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"TGFb\"            \"BMP\"             \"WNT\"             \"ncWNT\"          \n",
       " [5] \"EGF\"             \"NRG\"             \"FGF\"             \"PDGF\"           \n",
       " [9] \"VEGF\"            \"IGF\"             \"CCL\"             \"CXCL\"           \n",
       "[13] \"MIF\"             \"TWEAK\"           \"VISFATIN\"        \"ANGPTL\"         \n",
       "[17] \"ANGPT\"           \"MK\"              \"PTN\"             \"NT\"             \n",
       "[21] \"SEMA3\"           \"CALCR\"           \"GAS\"             \"GRN\"            \n",
       "[25] \"PROS\"            \"PTPR\"            \"IGFBP\"           \"CypA\"           \n",
       "[29] \"COLLAGEN\"        \"FN1\"             \"LAMININ\"         \"RELN\"           \n",
       "[33] \"THBS\"            \"TENASCIN\"        \"HSPG\"            \"27HC\"           \n",
       "[37] \"Androstenedione\" \"Androsterone\"    \"Cholesterol\"     \"Desmosterol\"    \n",
       "[41] \"Testosterone\"    \"APP\"             \"CADM\"            \"CD46\"           \n",
       "[45] \"CD99\"            \"CDH\"             \"CDH5\"            \"CLDN\"           \n",
       "[49] \"CSPG4\"           \"EPHA\"            \"EPHB\"            \"ESAM\"           \n",
       "[53] \"JAM\"             \"MAG\"             \"MPZ\"             \"NCAM\"           \n",
       "[57] \"NECTIN\"          \"NEGR\"            \"NOTCH\"           \"NRXN\"           \n",
       "[61] \"PECAM1\"          \"PTPRM\"           \"SEMA4\"           \"SEMA6\"          \n",
       "[65] \"SEMA7\"           \"VCAM\"            \"VISTA\"           \"GAP\"            \n",
       "[69] \"PCDH\"            \"ADGRA\"           \"ADGRE\"           \"ADGRG\"          \n",
       "[73] \"ADGRL\"           \"FLRT\"            \"UNC5\"            \"CNTN\"           \n",
       "[77] \"Netrin\"          \"LIFR\"            \"SPP1\"            \"SLIT\"           \n",
       "[81] \"AGRN\"            \"Prostaglandin\"   \"RA\"              \"CD34\"           \n",
       "[85] \"SELE\"            \"EDN\"             \"DHEA\"           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"in all samples\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TGFb'</li><li>'WNT'</li><li>'EGF'</li><li>'NRG'</li><li>'FGF'</li><li>'PDGF'</li><li>'VEGF'</li><li>'IGF'</li><li>'CCL'</li><li>'CXCL'</li><li>'MIF'</li><li>'TWEAK'</li><li>'VISFATIN'</li><li>'ANGPTL'</li><li>'ANGPT'</li><li>'MK'</li><li>'PTN'</li><li>'SEMA3'</li><li>'GAS'</li><li>'GRN'</li><li>'PROS'</li><li>'PTPR'</li><li>'IGFBP'</li><li>'CypA'</li><li>'COLLAGEN'</li><li>'FN1'</li><li>'LAMININ'</li><li>'THBS'</li><li>'TENASCIN'</li><li>'HSPG'</li><li>'Androstenedione'</li><li>'Cholesterol'</li><li>'Desmosterol'</li><li>'Testosterone'</li><li>'APP'</li><li>'CADM'</li><li>'CD46'</li><li>'CD99'</li><li>'CDH'</li><li>'CDH5'</li><li>'CLDN'</li><li>'CSPG4'</li><li>'EPHA'</li><li>'EPHB'</li><li>'ESAM'</li><li>'JAM'</li><li>'MPZ'</li><li>'NCAM'</li><li>'NECTIN'</li><li>'NEGR'</li><li>'NOTCH'</li><li>'NRXN'</li><li>'PECAM1'</li><li>'PTPRM'</li><li>'SEMA4'</li><li>'SEMA6'</li><li>'VCAM'</li><li>'GAP'</li><li>'PCDH'</li><li>'ADGRA'</li><li>'ADGRE'</li><li>'ADGRG'</li><li>'ADGRL'</li><li>'FLRT'</li><li>'UNC5'</li><li>'CNTN'</li><li>'Netrin'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TGFb'\n",
       "\\item 'WNT'\n",
       "\\item 'EGF'\n",
       "\\item 'NRG'\n",
       "\\item 'FGF'\n",
       "\\item 'PDGF'\n",
       "\\item 'VEGF'\n",
       "\\item 'IGF'\n",
       "\\item 'CCL'\n",
       "\\item 'CXCL'\n",
       "\\item 'MIF'\n",
       "\\item 'TWEAK'\n",
       "\\item 'VISFATIN'\n",
       "\\item 'ANGPTL'\n",
       "\\item 'ANGPT'\n",
       "\\item 'MK'\n",
       "\\item 'PTN'\n",
       "\\item 'SEMA3'\n",
       "\\item 'GAS'\n",
       "\\item 'GRN'\n",
       "\\item 'PROS'\n",
       "\\item 'PTPR'\n",
       "\\item 'IGFBP'\n",
       "\\item 'CypA'\n",
       "\\item 'COLLAGEN'\n",
       "\\item 'FN1'\n",
       "\\item 'LAMININ'\n",
       "\\item 'THBS'\n",
       "\\item 'TENASCIN'\n",
       "\\item 'HSPG'\n",
       "\\item 'Androstenedione'\n",
       "\\item 'Cholesterol'\n",
       "\\item 'Desmosterol'\n",
       "\\item 'Testosterone'\n",
       "\\item 'APP'\n",
       "\\item 'CADM'\n",
       "\\item 'CD46'\n",
       "\\item 'CD99'\n",
       "\\item 'CDH'\n",
       "\\item 'CDH5'\n",
       "\\item 'CLDN'\n",
       "\\item 'CSPG4'\n",
       "\\item 'EPHA'\n",
       "\\item 'EPHB'\n",
       "\\item 'ESAM'\n",
       "\\item 'JAM'\n",
       "\\item 'MPZ'\n",
       "\\item 'NCAM'\n",
       "\\item 'NECTIN'\n",
       "\\item 'NEGR'\n",
       "\\item 'NOTCH'\n",
       "\\item 'NRXN'\n",
       "\\item 'PECAM1'\n",
       "\\item 'PTPRM'\n",
       "\\item 'SEMA4'\n",
       "\\item 'SEMA6'\n",
       "\\item 'VCAM'\n",
       "\\item 'GAP'\n",
       "\\item 'PCDH'\n",
       "\\item 'ADGRA'\n",
       "\\item 'ADGRE'\n",
       "\\item 'ADGRG'\n",
       "\\item 'ADGRL'\n",
       "\\item 'FLRT'\n",
       "\\item 'UNC5'\n",
       "\\item 'CNTN'\n",
       "\\item 'Netrin'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TGFb'\n",
       "2. 'WNT'\n",
       "3. 'EGF'\n",
       "4. 'NRG'\n",
       "5. 'FGF'\n",
       "6. 'PDGF'\n",
       "7. 'VEGF'\n",
       "8. 'IGF'\n",
       "9. 'CCL'\n",
       "10. 'CXCL'\n",
       "11. 'MIF'\n",
       "12. 'TWEAK'\n",
       "13. 'VISFATIN'\n",
       "14. 'ANGPTL'\n",
       "15. 'ANGPT'\n",
       "16. 'MK'\n",
       "17. 'PTN'\n",
       "18. 'SEMA3'\n",
       "19. 'GAS'\n",
       "20. 'GRN'\n",
       "21. 'PROS'\n",
       "22. 'PTPR'\n",
       "23. 'IGFBP'\n",
       "24. 'CypA'\n",
       "25. 'COLLAGEN'\n",
       "26. 'FN1'\n",
       "27. 'LAMININ'\n",
       "28. 'THBS'\n",
       "29. 'TENASCIN'\n",
       "30. 'HSPG'\n",
       "31. 'Androstenedione'\n",
       "32. 'Cholesterol'\n",
       "33. 'Desmosterol'\n",
       "34. 'Testosterone'\n",
       "35. 'APP'\n",
       "36. 'CADM'\n",
       "37. 'CD46'\n",
       "38. 'CD99'\n",
       "39. 'CDH'\n",
       "40. 'CDH5'\n",
       "41. 'CLDN'\n",
       "42. 'CSPG4'\n",
       "43. 'EPHA'\n",
       "44. 'EPHB'\n",
       "45. 'ESAM'\n",
       "46. 'JAM'\n",
       "47. 'MPZ'\n",
       "48. 'NCAM'\n",
       "49. 'NECTIN'\n",
       "50. 'NEGR'\n",
       "51. 'NOTCH'\n",
       "52. 'NRXN'\n",
       "53. 'PECAM1'\n",
       "54. 'PTPRM'\n",
       "55. 'SEMA4'\n",
       "56. 'SEMA6'\n",
       "57. 'VCAM'\n",
       "58. 'GAP'\n",
       "59. 'PCDH'\n",
       "60. 'ADGRA'\n",
       "61. 'ADGRE'\n",
       "62. 'ADGRG'\n",
       "63. 'ADGRL'\n",
       "64. 'FLRT'\n",
       "65. 'UNC5'\n",
       "66. 'CNTN'\n",
       "67. 'Netrin'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"TGFb\"            \"WNT\"             \"EGF\"             \"NRG\"            \n",
       " [5] \"FGF\"             \"PDGF\"            \"VEGF\"            \"IGF\"            \n",
       " [9] \"CCL\"             \"CXCL\"            \"MIF\"             \"TWEAK\"          \n",
       "[13] \"VISFATIN\"        \"ANGPTL\"          \"ANGPT\"           \"MK\"             \n",
       "[17] \"PTN\"             \"SEMA3\"           \"GAS\"             \"GRN\"            \n",
       "[21] \"PROS\"            \"PTPR\"            \"IGFBP\"           \"CypA\"           \n",
       "[25] \"COLLAGEN\"        \"FN1\"             \"LAMININ\"         \"THBS\"           \n",
       "[29] \"TENASCIN\"        \"HSPG\"            \"Androstenedione\" \"Cholesterol\"    \n",
       "[33] \"Desmosterol\"     \"Testosterone\"    \"APP\"             \"CADM\"           \n",
       "[37] \"CD46\"            \"CD99\"            \"CDH\"             \"CDH5\"           \n",
       "[41] \"CLDN\"            \"CSPG4\"           \"EPHA\"            \"EPHB\"           \n",
       "[45] \"ESAM\"            \"JAM\"             \"MPZ\"             \"NCAM\"           \n",
       "[49] \"NECTIN\"          \"NEGR\"            \"NOTCH\"           \"NRXN\"           \n",
       "[53] \"PECAM1\"          \"PTPRM\"           \"SEMA4\"           \"SEMA6\"          \n",
       "[57] \"VCAM\"            \"GAP\"             \"PCDH\"            \"ADGRA\"          \n",
       "[61] \"ADGRE\"           \"ADGRG\"           \"ADGRL\"           \"FLRT\"           \n",
       "[65] \"UNC5\"            \"CNTN\"            \"Netrin\"         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"fresh and slow\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'BMP'</li><li>'ncWNT'</li><li>'NT'</li><li>'CALCR'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'BMP'\n",
       "\\item 'ncWNT'\n",
       "\\item 'NT'\n",
       "\\item 'CALCR'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'BMP'\n",
       "2. 'ncWNT'\n",
       "3. 'NT'\n",
       "4. 'CALCR'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"BMP\"   \"ncWNT\" \"NT\"    \"CALCR\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"fresh and vitri\"\n"
     ]
    },
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"slow and vitri\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'SPP1'</li><li>'Prostaglandin'</li><li>'RA'</li><li>'CD34'</li><li>'SELE'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'SPP1'\n",
       "\\item 'Prostaglandin'\n",
       "\\item 'RA'\n",
       "\\item 'CD34'\n",
       "\\item 'SELE'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'SPP1'\n",
       "2. 'Prostaglandin'\n",
       "3. 'RA'\n",
       "4. 'CD34'\n",
       "5. 'SELE'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"SPP1\"          \"Prostaglandin\" \"RA\"            \"CD34\"         \n",
       "[5] \"SELE\"         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"pathways unique to fresh \"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'RELN'</li><li>'27HC'</li><li>'Androsterone'</li><li>'MAG'</li><li>'SEMA7'</li><li>'VISTA'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'RELN'\n",
       "\\item '27HC'\n",
       "\\item 'Androsterone'\n",
       "\\item 'MAG'\n",
       "\\item 'SEMA7'\n",
       "\\item 'VISTA'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'RELN'\n",
       "2. '27HC'\n",
       "3. 'Androsterone'\n",
       "4. 'MAG'\n",
       "5. 'SEMA7'\n",
       "6. 'VISTA'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"RELN\"         \"27HC\"         \"Androsterone\" \"MAG\"          \"SEMA7\"       \n",
       "[6] \"VISTA\"       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"pathways unique to slow \"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'LIFR'</li><li>'SLIT'</li><li>'AGRN'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'LIFR'\n",
       "\\item 'SLIT'\n",
       "\\item 'AGRN'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'LIFR'\n",
       "2. 'SLIT'\n",
       "3. 'AGRN'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"LIFR\" \"SLIT\" \"AGRN\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"pathways unique to vitri\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'EDN'</li><li>'DHEA'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'EDN'\n",
       "\\item 'DHEA'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'EDN'\n",
       "2. 'DHEA'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"EDN\"  \"DHEA\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('union of all samples')\n",
    "union(union(unique(df$fresh$pathway_name), unique(df$slow$pathway_name)), unique(df$vitri$pathway_name))\n",
    "print('in all samples')\n",
    "all = intersect(intersect(unique(df$fresh$pathway_name), unique(df$slow$pathway_name)), unique(df$vitri$pathway_name))\n",
    "all\n",
    "print('fresh and slow')\n",
    "setdiff(intersect(unique(df$fresh$pathway_name), unique(df$slow$pathway_name)), all)\n",
    "print('fresh and vitri')\n",
    "setdiff(intersect(unique(df$fresh$pathway_name), unique(df$vitri$pathway_name)), all)\n",
    "print('slow and vitri')\n",
    "setdiff(intersect(unique(df$slow$pathway_name), unique(df$vitri$pathway_name)), all)\n",
    "\n",
    "\n",
    "# pathways unique to each slice\n",
    "print('pathways unique to fresh ')\n",
    "setdiff(unique(df$fresh$pathway_name), union(unique(df$slow$pathway_name), unique(df$vitri$pathway_name)))\n",
    "print('pathways unique to slow ')\n",
    "setdiff(unique(df$slow$pathway_name), union(unique(df$fresh$pathway_name), unique(df$vitri$pathway_name)))\n",
    "print('pathways unique to vitri')\n",
    "setdiff(unique(df$vitri$pathway_name), union(unique(df$fresh$pathway_name), unique(df$slow$pathway_name)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"union of all samples\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TGFB1_TGFBR1_TGFBR2'</li><li>'TGFB2_TGFBR1_TGFBR2'</li><li>'TGFB1_ACVR1_TGFBR1'</li><li>'TGFB2_ACVR1_TGFBR1'</li><li>'BMP6_ACVR1_BMPR2'</li><li>'BMP6_BMPR1A_BMPR2'</li><li>'BMP6_BMPR1B_BMPR2'</li><li>'WNT4_FZD1_LRP5'</li><li>'WNT4_FZD3_LRP5'</li><li>'WNT4_FZD1_LRP6'</li><li>'WNT4_FZD3_LRP6'</li><li>'WNT4_FZD7_LRP6'</li><li>'WNT4_FZD8_LRP6'</li><li>'WNT5B_FZD1'</li><li>'WNT5B_FZD3'</li><li>'WNT5B_FZD7'</li><li>'WNT5B_FZD8'</li><li>'AREG_EGFR'</li><li>'AREG_EGFR_ERBB2'</li><li>'NRG1_ERBB4'</li><li>'FGF2_FGFR1'</li><li>'FGF2_FGFR2'</li><li>'FGF7_FGFR1'</li><li>'FGF7_FGFR2'</li><li>'PDGFB_PDGFRA'</li><li>'PDGFB_PDGFRB'</li><li>'PDGFC_PDGFRA'</li><li>'PDGFD_PDGFRB'</li><li>'VEGFB_VEGFR1'</li><li>'VEGFC_VEGFR2'</li><li>'PGF_VEGFR1'</li><li>'IGF1_IGF1R'</li><li>'IGF2_IGF1R'</li><li>'IGF2_IGF2R'</li><li>'CCL2_ACKR1'</li><li>'CCL5_ACKR1'</li><li>'CCL14_ACKR1'</li><li>'CXCL8_ACKR1'</li><li>'CXCL12_CXCR4'</li><li>'MIF_CD74_CXCR4'</li><li>'MIF_CD74_CD44'</li><li>'TNFSF12_TNFRSF12A'</li><li>'NAMPT_ITGA5_ITGB1'</li><li>'ANGPTL1_ITGA1_ITGB1'</li><li>'ANGPTL2_ITGA5_ITGB1'</li><li>'ANGPT2_TEK'</li><li>'ANGPT2_ITGA5_ITGB1'</li><li>'MDK_SDC2'</li><li>'MDK_SDC4'</li><li>'MDK_ITGA6_ITGB1'</li><li>'MDK_LRP1'</li><li>'MDK_NCL'</li><li>'PTN_SDC2'</li><li>'PTN_SDC4'</li><li>'PTN_NCL'</li><li>'NTF3_NTRK2'</li><li>'NTF3_NTRK3'</li><li>'SEMA3A_NRP1_PLXNA2'</li><li>'SEMA3A_NRP1_PLXNA4'</li><li>'SEMA3C_NRP1_PLXNA2'</li><li>'SEMA3C_NRP1_PLXNA4'</li><li>'SEMA3C_NRP2_PLXNA2'</li><li>'SEMA3C_NRP2_PLXNA4'</li><li>'SEMA3G_NRP2_PLXNA2'</li><li>'SEMA3G_NRP2_PLXNA4'</li><li>'SEMA3C_NRP1_NRP2'</li><li>'SEMA3C_PLXND1'</li><li>'ADM_CALCRL'</li><li>'GAS6_AXL'</li><li>'GAS6_TYRO3'</li><li>'GRN_SORT1'</li><li>'PROS1_AXL'</li><li>'PROS1_TYRO3'</li><li>'PTPRS_NTRK3'</li><li>'IGFBP3_TMEM219'</li><li>'PPIA_BSG'</li><li>'COL1A1_ITGA1_ITGB1'</li><li>'COL1A2_ITGA1_ITGB1'</li><li>'COL4A1_ITGA1_ITGB1'</li><li>'COL4A2_ITGA1_ITGB1'</li><li>'COL4A5_ITGA1_ITGB1'</li><li>'COL6A1_ITGA1_ITGB1'</li><li>'COL6A2_ITGA1_ITGB1'</li><li>'COL6A3_ITGA1_ITGB1'</li><li>'FN1_ITGA3_ITGB1'</li><li>'FN1_ITGA5_ITGB1'</li><li>'FN1_ITGA8_ITGB1'</li><li>'LAMA1_ITGA1_ITGB1'</li><li>'LAMA2_ITGA1_ITGB1'</li><li>'LAMA3_ITGA1_ITGB1'</li><li>'LAMA4_ITGA1_ITGB1'</li><li>'LAMA5_ITGA1_ITGB1'</li><li>'LAMB1_ITGA1_ITGB1'</li><li>'LAMB2_ITGA1_ITGB1'</li><li>'LAMC1_ITGA1_ITGB1'</li><li>'LAMA1_ITGA3_ITGB1'</li><li>'LAMA2_ITGA3_ITGB1'</li><li>'LAMA3_ITGA3_ITGB1'</li><li>'LAMA4_ITGA3_ITGB1'</li><li>'LAMA5_ITGA3_ITGB1'</li><li>'LAMB1_ITGA3_ITGB1'</li><li>'LAMB2_ITGA3_ITGB1'</li><li>'LAMC1_ITGA3_ITGB1'</li><li>'RELN_ITGA3_ITGB1'</li><li>'THBS1_ITGA3_ITGB1'</li><li>'THBS2_ITGA3_ITGB1'</li><li>'COL1A1_ITGA3_ITGB1'</li><li>'COL1A2_ITGA3_ITGB1'</li><li>'COL4A1_ITGA3_ITGB1'</li><li>'COL4A2_ITGA3_ITGB1'</li><li>'COL4A5_ITGA3_ITGB1'</li><li>'COL6A1_ITGA3_ITGB1'</li><li>'COL6A2_ITGA3_ITGB1'</li><li>'COL6A3_ITGA3_ITGB1'</li><li>'LAMA1_ITGA6_ITGB1'</li><li>'LAMA2_ITGA6_ITGB1'</li><li>'LAMA3_ITGA6_ITGB1'</li><li>'LAMA4_ITGA6_ITGB1'</li><li>'LAMA5_ITGA6_ITGB1'</li><li>'LAMB1_ITGA6_ITGB1'</li><li>'LAMB2_ITGA6_ITGB1'</li><li>'LAMC1_ITGA6_ITGB1'</li><li>'LAMA1_ITGA7_ITGB1'</li><li>'LAMA2_ITGA7_ITGB1'</li><li>'LAMA3_ITGA7_ITGB1'</li><li>'LAMA4_ITGA7_ITGB1'</li><li>'LAMA5_ITGA7_ITGB1'</li><li>'LAMB1_ITGA7_ITGB1'</li><li>'LAMB2_ITGA7_ITGB1'</li><li>'LAMC1_ITGA7_ITGB1'</li><li>'TNXB_ITGA8_ITGB1'</li><li>'TNXB_ITGA9_ITGB1'</li><li>'COL1A1_ITGA9_ITGB1'</li><li>'COL1A2_ITGA9_ITGB1'</li><li>'COL4A1_ITGA9_ITGB1'</li><li>'COL4A2_ITGA9_ITGB1'</li><li>'COL4A5_ITGA9_ITGB1'</li><li>'COL6A1_ITGA9_ITGB1'</li><li>'COL6A2_ITGA9_ITGB1'</li><li>'COL6A3_ITGA9_ITGB1'</li><li>'LAMA1_ITGA9_ITGB1'</li><li>'LAMA2_ITGA9_ITGB1'</li><li>'LAMA3_ITGA9_ITGB1'</li><li>'LAMA4_ITGA9_ITGB1'</li><li>'LAMA5_ITGA9_ITGB1'</li><li>'LAMB1_ITGA9_ITGB1'</li><li>'LAMB2_ITGA9_ITGB1'</li><li>'LAMC1_ITGA9_ITGB1'</li><li>'FN1_ITGAV_ITGB1'</li><li>'FN1_ITGAV_ITGB8'</li><li>'LAMA1_ITGAV_ITGB8'</li><li>'LAMA2_ITGAV_ITGB8'</li><li>'LAMA3_ITGAV_ITGB8'</li><li>'LAMA4_ITGAV_ITGB8'</li><li>'LAMA5_ITGAV_ITGB8'</li><li>'LAMB1_ITGAV_ITGB8'</li><li>'LAMB2_ITGAV_ITGB8'</li><li>'LAMC1_ITGAV_ITGB8'</li><li>'COL1A1_ITGAV_ITGB8'</li><li>'COL1A2_ITGAV_ITGB8'</li><li>'COL4A1_ITGAV_ITGB8'</li><li>'COL4A2_ITGAV_ITGB8'</li><li>'COL4A5_ITGAV_ITGB8'</li><li>'COL6A1_ITGAV_ITGB8'</li><li>'COL6A2_ITGAV_ITGB8'</li><li>'COL6A3_ITGAV_ITGB8'</li><li>'FN1_CD44'</li><li>'COL1A1_CD44'</li><li>'COL1A2_CD44'</li><li>'COL4A1_CD44'</li><li>'COL4A2_CD44'</li><li>'COL4A5_CD44'</li><li>'COL6A1_CD44'</li><li>'COL6A2_CD44'</li><li>'COL6A3_CD44'</li><li>'LAMA1_CD44'</li><li>'LAMA2_CD44'</li><li>'LAMA3_CD44'</li><li>'LAMA4_CD44'</li><li>'LAMA5_CD44'</li><li>'LAMB1_CD44'</li><li>'LAMB2_CD44'</li><li>'LAMC1_CD44'</li><li>'COL1A1_SDC4'</li><li>'COL1A2_SDC4'</li><li>'COL4A1_SDC4'</li><li>'COL4A2_SDC4'</li><li>'COL4A5_SDC4'</li><li>'COL6A1_SDC4'</li><li>'COL6A2_SDC4'</li><li>'COL6A3_SDC4'</li><li>'FN1_SDC4'</li><li>'TNXB_SDC4'</li><li>'THBS1_SDC4'</li><li>'THBS2_SDC4'</li><li>'THBS1_CD36'</li><li>'THBS2_CD36'</li><li>'THBS1_CD47'</li><li>'THBS2_CD47'</li><li>'HSPG2_DAG1'</li><li>⋯</li><li>'EFNB1_EPHB4'</li><li>'EFNB2_EPHA4'</li><li>'EFNB2_EPHB1'</li><li>'EFNB2_EPHB2'</li><li>'EFNB2_EPHB4'</li><li>'ESAM_ESAM'</li><li>'JAM2_ITGA3_ITGB1'</li><li>'JAM2_ITGAV_ITGB1'</li><li>'JAM2_JAM3'</li><li>'F11R_F11R'</li><li>'JAM2_F11R'</li><li>'JAM2_JAM2'</li><li>'JAM3_F11R'</li><li>'JAM3_JAM3'</li><li>'MAG_MAG'</li><li>'MPZL1_MPZL1'</li><li>'NCAM1_FGFR1'</li><li>'NCAM1_NCAM1'</li><li>'NECTIN3_NECTIN2'</li><li>'NEGR1_NEGR1'</li><li>'DLK1_NOTCH1'</li><li>'DLK1_NOTCH2'</li><li>'DLK1_NOTCH3'</li><li>'DLK1_NOTCH4'</li><li>'DLL4_NOTCH1'</li><li>'DLL4_NOTCH2'</li><li>'DLL4_NOTCH3'</li><li>'DLL4_NOTCH4'</li><li>'JAG1_NOTCH1'</li><li>'JAG1_NOTCH2'</li><li>'JAG1_NOTCH3'</li><li>'JAG1_NOTCH4'</li><li>'NRXN3_NLGN1'</li><li>'NRXN3_NLGN2'</li><li>'PECAM1_PECAM1'</li><li>'PTPRM_PTPRM'</li><li>'SEMA4C_PLXNB2'</li><li>'SEMA6A_PLXNA2'</li><li>'SEMA6A_PLXNA4'</li><li>'SEMA7A_ITGB1_ITGA1'</li><li>'SEMA7A_PLXNC1'</li><li>'ITGA9_ITGB1_VCAM1'</li><li>'VSIR_IGSF11'</li><li>'CDH3_CDH3'</li><li>'GJA5_GJA5'</li><li>'GJA1_GJA5'</li><li>'GJC1_GJA5'</li><li>'GJA1_GJA1'</li><li>'GJC1_GJA1'</li><li>'GJC1_GJC1'</li><li>'PCDHGA1_PCDHGA1'</li><li>'SDC2_ADGRA2'</li><li>'CD55_ADGRE5'</li><li>'LPAR1_ADGRE5'</li><li>'THY1_ADGRE5'</li><li>'ITGAV_ITGB1_ADGRE5'</li><li>'COL4A1_ADGRG6'</li><li>'COL4A2_ADGRG6'</li><li>'COL4A5_ADGRG6'</li><li>'LAMA2_ADGRG6'</li><li>'PRNP_ADGRG6'</li><li>'FLRT2_ADGRL2'</li><li>'TENM2_ADGRL2'</li><li>'TENM3_ADGRL2'</li><li>'TENM4_ADGRL2'</li><li>'FLRT2_ADGRL3'</li><li>'TENM2_ADGRL3'</li><li>'TENM3_ADGRL3'</li><li>'TENM4_ADGRL3'</li><li>'FLRT2_FLRT2'</li><li>'FLRT2_UNC5C'</li><li>'Q9HDB5_CLSTN1'</li><li>'Q9HDB5_CLSTN2'</li><li>'NRXN3_LRRTM3'</li><li>'APP_SORL1'</li><li>'CADM1_NECTIN3'</li><li>'CD99_PILRB'</li><li>'CNTN1_NOTCH1'</li><li>'CNTN1_NOTCH2'</li><li>'NTN1_DCC'</li><li>'NTN1_DSCAM'</li><li>'NTN1_UNC5C'</li><li>'NTN4_NTRK2'</li><li>'BMP6_ACVR1_ACVR2A'</li><li>'WNT2B_FZD1_LRP5'</li><li>'WNT2B_FZD6_LRP5'</li><li>'WNT2B_FZD7_LRP5'</li><li>'WNT2B_FZD8_LRP5'</li><li>'WNT8A_FZD6_LRP5'</li><li>'WNT2B_FZD1_LRP6'</li><li>'WNT2B_FZD6_LRP6'</li><li>'WNT2B_FZD8_LRP6'</li><li>'WNT8A_FZD6_LRP6'</li><li>'WNT5B_FZD6'</li><li>'NRG3_ERBB4'</li><li>'FGF2_FGFR4'</li><li>'PDGFA_PDGFRA'</li><li>'PDGFA_PDGFRB'</li><li>'VEGFA_VEGFR1R2'</li><li>'CCL8_ACKR1'</li><li>'CXCL12_ACKR3'</li><li>'MIF_ACKR3'</li><li>'CTF1_LIFR_IL6ST'</li><li>'SPP1_CD44'</li><li>'SPP1_ITGAV_ITGB1'</li><li>'SPP1_ITGAV_ITGB5'</li><li>'SPP1_ITGA9_ITGB1'</li><li>'SPP1_ITGA5_ITGB1'</li><li>'ANGPTL2_TLR4'</li><li>'PTN_SDC3'</li><li>'SEMA3C_NRP1_PLXNA1'</li><li>'SEMA3C_NRP2_PLXNA1'</li><li>'SEMA3F_NRP2_PLXNA1'</li><li>'SEMA3F_NRP2_PLXNA4'</li><li>'GAS6_MERTK'</li><li>'SLIT2_ROBO1'</li><li>'SLIT2_ROBO2'</li><li>'PROS1_MERTK'</li><li>'THBS4_ITGA3_ITGB1'</li><li>'COL6A6_ITGA9_ITGB1'</li><li>'LAMC2_ITGA9_ITGB1'</li><li>'LAMC2_ITGAV_ITGB8'</li><li>'COL6A6_ITGAV_ITGB8'</li><li>'COL6A6_CD44'</li><li>'COL6A6_SDC4'</li><li>'THBS4_SDC4'</li><li>'LAMA2_SV2A'</li><li>'LAMA4_SV2A'</li><li>'LAMB2_SV2A'</li><li>'LAMC1_SV2A'</li><li>'LAMC2_SV2A'</li><li>'THBS4_CD36'</li><li>'THBS4_CD47'</li><li>'AGRN_DAG1'</li><li>'LAMC2_DAG1'</li><li>'PGF2a-AKR1C3_PTGFR'</li><li>'PGF2a-CBR1_PTGFR'</li><li>'RetinoicAcid-RA-ALDH1A1_RARA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A1_RARA_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A1_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A2_RARA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A2_RARA_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A2_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A3_RARA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A3_RARA_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A3_RXRA_CRABP2'</li><li>'CD34_SELP'</li><li>'CNTN1_NRCAM'</li><li>'NCAM1_NCAM2'</li><li>'NECTIN3_NECTIN1'</li><li>'NECTIN3_PVR'</li><li>'JAG2_NOTCH1'</li><li>'JAG2_NOTCH2'</li><li>'JAG2_NOTCH3'</li><li>'NRXN1_NLGN1'</li><li>'NRXN1_NLGN2'</li><li>'NRXN2_NLGN2'</li><li>'SELE_CD44'</li><li>'SELE_GLG1'</li><li>'SEMA4D_PLXNB1'</li><li>'SEMA4D_PLXNB2'</li><li>'SEMA4G_PLXNB2'</li><li>'CDH12_ITGA1_ITGB1'</li><li>'COL3A1_ADGRG1'</li><li>'TGM2_ADGRG1'</li><li>'FLRT2_UNC5B'</li><li>'NRXN1_CLSTN1'</li><li>'NRXN1_CLSTN2'</li><li>'NRXN1_DAG1'</li><li>'NRXN2_DAG1'</li><li>'LRFN5_PTPRD'</li><li>'LRFN5_PTPRF'</li><li>'LRFN5_PTPRS'</li><li>'LRRC4C_PTPRF'</li><li>'NRXN1_LRRTM4'</li><li>'NRXN3_LRRTM4'</li><li>'LRRTM4_PTPRS'</li><li>'NTN4_UNC5B'</li><li>'TGFB3_TGFBR1_TGFBR2'</li><li>'TGFB3_ACVR1_TGFBR1'</li><li>'SPP1_ITGA8_ITGB1'</li><li>'ANGPTL4_ITGA5_ITGB1'</li><li>'ANGPTL4_CDH11'</li><li>'ANGPTL4_SDC2'</li><li>'ANGPTL4_SDC4'</li><li>'EDN1_EDNRA'</li><li>'EDN1_EDNRB'</li><li>'PGE2-PTGES3_PTGER2'</li><li>'RetinoicAcid-RA-ALDH1A1_RORB'</li><li>'RetinoicAcid-RA-ALDH1A2_RORB'</li><li>'RetinoicAcid-RA-ALDH1A3_RORB'</li><li>'Dehydroepiandrosterone-DHEA-STS_ESR1'</li><li>'CDH1_CDH1'</li><li>'FLRT3_ADGRL2'</li><li>'FLRT3_ADGRL3'</li><li>'FLRT3_FLRT3'</li><li>'FLRT3_UNC5C'</li><li>'NRXN2_CLSTN1'</li><li>'NRXN2_CLSTN2'</li><li>'NRXN1_LRRTM3'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TGFB1\\_TGFBR1\\_TGFBR2'\n",
       "\\item 'TGFB2\\_TGFBR1\\_TGFBR2'\n",
       "\\item 'TGFB1\\_ACVR1\\_TGFBR1'\n",
       "\\item 'TGFB2\\_ACVR1\\_TGFBR1'\n",
       "\\item 'BMP6\\_ACVR1\\_BMPR2'\n",
       "\\item 'BMP6\\_BMPR1A\\_BMPR2'\n",
       "\\item 'BMP6\\_BMPR1B\\_BMPR2'\n",
       "\\item 'WNT4\\_FZD1\\_LRP5'\n",
       "\\item 'WNT4\\_FZD3\\_LRP5'\n",
       "\\item 'WNT4\\_FZD1\\_LRP6'\n",
       "\\item 'WNT4\\_FZD3\\_LRP6'\n",
       "\\item 'WNT4\\_FZD7\\_LRP6'\n",
       "\\item 'WNT4\\_FZD8\\_LRP6'\n",
       "\\item 'WNT5B\\_FZD1'\n",
       "\\item 'WNT5B\\_FZD3'\n",
       "\\item 'WNT5B\\_FZD7'\n",
       "\\item 'WNT5B\\_FZD8'\n",
       "\\item 'AREG\\_EGFR'\n",
       "\\item 'AREG\\_EGFR\\_ERBB2'\n",
       "\\item 'NRG1\\_ERBB4'\n",
       "\\item 'FGF2\\_FGFR1'\n",
       "\\item 'FGF2\\_FGFR2'\n",
       "\\item 'FGF7\\_FGFR1'\n",
       "\\item 'FGF7\\_FGFR2'\n",
       "\\item 'PDGFB\\_PDGFRA'\n",
       "\\item 'PDGFB\\_PDGFRB'\n",
       "\\item 'PDGFC\\_PDGFRA'\n",
       "\\item 'PDGFD\\_PDGFRB'\n",
       "\\item 'VEGFB\\_VEGFR1'\n",
       "\\item 'VEGFC\\_VEGFR2'\n",
       "\\item 'PGF\\_VEGFR1'\n",
       "\\item 'IGF1\\_IGF1R'\n",
       "\\item 'IGF2\\_IGF1R'\n",
       "\\item 'IGF2\\_IGF2R'\n",
       "\\item 'CCL2\\_ACKR1'\n",
       "\\item 'CCL5\\_ACKR1'\n",
       "\\item 'CCL14\\_ACKR1'\n",
       "\\item 'CXCL8\\_ACKR1'\n",
       "\\item 'CXCL12\\_CXCR4'\n",
       "\\item 'MIF\\_CD74\\_CXCR4'\n",
       "\\item 'MIF\\_CD74\\_CD44'\n",
       "\\item 'TNFSF12\\_TNFRSF12A'\n",
       "\\item 'NAMPT\\_ITGA5\\_ITGB1'\n",
       "\\item 'ANGPTL1\\_ITGA1\\_ITGB1'\n",
       "\\item 'ANGPTL2\\_ITGA5\\_ITGB1'\n",
       "\\item 'ANGPT2\\_TEK'\n",
       "\\item 'ANGPT2\\_ITGA5\\_ITGB1'\n",
       "\\item 'MDK\\_SDC2'\n",
       "\\item 'MDK\\_SDC4'\n",
       "\\item 'MDK\\_ITGA6\\_ITGB1'\n",
       "\\item 'MDK\\_LRP1'\n",
       "\\item 'MDK\\_NCL'\n",
       "\\item 'PTN\\_SDC2'\n",
       "\\item 'PTN\\_SDC4'\n",
       "\\item 'PTN\\_NCL'\n",
       "\\item 'NTF3\\_NTRK2'\n",
       "\\item 'NTF3\\_NTRK3'\n",
       "\\item 'SEMA3A\\_NRP1\\_PLXNA2'\n",
       "\\item 'SEMA3A\\_NRP1\\_PLXNA4'\n",
       "\\item 'SEMA3C\\_NRP1\\_PLXNA2'\n",
       "\\item 'SEMA3C\\_NRP1\\_PLXNA4'\n",
       "\\item 'SEMA3C\\_NRP2\\_PLXNA2'\n",
       "\\item 'SEMA3C\\_NRP2\\_PLXNA4'\n",
       "\\item 'SEMA3G\\_NRP2\\_PLXNA2'\n",
       "\\item 'SEMA3G\\_NRP2\\_PLXNA4'\n",
       "\\item 'SEMA3C\\_NRP1\\_NRP2'\n",
       "\\item 'SEMA3C\\_PLXND1'\n",
       "\\item 'ADM\\_CALCRL'\n",
       "\\item 'GAS6\\_AXL'\n",
       "\\item 'GAS6\\_TYRO3'\n",
       "\\item 'GRN\\_SORT1'\n",
       "\\item 'PROS1\\_AXL'\n",
       "\\item 'PROS1\\_TYRO3'\n",
       "\\item 'PTPRS\\_NTRK3'\n",
       "\\item 'IGFBP3\\_TMEM219'\n",
       "\\item 'PPIA\\_BSG'\n",
       "\\item 'COL1A1\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL1A2\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL4A1\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL4A2\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL4A5\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL6A1\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL6A2\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL6A3\\_ITGA1\\_ITGB1'\n",
       "\\item 'FN1\\_ITGA3\\_ITGB1'\n",
       "\\item 'FN1\\_ITGA5\\_ITGB1'\n",
       "\\item 'FN1\\_ITGA8\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA3\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA3\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA3\\_ITGB1'\n",
       "\\item 'RELN\\_ITGA3\\_ITGB1'\n",
       "\\item 'THBS1\\_ITGA3\\_ITGB1'\n",
       "\\item 'THBS2\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL1A1\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL1A2\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL4A1\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL4A2\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL4A5\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL6A1\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL6A2\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL6A3\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA3\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMA3\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA7\\_ITGB1'\n",
       "\\item 'TNXB\\_ITGA8\\_ITGB1'\n",
       "\\item 'TNXB\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL1A1\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL1A2\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL4A1\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL4A2\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL4A5\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL6A1\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL6A2\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL6A3\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA3\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA9\\_ITGB1'\n",
       "\\item 'FN1\\_ITGAV\\_ITGB1'\n",
       "\\item 'FN1\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMA1\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMA2\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMA3\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMA4\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMA5\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMB1\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMB2\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMC1\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL1A1\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL1A2\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL4A1\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL4A2\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL4A5\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A1\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A2\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A3\\_ITGAV\\_ITGB8'\n",
       "\\item 'FN1\\_CD44'\n",
       "\\item 'COL1A1\\_CD44'\n",
       "\\item 'COL1A2\\_CD44'\n",
       "\\item 'COL4A1\\_CD44'\n",
       "\\item 'COL4A2\\_CD44'\n",
       "\\item 'COL4A5\\_CD44'\n",
       "\\item 'COL6A1\\_CD44'\n",
       "\\item 'COL6A2\\_CD44'\n",
       "\\item 'COL6A3\\_CD44'\n",
       "\\item 'LAMA1\\_CD44'\n",
       "\\item 'LAMA2\\_CD44'\n",
       "\\item 'LAMA3\\_CD44'\n",
       "\\item 'LAMA4\\_CD44'\n",
       "\\item 'LAMA5\\_CD44'\n",
       "\\item 'LAMB1\\_CD44'\n",
       "\\item 'LAMB2\\_CD44'\n",
       "\\item 'LAMC1\\_CD44'\n",
       "\\item 'COL1A1\\_SDC4'\n",
       "\\item 'COL1A2\\_SDC4'\n",
       "\\item 'COL4A1\\_SDC4'\n",
       "\\item 'COL4A2\\_SDC4'\n",
       "\\item 'COL4A5\\_SDC4'\n",
       "\\item 'COL6A1\\_SDC4'\n",
       "\\item 'COL6A2\\_SDC4'\n",
       "\\item 'COL6A3\\_SDC4'\n",
       "\\item 'FN1\\_SDC4'\n",
       "\\item 'TNXB\\_SDC4'\n",
       "\\item 'THBS1\\_SDC4'\n",
       "\\item 'THBS2\\_SDC4'\n",
       "\\item 'THBS1\\_CD36'\n",
       "\\item 'THBS2\\_CD36'\n",
       "\\item 'THBS1\\_CD47'\n",
       "\\item 'THBS2\\_CD47'\n",
       "\\item 'HSPG2\\_DAG1'\n",
       "\\item ⋯\n",
       "\\item 'EFNB1\\_EPHB4'\n",
       "\\item 'EFNB2\\_EPHA4'\n",
       "\\item 'EFNB2\\_EPHB1'\n",
       "\\item 'EFNB2\\_EPHB2'\n",
       "\\item 'EFNB2\\_EPHB4'\n",
       "\\item 'ESAM\\_ESAM'\n",
       "\\item 'JAM2\\_ITGA3\\_ITGB1'\n",
       "\\item 'JAM2\\_ITGAV\\_ITGB1'\n",
       "\\item 'JAM2\\_JAM3'\n",
       "\\item 'F11R\\_F11R'\n",
       "\\item 'JAM2\\_F11R'\n",
       "\\item 'JAM2\\_JAM2'\n",
       "\\item 'JAM3\\_F11R'\n",
       "\\item 'JAM3\\_JAM3'\n",
       "\\item 'MAG\\_MAG'\n",
       "\\item 'MPZL1\\_MPZL1'\n",
       "\\item 'NCAM1\\_FGFR1'\n",
       "\\item 'NCAM1\\_NCAM1'\n",
       "\\item 'NECTIN3\\_NECTIN2'\n",
       "\\item 'NEGR1\\_NEGR1'\n",
       "\\item 'DLK1\\_NOTCH1'\n",
       "\\item 'DLK1\\_NOTCH2'\n",
       "\\item 'DLK1\\_NOTCH3'\n",
       "\\item 'DLK1\\_NOTCH4'\n",
       "\\item 'DLL4\\_NOTCH1'\n",
       "\\item 'DLL4\\_NOTCH2'\n",
       "\\item 'DLL4\\_NOTCH3'\n",
       "\\item 'DLL4\\_NOTCH4'\n",
       "\\item 'JAG1\\_NOTCH1'\n",
       "\\item 'JAG1\\_NOTCH2'\n",
       "\\item 'JAG1\\_NOTCH3'\n",
       "\\item 'JAG1\\_NOTCH4'\n",
       "\\item 'NRXN3\\_NLGN1'\n",
       "\\item 'NRXN3\\_NLGN2'\n",
       "\\item 'PECAM1\\_PECAM1'\n",
       "\\item 'PTPRM\\_PTPRM'\n",
       "\\item 'SEMA4C\\_PLXNB2'\n",
       "\\item 'SEMA6A\\_PLXNA2'\n",
       "\\item 'SEMA6A\\_PLXNA4'\n",
       "\\item 'SEMA7A\\_ITGB1\\_ITGA1'\n",
       "\\item 'SEMA7A\\_PLXNC1'\n",
       "\\item 'ITGA9\\_ITGB1\\_VCAM1'\n",
       "\\item 'VSIR\\_IGSF11'\n",
       "\\item 'CDH3\\_CDH3'\n",
       "\\item 'GJA5\\_GJA5'\n",
       "\\item 'GJA1\\_GJA5'\n",
       "\\item 'GJC1\\_GJA5'\n",
       "\\item 'GJA1\\_GJA1'\n",
       "\\item 'GJC1\\_GJA1'\n",
       "\\item 'GJC1\\_GJC1'\n",
       "\\item 'PCDHGA1\\_PCDHGA1'\n",
       "\\item 'SDC2\\_ADGRA2'\n",
       "\\item 'CD55\\_ADGRE5'\n",
       "\\item 'LPAR1\\_ADGRE5'\n",
       "\\item 'THY1\\_ADGRE5'\n",
       "\\item 'ITGAV\\_ITGB1\\_ADGRE5'\n",
       "\\item 'COL4A1\\_ADGRG6'\n",
       "\\item 'COL4A2\\_ADGRG6'\n",
       "\\item 'COL4A5\\_ADGRG6'\n",
       "\\item 'LAMA2\\_ADGRG6'\n",
       "\\item 'PRNP\\_ADGRG6'\n",
       "\\item 'FLRT2\\_ADGRL2'\n",
       "\\item 'TENM2\\_ADGRL2'\n",
       "\\item 'TENM3\\_ADGRL2'\n",
       "\\item 'TENM4\\_ADGRL2'\n",
       "\\item 'FLRT2\\_ADGRL3'\n",
       "\\item 'TENM2\\_ADGRL3'\n",
       "\\item 'TENM3\\_ADGRL3'\n",
       "\\item 'TENM4\\_ADGRL3'\n",
       "\\item 'FLRT2\\_FLRT2'\n",
       "\\item 'FLRT2\\_UNC5C'\n",
       "\\item 'Q9HDB5\\_CLSTN1'\n",
       "\\item 'Q9HDB5\\_CLSTN2'\n",
       "\\item 'NRXN3\\_LRRTM3'\n",
       "\\item 'APP\\_SORL1'\n",
       "\\item 'CADM1\\_NECTIN3'\n",
       "\\item 'CD99\\_PILRB'\n",
       "\\item 'CNTN1\\_NOTCH1'\n",
       "\\item 'CNTN1\\_NOTCH2'\n",
       "\\item 'NTN1\\_DCC'\n",
       "\\item 'NTN1\\_DSCAM'\n",
       "\\item 'NTN1\\_UNC5C'\n",
       "\\item 'NTN4\\_NTRK2'\n",
       "\\item 'BMP6\\_ACVR1\\_ACVR2A'\n",
       "\\item 'WNT2B\\_FZD1\\_LRP5'\n",
       "\\item 'WNT2B\\_FZD6\\_LRP5'\n",
       "\\item 'WNT2B\\_FZD7\\_LRP5'\n",
       "\\item 'WNT2B\\_FZD8\\_LRP5'\n",
       "\\item 'WNT8A\\_FZD6\\_LRP5'\n",
       "\\item 'WNT2B\\_FZD1\\_LRP6'\n",
       "\\item 'WNT2B\\_FZD6\\_LRP6'\n",
       "\\item 'WNT2B\\_FZD8\\_LRP6'\n",
       "\\item 'WNT8A\\_FZD6\\_LRP6'\n",
       "\\item 'WNT5B\\_FZD6'\n",
       "\\item 'NRG3\\_ERBB4'\n",
       "\\item 'FGF2\\_FGFR4'\n",
       "\\item 'PDGFA\\_PDGFRA'\n",
       "\\item 'PDGFA\\_PDGFRB'\n",
       "\\item 'VEGFA\\_VEGFR1R2'\n",
       "\\item 'CCL8\\_ACKR1'\n",
       "\\item 'CXCL12\\_ACKR3'\n",
       "\\item 'MIF\\_ACKR3'\n",
       "\\item 'CTF1\\_LIFR\\_IL6ST'\n",
       "\\item 'SPP1\\_CD44'\n",
       "\\item 'SPP1\\_ITGAV\\_ITGB1'\n",
       "\\item 'SPP1\\_ITGAV\\_ITGB5'\n",
       "\\item 'SPP1\\_ITGA9\\_ITGB1'\n",
       "\\item 'SPP1\\_ITGA5\\_ITGB1'\n",
       "\\item 'ANGPTL2\\_TLR4'\n",
       "\\item 'PTN\\_SDC3'\n",
       "\\item 'SEMA3C\\_NRP1\\_PLXNA1'\n",
       "\\item 'SEMA3C\\_NRP2\\_PLXNA1'\n",
       "\\item 'SEMA3F\\_NRP2\\_PLXNA1'\n",
       "\\item 'SEMA3F\\_NRP2\\_PLXNA4'\n",
       "\\item 'GAS6\\_MERTK'\n",
       "\\item 'SLIT2\\_ROBO1'\n",
       "\\item 'SLIT2\\_ROBO2'\n",
       "\\item 'PROS1\\_MERTK'\n",
       "\\item 'THBS4\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL6A6\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMC2\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMC2\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A6\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A6\\_CD44'\n",
       "\\item 'COL6A6\\_SDC4'\n",
       "\\item 'THBS4\\_SDC4'\n",
       "\\item 'LAMA2\\_SV2A'\n",
       "\\item 'LAMA4\\_SV2A'\n",
       "\\item 'LAMB2\\_SV2A'\n",
       "\\item 'LAMC1\\_SV2A'\n",
       "\\item 'LAMC2\\_SV2A'\n",
       "\\item 'THBS4\\_CD36'\n",
       "\\item 'THBS4\\_CD47'\n",
       "\\item 'AGRN\\_DAG1'\n",
       "\\item 'LAMC2\\_DAG1'\n",
       "\\item 'PGF2a-AKR1C3\\_PTGFR'\n",
       "\\item 'PGF2a-CBR1\\_PTGFR'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A1\\_RARA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A1\\_RARA\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A1\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A2\\_RARA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A2\\_RARA\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A2\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A3\\_RARA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A3\\_RARA\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A3\\_RXRA\\_CRABP2'\n",
       "\\item 'CD34\\_SELP'\n",
       "\\item 'CNTN1\\_NRCAM'\n",
       "\\item 'NCAM1\\_NCAM2'\n",
       "\\item 'NECTIN3\\_NECTIN1'\n",
       "\\item 'NECTIN3\\_PVR'\n",
       "\\item 'JAG2\\_NOTCH1'\n",
       "\\item 'JAG2\\_NOTCH2'\n",
       "\\item 'JAG2\\_NOTCH3'\n",
       "\\item 'NRXN1\\_NLGN1'\n",
       "\\item 'NRXN1\\_NLGN2'\n",
       "\\item 'NRXN2\\_NLGN2'\n",
       "\\item 'SELE\\_CD44'\n",
       "\\item 'SELE\\_GLG1'\n",
       "\\item 'SEMA4D\\_PLXNB1'\n",
       "\\item 'SEMA4D\\_PLXNB2'\n",
       "\\item 'SEMA4G\\_PLXNB2'\n",
       "\\item 'CDH12\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL3A1\\_ADGRG1'\n",
       "\\item 'TGM2\\_ADGRG1'\n",
       "\\item 'FLRT2\\_UNC5B'\n",
       "\\item 'NRXN1\\_CLSTN1'\n",
       "\\item 'NRXN1\\_CLSTN2'\n",
       "\\item 'NRXN1\\_DAG1'\n",
       "\\item 'NRXN2\\_DAG1'\n",
       "\\item 'LRFN5\\_PTPRD'\n",
       "\\item 'LRFN5\\_PTPRF'\n",
       "\\item 'LRFN5\\_PTPRS'\n",
       "\\item 'LRRC4C\\_PTPRF'\n",
       "\\item 'NRXN1\\_LRRTM4'\n",
       "\\item 'NRXN3\\_LRRTM4'\n",
       "\\item 'LRRTM4\\_PTPRS'\n",
       "\\item 'NTN4\\_UNC5B'\n",
       "\\item 'TGFB3\\_TGFBR1\\_TGFBR2'\n",
       "\\item 'TGFB3\\_ACVR1\\_TGFBR1'\n",
       "\\item 'SPP1\\_ITGA8\\_ITGB1'\n",
       "\\item 'ANGPTL4\\_ITGA5\\_ITGB1'\n",
       "\\item 'ANGPTL4\\_CDH11'\n",
       "\\item 'ANGPTL4\\_SDC2'\n",
       "\\item 'ANGPTL4\\_SDC4'\n",
       "\\item 'EDN1\\_EDNRA'\n",
       "\\item 'EDN1\\_EDNRB'\n",
       "\\item 'PGE2-PTGES3\\_PTGER2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A1\\_RORB'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A2\\_RORB'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A3\\_RORB'\n",
       "\\item 'Dehydroepiandrosterone-DHEA-STS\\_ESR1'\n",
       "\\item 'CDH1\\_CDH1'\n",
       "\\item 'FLRT3\\_ADGRL2'\n",
       "\\item 'FLRT3\\_ADGRL3'\n",
       "\\item 'FLRT3\\_FLRT3'\n",
       "\\item 'FLRT3\\_UNC5C'\n",
       "\\item 'NRXN2\\_CLSTN1'\n",
       "\\item 'NRXN2\\_CLSTN2'\n",
       "\\item 'NRXN1\\_LRRTM3'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TGFB1_TGFBR1_TGFBR2'\n",
       "2. 'TGFB2_TGFBR1_TGFBR2'\n",
       "3. 'TGFB1_ACVR1_TGFBR1'\n",
       "4. 'TGFB2_ACVR1_TGFBR1'\n",
       "5. 'BMP6_ACVR1_BMPR2'\n",
       "6. 'BMP6_BMPR1A_BMPR2'\n",
       "7. 'BMP6_BMPR1B_BMPR2'\n",
       "8. 'WNT4_FZD1_LRP5'\n",
       "9. 'WNT4_FZD3_LRP5'\n",
       "10. 'WNT4_FZD1_LRP6'\n",
       "11. 'WNT4_FZD3_LRP6'\n",
       "12. 'WNT4_FZD7_LRP6'\n",
       "13. 'WNT4_FZD8_LRP6'\n",
       "14. 'WNT5B_FZD1'\n",
       "15. 'WNT5B_FZD3'\n",
       "16. 'WNT5B_FZD7'\n",
       "17. 'WNT5B_FZD8'\n",
       "18. 'AREG_EGFR'\n",
       "19. 'AREG_EGFR_ERBB2'\n",
       "20. 'NRG1_ERBB4'\n",
       "21. 'FGF2_FGFR1'\n",
       "22. 'FGF2_FGFR2'\n",
       "23. 'FGF7_FGFR1'\n",
       "24. 'FGF7_FGFR2'\n",
       "25. 'PDGFB_PDGFRA'\n",
       "26. 'PDGFB_PDGFRB'\n",
       "27. 'PDGFC_PDGFRA'\n",
       "28. 'PDGFD_PDGFRB'\n",
       "29. 'VEGFB_VEGFR1'\n",
       "30. 'VEGFC_VEGFR2'\n",
       "31. 'PGF_VEGFR1'\n",
       "32. 'IGF1_IGF1R'\n",
       "33. 'IGF2_IGF1R'\n",
       "34. 'IGF2_IGF2R'\n",
       "35. 'CCL2_ACKR1'\n",
       "36. 'CCL5_ACKR1'\n",
       "37. 'CCL14_ACKR1'\n",
       "38. 'CXCL8_ACKR1'\n",
       "39. 'CXCL12_CXCR4'\n",
       "40. 'MIF_CD74_CXCR4'\n",
       "41. 'MIF_CD74_CD44'\n",
       "42. 'TNFSF12_TNFRSF12A'\n",
       "43. 'NAMPT_ITGA5_ITGB1'\n",
       "44. 'ANGPTL1_ITGA1_ITGB1'\n",
       "45. 'ANGPTL2_ITGA5_ITGB1'\n",
       "46. 'ANGPT2_TEK'\n",
       "47. 'ANGPT2_ITGA5_ITGB1'\n",
       "48. 'MDK_SDC2'\n",
       "49. 'MDK_SDC4'\n",
       "50. 'MDK_ITGA6_ITGB1'\n",
       "51. 'MDK_LRP1'\n",
       "52. 'MDK_NCL'\n",
       "53. 'PTN_SDC2'\n",
       "54. 'PTN_SDC4'\n",
       "55. 'PTN_NCL'\n",
       "56. 'NTF3_NTRK2'\n",
       "57. 'NTF3_NTRK3'\n",
       "58. 'SEMA3A_NRP1_PLXNA2'\n",
       "59. 'SEMA3A_NRP1_PLXNA4'\n",
       "60. 'SEMA3C_NRP1_PLXNA2'\n",
       "61. 'SEMA3C_NRP1_PLXNA4'\n",
       "62. 'SEMA3C_NRP2_PLXNA2'\n",
       "63. 'SEMA3C_NRP2_PLXNA4'\n",
       "64. 'SEMA3G_NRP2_PLXNA2'\n",
       "65. 'SEMA3G_NRP2_PLXNA4'\n",
       "66. 'SEMA3C_NRP1_NRP2'\n",
       "67. 'SEMA3C_PLXND1'\n",
       "68. 'ADM_CALCRL'\n",
       "69. 'GAS6_AXL'\n",
       "70. 'GAS6_TYRO3'\n",
       "71. 'GRN_SORT1'\n",
       "72. 'PROS1_AXL'\n",
       "73. 'PROS1_TYRO3'\n",
       "74. 'PTPRS_NTRK3'\n",
       "75. 'IGFBP3_TMEM219'\n",
       "76. 'PPIA_BSG'\n",
       "77. 'COL1A1_ITGA1_ITGB1'\n",
       "78. 'COL1A2_ITGA1_ITGB1'\n",
       "79. 'COL4A1_ITGA1_ITGB1'\n",
       "80. 'COL4A2_ITGA1_ITGB1'\n",
       "81. 'COL4A5_ITGA1_ITGB1'\n",
       "82. 'COL6A1_ITGA1_ITGB1'\n",
       "83. 'COL6A2_ITGA1_ITGB1'\n",
       "84. 'COL6A3_ITGA1_ITGB1'\n",
       "85. 'FN1_ITGA3_ITGB1'\n",
       "86. 'FN1_ITGA5_ITGB1'\n",
       "87. 'FN1_ITGA8_ITGB1'\n",
       "88. 'LAMA1_ITGA1_ITGB1'\n",
       "89. 'LAMA2_ITGA1_ITGB1'\n",
       "90. 'LAMA3_ITGA1_ITGB1'\n",
       "91. 'LAMA4_ITGA1_ITGB1'\n",
       "92. 'LAMA5_ITGA1_ITGB1'\n",
       "93. 'LAMB1_ITGA1_ITGB1'\n",
       "94. 'LAMB2_ITGA1_ITGB1'\n",
       "95. 'LAMC1_ITGA1_ITGB1'\n",
       "96. 'LAMA1_ITGA3_ITGB1'\n",
       "97. 'LAMA2_ITGA3_ITGB1'\n",
       "98. 'LAMA3_ITGA3_ITGB1'\n",
       "99. 'LAMA4_ITGA3_ITGB1'\n",
       "100. 'LAMA5_ITGA3_ITGB1'\n",
       "101. 'LAMB1_ITGA3_ITGB1'\n",
       "102. 'LAMB2_ITGA3_ITGB1'\n",
       "103. 'LAMC1_ITGA3_ITGB1'\n",
       "104. 'RELN_ITGA3_ITGB1'\n",
       "105. 'THBS1_ITGA3_ITGB1'\n",
       "106. 'THBS2_ITGA3_ITGB1'\n",
       "107. 'COL1A1_ITGA3_ITGB1'\n",
       "108. 'COL1A2_ITGA3_ITGB1'\n",
       "109. 'COL4A1_ITGA3_ITGB1'\n",
       "110. 'COL4A2_ITGA3_ITGB1'\n",
       "111. 'COL4A5_ITGA3_ITGB1'\n",
       "112. 'COL6A1_ITGA3_ITGB1'\n",
       "113. 'COL6A2_ITGA3_ITGB1'\n",
       "114. 'COL6A3_ITGA3_ITGB1'\n",
       "115. 'LAMA1_ITGA6_ITGB1'\n",
       "116. 'LAMA2_ITGA6_ITGB1'\n",
       "117. 'LAMA3_ITGA6_ITGB1'\n",
       "118. 'LAMA4_ITGA6_ITGB1'\n",
       "119. 'LAMA5_ITGA6_ITGB1'\n",
       "120. 'LAMB1_ITGA6_ITGB1'\n",
       "121. 'LAMB2_ITGA6_ITGB1'\n",
       "122. 'LAMC1_ITGA6_ITGB1'\n",
       "123. 'LAMA1_ITGA7_ITGB1'\n",
       "124. 'LAMA2_ITGA7_ITGB1'\n",
       "125. 'LAMA3_ITGA7_ITGB1'\n",
       "126. 'LAMA4_ITGA7_ITGB1'\n",
       "127. 'LAMA5_ITGA7_ITGB1'\n",
       "128. 'LAMB1_ITGA7_ITGB1'\n",
       "129. 'LAMB2_ITGA7_ITGB1'\n",
       "130. 'LAMC1_ITGA7_ITGB1'\n",
       "131. 'TNXB_ITGA8_ITGB1'\n",
       "132. 'TNXB_ITGA9_ITGB1'\n",
       "133. 'COL1A1_ITGA9_ITGB1'\n",
       "134. 'COL1A2_ITGA9_ITGB1'\n",
       "135. 'COL4A1_ITGA9_ITGB1'\n",
       "136. 'COL4A2_ITGA9_ITGB1'\n",
       "137. 'COL4A5_ITGA9_ITGB1'\n",
       "138. 'COL6A1_ITGA9_ITGB1'\n",
       "139. 'COL6A2_ITGA9_ITGB1'\n",
       "140. 'COL6A3_ITGA9_ITGB1'\n",
       "141. 'LAMA1_ITGA9_ITGB1'\n",
       "142. 'LAMA2_ITGA9_ITGB1'\n",
       "143. 'LAMA3_ITGA9_ITGB1'\n",
       "144. 'LAMA4_ITGA9_ITGB1'\n",
       "145. 'LAMA5_ITGA9_ITGB1'\n",
       "146. 'LAMB1_ITGA9_ITGB1'\n",
       "147. 'LAMB2_ITGA9_ITGB1'\n",
       "148. 'LAMC1_ITGA9_ITGB1'\n",
       "149. 'FN1_ITGAV_ITGB1'\n",
       "150. 'FN1_ITGAV_ITGB8'\n",
       "151. 'LAMA1_ITGAV_ITGB8'\n",
       "152. 'LAMA2_ITGAV_ITGB8'\n",
       "153. 'LAMA3_ITGAV_ITGB8'\n",
       "154. 'LAMA4_ITGAV_ITGB8'\n",
       "155. 'LAMA5_ITGAV_ITGB8'\n",
       "156. 'LAMB1_ITGAV_ITGB8'\n",
       "157. 'LAMB2_ITGAV_ITGB8'\n",
       "158. 'LAMC1_ITGAV_ITGB8'\n",
       "159. 'COL1A1_ITGAV_ITGB8'\n",
       "160. 'COL1A2_ITGAV_ITGB8'\n",
       "161. 'COL4A1_ITGAV_ITGB8'\n",
       "162. 'COL4A2_ITGAV_ITGB8'\n",
       "163. 'COL4A5_ITGAV_ITGB8'\n",
       "164. 'COL6A1_ITGAV_ITGB8'\n",
       "165. 'COL6A2_ITGAV_ITGB8'\n",
       "166. 'COL6A3_ITGAV_ITGB8'\n",
       "167. 'FN1_CD44'\n",
       "168. 'COL1A1_CD44'\n",
       "169. 'COL1A2_CD44'\n",
       "170. 'COL4A1_CD44'\n",
       "171. 'COL4A2_CD44'\n",
       "172. 'COL4A5_CD44'\n",
       "173. 'COL6A1_CD44'\n",
       "174. 'COL6A2_CD44'\n",
       "175. 'COL6A3_CD44'\n",
       "176. 'LAMA1_CD44'\n",
       "177. 'LAMA2_CD44'\n",
       "178. 'LAMA3_CD44'\n",
       "179. 'LAMA4_CD44'\n",
       "180. 'LAMA5_CD44'\n",
       "181. 'LAMB1_CD44'\n",
       "182. 'LAMB2_CD44'\n",
       "183. 'LAMC1_CD44'\n",
       "184. 'COL1A1_SDC4'\n",
       "185. 'COL1A2_SDC4'\n",
       "186. 'COL4A1_SDC4'\n",
       "187. 'COL4A2_SDC4'\n",
       "188. 'COL4A5_SDC4'\n",
       "189. 'COL6A1_SDC4'\n",
       "190. 'COL6A2_SDC4'\n",
       "191. 'COL6A3_SDC4'\n",
       "192. 'FN1_SDC4'\n",
       "193. 'TNXB_SDC4'\n",
       "194. 'THBS1_SDC4'\n",
       "195. 'THBS2_SDC4'\n",
       "196. 'THBS1_CD36'\n",
       "197. 'THBS2_CD36'\n",
       "198. 'THBS1_CD47'\n",
       "199. 'THBS2_CD47'\n",
       "200. 'HSPG2_DAG1'\n",
       "201. ⋯\n",
       "202. 'EFNB1_EPHB4'\n",
       "203. 'EFNB2_EPHA4'\n",
       "204. 'EFNB2_EPHB1'\n",
       "205. 'EFNB2_EPHB2'\n",
       "206. 'EFNB2_EPHB4'\n",
       "207. 'ESAM_ESAM'\n",
       "208. 'JAM2_ITGA3_ITGB1'\n",
       "209. 'JAM2_ITGAV_ITGB1'\n",
       "210. 'JAM2_JAM3'\n",
       "211. 'F11R_F11R'\n",
       "212. 'JAM2_F11R'\n",
       "213. 'JAM2_JAM2'\n",
       "214. 'JAM3_F11R'\n",
       "215. 'JAM3_JAM3'\n",
       "216. 'MAG_MAG'\n",
       "217. 'MPZL1_MPZL1'\n",
       "218. 'NCAM1_FGFR1'\n",
       "219. 'NCAM1_NCAM1'\n",
       "220. 'NECTIN3_NECTIN2'\n",
       "221. 'NEGR1_NEGR1'\n",
       "222. 'DLK1_NOTCH1'\n",
       "223. 'DLK1_NOTCH2'\n",
       "224. 'DLK1_NOTCH3'\n",
       "225. 'DLK1_NOTCH4'\n",
       "226. 'DLL4_NOTCH1'\n",
       "227. 'DLL4_NOTCH2'\n",
       "228. 'DLL4_NOTCH3'\n",
       "229. 'DLL4_NOTCH4'\n",
       "230. 'JAG1_NOTCH1'\n",
       "231. 'JAG1_NOTCH2'\n",
       "232. 'JAG1_NOTCH3'\n",
       "233. 'JAG1_NOTCH4'\n",
       "234. 'NRXN3_NLGN1'\n",
       "235. 'NRXN3_NLGN2'\n",
       "236. 'PECAM1_PECAM1'\n",
       "237. 'PTPRM_PTPRM'\n",
       "238. 'SEMA4C_PLXNB2'\n",
       "239. 'SEMA6A_PLXNA2'\n",
       "240. 'SEMA6A_PLXNA4'\n",
       "241. 'SEMA7A_ITGB1_ITGA1'\n",
       "242. 'SEMA7A_PLXNC1'\n",
       "243. 'ITGA9_ITGB1_VCAM1'\n",
       "244. 'VSIR_IGSF11'\n",
       "245. 'CDH3_CDH3'\n",
       "246. 'GJA5_GJA5'\n",
       "247. 'GJA1_GJA5'\n",
       "248. 'GJC1_GJA5'\n",
       "249. 'GJA1_GJA1'\n",
       "250. 'GJC1_GJA1'\n",
       "251. 'GJC1_GJC1'\n",
       "252. 'PCDHGA1_PCDHGA1'\n",
       "253. 'SDC2_ADGRA2'\n",
       "254. 'CD55_ADGRE5'\n",
       "255. 'LPAR1_ADGRE5'\n",
       "256. 'THY1_ADGRE5'\n",
       "257. 'ITGAV_ITGB1_ADGRE5'\n",
       "258. 'COL4A1_ADGRG6'\n",
       "259. 'COL4A2_ADGRG6'\n",
       "260. 'COL4A5_ADGRG6'\n",
       "261. 'LAMA2_ADGRG6'\n",
       "262. 'PRNP_ADGRG6'\n",
       "263. 'FLRT2_ADGRL2'\n",
       "264. 'TENM2_ADGRL2'\n",
       "265. 'TENM3_ADGRL2'\n",
       "266. 'TENM4_ADGRL2'\n",
       "267. 'FLRT2_ADGRL3'\n",
       "268. 'TENM2_ADGRL3'\n",
       "269. 'TENM3_ADGRL3'\n",
       "270. 'TENM4_ADGRL3'\n",
       "271. 'FLRT2_FLRT2'\n",
       "272. 'FLRT2_UNC5C'\n",
       "273. 'Q9HDB5_CLSTN1'\n",
       "274. 'Q9HDB5_CLSTN2'\n",
       "275. 'NRXN3_LRRTM3'\n",
       "276. 'APP_SORL1'\n",
       "277. 'CADM1_NECTIN3'\n",
       "278. 'CD99_PILRB'\n",
       "279. 'CNTN1_NOTCH1'\n",
       "280. 'CNTN1_NOTCH2'\n",
       "281. 'NTN1_DCC'\n",
       "282. 'NTN1_DSCAM'\n",
       "283. 'NTN1_UNC5C'\n",
       "284. 'NTN4_NTRK2'\n",
       "285. 'BMP6_ACVR1_ACVR2A'\n",
       "286. 'WNT2B_FZD1_LRP5'\n",
       "287. 'WNT2B_FZD6_LRP5'\n",
       "288. 'WNT2B_FZD7_LRP5'\n",
       "289. 'WNT2B_FZD8_LRP5'\n",
       "290. 'WNT8A_FZD6_LRP5'\n",
       "291. 'WNT2B_FZD1_LRP6'\n",
       "292. 'WNT2B_FZD6_LRP6'\n",
       "293. 'WNT2B_FZD8_LRP6'\n",
       "294. 'WNT8A_FZD6_LRP6'\n",
       "295. 'WNT5B_FZD6'\n",
       "296. 'NRG3_ERBB4'\n",
       "297. 'FGF2_FGFR4'\n",
       "298. 'PDGFA_PDGFRA'\n",
       "299. 'PDGFA_PDGFRB'\n",
       "300. 'VEGFA_VEGFR1R2'\n",
       "301. 'CCL8_ACKR1'\n",
       "302. 'CXCL12_ACKR3'\n",
       "303. 'MIF_ACKR3'\n",
       "304. 'CTF1_LIFR_IL6ST'\n",
       "305. 'SPP1_CD44'\n",
       "306. 'SPP1_ITGAV_ITGB1'\n",
       "307. 'SPP1_ITGAV_ITGB5'\n",
       "308. 'SPP1_ITGA9_ITGB1'\n",
       "309. 'SPP1_ITGA5_ITGB1'\n",
       "310. 'ANGPTL2_TLR4'\n",
       "311. 'PTN_SDC3'\n",
       "312. 'SEMA3C_NRP1_PLXNA1'\n",
       "313. 'SEMA3C_NRP2_PLXNA1'\n",
       "314. 'SEMA3F_NRP2_PLXNA1'\n",
       "315. 'SEMA3F_NRP2_PLXNA4'\n",
       "316. 'GAS6_MERTK'\n",
       "317. 'SLIT2_ROBO1'\n",
       "318. 'SLIT2_ROBO2'\n",
       "319. 'PROS1_MERTK'\n",
       "320. 'THBS4_ITGA3_ITGB1'\n",
       "321. 'COL6A6_ITGA9_ITGB1'\n",
       "322. 'LAMC2_ITGA9_ITGB1'\n",
       "323. 'LAMC2_ITGAV_ITGB8'\n",
       "324. 'COL6A6_ITGAV_ITGB8'\n",
       "325. 'COL6A6_CD44'\n",
       "326. 'COL6A6_SDC4'\n",
       "327. 'THBS4_SDC4'\n",
       "328. 'LAMA2_SV2A'\n",
       "329. 'LAMA4_SV2A'\n",
       "330. 'LAMB2_SV2A'\n",
       "331. 'LAMC1_SV2A'\n",
       "332. 'LAMC2_SV2A'\n",
       "333. 'THBS4_CD36'\n",
       "334. 'THBS4_CD47'\n",
       "335. 'AGRN_DAG1'\n",
       "336. 'LAMC2_DAG1'\n",
       "337. 'PGF2a-AKR1C3_PTGFR'\n",
       "338. 'PGF2a-CBR1_PTGFR'\n",
       "339. 'RetinoicAcid-RA-ALDH1A1_RARA_CRABP2'\n",
       "340. 'RetinoicAcid-RA-ALDH1A1_RARA_RXRA_CRABP2'\n",
       "341. 'RetinoicAcid-RA-ALDH1A1_RXRA_CRABP2'\n",
       "342. 'RetinoicAcid-RA-ALDH1A2_RARA_CRABP2'\n",
       "343. 'RetinoicAcid-RA-ALDH1A2_RARA_RXRA_CRABP2'\n",
       "344. 'RetinoicAcid-RA-ALDH1A2_RXRA_CRABP2'\n",
       "345. 'RetinoicAcid-RA-ALDH1A3_RARA_CRABP2'\n",
       "346. 'RetinoicAcid-RA-ALDH1A3_RARA_RXRA_CRABP2'\n",
       "347. 'RetinoicAcid-RA-ALDH1A3_RXRA_CRABP2'\n",
       "348. 'CD34_SELP'\n",
       "349. 'CNTN1_NRCAM'\n",
       "350. 'NCAM1_NCAM2'\n",
       "351. 'NECTIN3_NECTIN1'\n",
       "352. 'NECTIN3_PVR'\n",
       "353. 'JAG2_NOTCH1'\n",
       "354. 'JAG2_NOTCH2'\n",
       "355. 'JAG2_NOTCH3'\n",
       "356. 'NRXN1_NLGN1'\n",
       "357. 'NRXN1_NLGN2'\n",
       "358. 'NRXN2_NLGN2'\n",
       "359. 'SELE_CD44'\n",
       "360. 'SELE_GLG1'\n",
       "361. 'SEMA4D_PLXNB1'\n",
       "362. 'SEMA4D_PLXNB2'\n",
       "363. 'SEMA4G_PLXNB2'\n",
       "364. 'CDH12_ITGA1_ITGB1'\n",
       "365. 'COL3A1_ADGRG1'\n",
       "366. 'TGM2_ADGRG1'\n",
       "367. 'FLRT2_UNC5B'\n",
       "368. 'NRXN1_CLSTN1'\n",
       "369. 'NRXN1_CLSTN2'\n",
       "370. 'NRXN1_DAG1'\n",
       "371. 'NRXN2_DAG1'\n",
       "372. 'LRFN5_PTPRD'\n",
       "373. 'LRFN5_PTPRF'\n",
       "374. 'LRFN5_PTPRS'\n",
       "375. 'LRRC4C_PTPRF'\n",
       "376. 'NRXN1_LRRTM4'\n",
       "377. 'NRXN3_LRRTM4'\n",
       "378. 'LRRTM4_PTPRS'\n",
       "379. 'NTN4_UNC5B'\n",
       "380. 'TGFB3_TGFBR1_TGFBR2'\n",
       "381. 'TGFB3_ACVR1_TGFBR1'\n",
       "382. 'SPP1_ITGA8_ITGB1'\n",
       "383. 'ANGPTL4_ITGA5_ITGB1'\n",
       "384. 'ANGPTL4_CDH11'\n",
       "385. 'ANGPTL4_SDC2'\n",
       "386. 'ANGPTL4_SDC4'\n",
       "387. 'EDN1_EDNRA'\n",
       "388. 'EDN1_EDNRB'\n",
       "389. 'PGE2-PTGES3_PTGER2'\n",
       "390. 'RetinoicAcid-RA-ALDH1A1_RORB'\n",
       "391. 'RetinoicAcid-RA-ALDH1A2_RORB'\n",
       "392. 'RetinoicAcid-RA-ALDH1A3_RORB'\n",
       "393. 'Dehydroepiandrosterone-DHEA-STS_ESR1'\n",
       "394. 'CDH1_CDH1'\n",
       "395. 'FLRT3_ADGRL2'\n",
       "396. 'FLRT3_ADGRL3'\n",
       "397. 'FLRT3_FLRT3'\n",
       "398. 'FLRT3_UNC5C'\n",
       "399. 'NRXN2_CLSTN1'\n",
       "400. 'NRXN2_CLSTN2'\n",
       "401. 'NRXN1_LRRTM3'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"TGFB1_TGFBR1_TGFBR2\"                     \n",
       "  [2] \"TGFB2_TGFBR1_TGFBR2\"                     \n",
       "  [3] \"TGFB1_ACVR1_TGFBR1\"                      \n",
       "  [4] \"TGFB2_ACVR1_TGFBR1\"                      \n",
       "  [5] \"BMP6_ACVR1_BMPR2\"                        \n",
       "  [6] \"BMP6_BMPR1A_BMPR2\"                       \n",
       "  [7] \"BMP6_BMPR1B_BMPR2\"                       \n",
       "  [8] \"WNT4_FZD1_LRP5\"                          \n",
       "  [9] \"WNT4_FZD3_LRP5\"                          \n",
       " [10] \"WNT4_FZD1_LRP6\"                          \n",
       " [11] \"WNT4_FZD3_LRP6\"                          \n",
       " [12] \"WNT4_FZD7_LRP6\"                          \n",
       " [13] \"WNT4_FZD8_LRP6\"                          \n",
       " [14] \"WNT5B_FZD1\"                              \n",
       " [15] \"WNT5B_FZD3\"                              \n",
       " [16] \"WNT5B_FZD7\"                              \n",
       " [17] \"WNT5B_FZD8\"                              \n",
       " [18] \"AREG_EGFR\"                               \n",
       " [19] \"AREG_EGFR_ERBB2\"                         \n",
       " [20] \"NRG1_ERBB4\"                              \n",
       " [21] \"FGF2_FGFR1\"                              \n",
       " [22] \"FGF2_FGFR2\"                              \n",
       " [23] \"FGF7_FGFR1\"                              \n",
       " [24] \"FGF7_FGFR2\"                              \n",
       " [25] \"PDGFB_PDGFRA\"                            \n",
       " [26] \"PDGFB_PDGFRB\"                            \n",
       " [27] \"PDGFC_PDGFRA\"                            \n",
       " [28] \"PDGFD_PDGFRB\"                            \n",
       " [29] \"VEGFB_VEGFR1\"                            \n",
       " [30] \"VEGFC_VEGFR2\"                            \n",
       " [31] \"PGF_VEGFR1\"                              \n",
       " [32] \"IGF1_IGF1R\"                              \n",
       " [33] \"IGF2_IGF1R\"                              \n",
       " [34] \"IGF2_IGF2R\"                              \n",
       " [35] \"CCL2_ACKR1\"                              \n",
       " [36] \"CCL5_ACKR1\"                              \n",
       " [37] \"CCL14_ACKR1\"                             \n",
       " [38] \"CXCL8_ACKR1\"                             \n",
       " [39] \"CXCL12_CXCR4\"                            \n",
       " [40] \"MIF_CD74_CXCR4\"                          \n",
       " [41] \"MIF_CD74_CD44\"                           \n",
       " [42] \"TNFSF12_TNFRSF12A\"                       \n",
       " [43] \"NAMPT_ITGA5_ITGB1\"                       \n",
       " [44] \"ANGPTL1_ITGA1_ITGB1\"                     \n",
       " [45] \"ANGPTL2_ITGA5_ITGB1\"                     \n",
       " [46] \"ANGPT2_TEK\"                              \n",
       " [47] \"ANGPT2_ITGA5_ITGB1\"                      \n",
       " [48] \"MDK_SDC2\"                                \n",
       " [49] \"MDK_SDC4\"                                \n",
       " [50] \"MDK_ITGA6_ITGB1\"                         \n",
       " [51] \"MDK_LRP1\"                                \n",
       " [52] \"MDK_NCL\"                                 \n",
       " [53] \"PTN_SDC2\"                                \n",
       " [54] \"PTN_SDC4\"                                \n",
       " [55] \"PTN_NCL\"                                 \n",
       " [56] \"NTF3_NTRK2\"                              \n",
       " [57] \"NTF3_NTRK3\"                              \n",
       " [58] \"SEMA3A_NRP1_PLXNA2\"                      \n",
       " [59] \"SEMA3A_NRP1_PLXNA4\"                      \n",
       " [60] \"SEMA3C_NRP1_PLXNA2\"                      \n",
       " [61] \"SEMA3C_NRP1_PLXNA4\"                      \n",
       " [62] \"SEMA3C_NRP2_PLXNA2\"                      \n",
       " [63] \"SEMA3C_NRP2_PLXNA4\"                      \n",
       " [64] \"SEMA3G_NRP2_PLXNA2\"                      \n",
       " [65] \"SEMA3G_NRP2_PLXNA4\"                      \n",
       " [66] \"SEMA3C_NRP1_NRP2\"                        \n",
       " [67] \"SEMA3C_PLXND1\"                           \n",
       " [68] \"ADM_CALCRL\"                              \n",
       " [69] \"GAS6_AXL\"                                \n",
       " [70] \"GAS6_TYRO3\"                              \n",
       " [71] \"GRN_SORT1\"                               \n",
       " [72] \"PROS1_AXL\"                               \n",
       " [73] \"PROS1_TYRO3\"                             \n",
       " [74] \"PTPRS_NTRK3\"                             \n",
       " [75] \"IGFBP3_TMEM219\"                          \n",
       " [76] \"PPIA_BSG\"                                \n",
       " [77] \"COL1A1_ITGA1_ITGB1\"                      \n",
       " [78] \"COL1A2_ITGA1_ITGB1\"                      \n",
       " [79] \"COL4A1_ITGA1_ITGB1\"                      \n",
       " [80] \"COL4A2_ITGA1_ITGB1\"                      \n",
       " [81] \"COL4A5_ITGA1_ITGB1\"                      \n",
       " [82] \"COL6A1_ITGA1_ITGB1\"                      \n",
       " [83] \"COL6A2_ITGA1_ITGB1\"                      \n",
       " [84] \"COL6A3_ITGA1_ITGB1\"                      \n",
       " [85] \"FN1_ITGA3_ITGB1\"                         \n",
       " [86] \"FN1_ITGA5_ITGB1\"                         \n",
       " [87] \"FN1_ITGA8_ITGB1\"                         \n",
       " [88] \"LAMA1_ITGA1_ITGB1\"                       \n",
       " [89] \"LAMA2_ITGA1_ITGB1\"                       \n",
       " [90] \"LAMA3_ITGA1_ITGB1\"                       \n",
       " [91] \"LAMA4_ITGA1_ITGB1\"                       \n",
       " [92] \"LAMA5_ITGA1_ITGB1\"                       \n",
       " [93] \"LAMB1_ITGA1_ITGB1\"                       \n",
       " [94] \"LAMB2_ITGA1_ITGB1\"                       \n",
       " [95] \"LAMC1_ITGA1_ITGB1\"                       \n",
       " [96] \"LAMA1_ITGA3_ITGB1\"                       \n",
       " [97] \"LAMA2_ITGA3_ITGB1\"                       \n",
       " [98] \"LAMA3_ITGA3_ITGB1\"                       \n",
       " [99] \"LAMA4_ITGA3_ITGB1\"                       \n",
       "[100] \"LAMA5_ITGA3_ITGB1\"                       \n",
       "[101] \"LAMB1_ITGA3_ITGB1\"                       \n",
       "[102] \"LAMB2_ITGA3_ITGB1\"                       \n",
       "[103] \"LAMC1_ITGA3_ITGB1\"                       \n",
       "[104] \"RELN_ITGA3_ITGB1\"                        \n",
       "[105] \"THBS1_ITGA3_ITGB1\"                       \n",
       "[106] \"THBS2_ITGA3_ITGB1\"                       \n",
       "[107] \"COL1A1_ITGA3_ITGB1\"                      \n",
       "[108] \"COL1A2_ITGA3_ITGB1\"                      \n",
       "[109] \"COL4A1_ITGA3_ITGB1\"                      \n",
       "[110] \"COL4A2_ITGA3_ITGB1\"                      \n",
       "[111] \"COL4A5_ITGA3_ITGB1\"                      \n",
       "[112] \"COL6A1_ITGA3_ITGB1\"                      \n",
       "[113] \"COL6A2_ITGA3_ITGB1\"                      \n",
       "[114] \"COL6A3_ITGA3_ITGB1\"                      \n",
       "[115] \"LAMA1_ITGA6_ITGB1\"                       \n",
       "[116] \"LAMA2_ITGA6_ITGB1\"                       \n",
       "[117] \"LAMA3_ITGA6_ITGB1\"                       \n",
       "[118] \"LAMA4_ITGA6_ITGB1\"                       \n",
       "[119] \"LAMA5_ITGA6_ITGB1\"                       \n",
       "[120] \"LAMB1_ITGA6_ITGB1\"                       \n",
       "[121] \"LAMB2_ITGA6_ITGB1\"                       \n",
       "[122] \"LAMC1_ITGA6_ITGB1\"                       \n",
       "[123] \"LAMA1_ITGA7_ITGB1\"                       \n",
       "[124] \"LAMA2_ITGA7_ITGB1\"                       \n",
       "[125] \"LAMA3_ITGA7_ITGB1\"                       \n",
       "[126] \"LAMA4_ITGA7_ITGB1\"                       \n",
       "[127] \"LAMA5_ITGA7_ITGB1\"                       \n",
       "[128] \"LAMB1_ITGA7_ITGB1\"                       \n",
       "[129] \"LAMB2_ITGA7_ITGB1\"                       \n",
       "[130] \"LAMC1_ITGA7_ITGB1\"                       \n",
       "[131] \"TNXB_ITGA8_ITGB1\"                        \n",
       "[132] \"TNXB_ITGA9_ITGB1\"                        \n",
       "[133] \"COL1A1_ITGA9_ITGB1\"                      \n",
       "[134] \"COL1A2_ITGA9_ITGB1\"                      \n",
       "[135] \"COL4A1_ITGA9_ITGB1\"                      \n",
       "[136] \"COL4A2_ITGA9_ITGB1\"                      \n",
       "[137] \"COL4A5_ITGA9_ITGB1\"                      \n",
       "[138] \"COL6A1_ITGA9_ITGB1\"                      \n",
       "[139] \"COL6A2_ITGA9_ITGB1\"                      \n",
       "[140] \"COL6A3_ITGA9_ITGB1\"                      \n",
       "[141] \"LAMA1_ITGA9_ITGB1\"                       \n",
       "[142] \"LAMA2_ITGA9_ITGB1\"                       \n",
       "[143] \"LAMA3_ITGA9_ITGB1\"                       \n",
       "[144] \"LAMA4_ITGA9_ITGB1\"                       \n",
       "[145] \"LAMA5_ITGA9_ITGB1\"                       \n",
       "[146] \"LAMB1_ITGA9_ITGB1\"                       \n",
       "[147] \"LAMB2_ITGA9_ITGB1\"                       \n",
       "[148] \"LAMC1_ITGA9_ITGB1\"                       \n",
       "[149] \"FN1_ITGAV_ITGB1\"                         \n",
       "[150] \"FN1_ITGAV_ITGB8\"                         \n",
       "[151] \"LAMA1_ITGAV_ITGB8\"                       \n",
       "[152] \"LAMA2_ITGAV_ITGB8\"                       \n",
       "[153] \"LAMA3_ITGAV_ITGB8\"                       \n",
       "[154] \"LAMA4_ITGAV_ITGB8\"                       \n",
       "[155] \"LAMA5_ITGAV_ITGB8\"                       \n",
       "[156] \"LAMB1_ITGAV_ITGB8\"                       \n",
       "[157] \"LAMB2_ITGAV_ITGB8\"                       \n",
       "[158] \"LAMC1_ITGAV_ITGB8\"                       \n",
       "[159] \"COL1A1_ITGAV_ITGB8\"                      \n",
       "[160] \"COL1A2_ITGAV_ITGB8\"                      \n",
       "[161] \"COL4A1_ITGAV_ITGB8\"                      \n",
       "[162] \"COL4A2_ITGAV_ITGB8\"                      \n",
       "[163] \"COL4A5_ITGAV_ITGB8\"                      \n",
       "[164] \"COL6A1_ITGAV_ITGB8\"                      \n",
       "[165] \"COL6A2_ITGAV_ITGB8\"                      \n",
       "[166] \"COL6A3_ITGAV_ITGB8\"                      \n",
       "[167] \"FN1_CD44\"                                \n",
       "[168] \"COL1A1_CD44\"                             \n",
       "[169] \"COL1A2_CD44\"                             \n",
       "[170] \"COL4A1_CD44\"                             \n",
       "[171] \"COL4A2_CD44\"                             \n",
       "[172] \"COL4A5_CD44\"                             \n",
       "[173] \"COL6A1_CD44\"                             \n",
       "[174] \"COL6A2_CD44\"                             \n",
       "[175] \"COL6A3_CD44\"                             \n",
       "[176] \"LAMA1_CD44\"                              \n",
       "[177] \"LAMA2_CD44\"                              \n",
       "[178] \"LAMA3_CD44\"                              \n",
       "[179] \"LAMA4_CD44\"                              \n",
       "[180] \"LAMA5_CD44\"                              \n",
       "[181] \"LAMB1_CD44\"                              \n",
       "[182] \"LAMB2_CD44\"                              \n",
       "[183] \"LAMC1_CD44\"                              \n",
       "[184] \"COL1A1_SDC4\"                             \n",
       "[185] \"COL1A2_SDC4\"                             \n",
       "[186] \"COL4A1_SDC4\"                             \n",
       "[187] \"COL4A2_SDC4\"                             \n",
       "[188] \"COL4A5_SDC4\"                             \n",
       "[189] \"COL6A1_SDC4\"                             \n",
       "[190] \"COL6A2_SDC4\"                             \n",
       "[191] \"COL6A3_SDC4\"                             \n",
       "[192] \"FN1_SDC4\"                                \n",
       "[193] \"TNXB_SDC4\"                               \n",
       "[194] \"THBS1_SDC4\"                              \n",
       "[195] \"THBS2_SDC4\"                              \n",
       "[196] \"THBS1_CD36\"                              \n",
       "[197] \"THBS2_CD36\"                              \n",
       "[198] \"THBS1_CD47\"                              \n",
       "[199] \"THBS2_CD47\"                              \n",
       "[200] \"HSPG2_DAG1\"                              \n",
       "[201] \"LAMA1_DAG1\"                              \n",
       "[202] \"LAMA2_DAG1\"                              \n",
       "[203] \"LAMA3_DAG1\"                              \n",
       "[204] \"LAMA4_DAG1\"                              \n",
       "[205] \"LAMA5_DAG1\"                              \n",
       "[206] \"LAMB1_DAG1\"                              \n",
       "[207] \"LAMB2_DAG1\"                              \n",
       "[208] \"LAMC1_DAG1\"                              \n",
       "[209] \"RELN_VLDLR\"                              \n",
       "[210] \"27Hydroxycholesterol-27HC-CYP11A1_NR1H4\" \n",
       "[211] \"ANDROSTENEDIONE-CYP17A1_AR\"              \n",
       "[212] \"ANDROSTERONE-HSD17B6_NR1H4\"              \n",
       "[213] \"Cholesterol-Cholesterol-DHCR24_RORA\"     \n",
       "[214] \"Cholesterol-Cholesterol-DHCR7_RORA\"      \n",
       "[215] \"Cholesterol-Cholesterol-LIPA_RORA\"       \n",
       "[216] \"Desmosterol-Desmosterol-DHCR7_NR1H2\"     \n",
       "[217] \"Testosterone-Testosterone-HSD17B12_AR\"   \n",
       "[218] \"APP_CD74\"                                \n",
       "[219] \"CADM1_CADM1\"                             \n",
       "[220] \"CD46_JAG1\"                               \n",
       "[221] \"CD99_CD99\"                               \n",
       "[222] \"CD99_CD99L2\"                             \n",
       "[223] \"CDH4_CDH4\"                               \n",
       "[224] \"CDH2_CDH2\"                               \n",
       "[225] \"CDH5_CDH5\"                               \n",
       "[226] \"CLDN11_CLDN11\"                           \n",
       "[227] \"CSPG4_ITGA2_ITGB1\"                       \n",
       "[228] \"EFNA1_EPHA3\"                             \n",
       "[229] \"EFNA1_EPHA4\"                             \n",
       "[230] \"EFNA5_EPHA3\"                             \n",
       "[231] \"EFNA5_EPHA4\"                             \n",
       "[232] \"EFNA5_EPHB2\"                             \n",
       "[233] \"EFNB1_EPHA4\"                             \n",
       "[234] \"EFNB1_EPHB1\"                             \n",
       "[235] \"EFNB1_EPHB2\"                             \n",
       "[236] \"EFNB1_EPHB4\"                             \n",
       "[237] \"EFNB2_EPHA4\"                             \n",
       "[238] \"EFNB2_EPHB1\"                             \n",
       "[239] \"EFNB2_EPHB2\"                             \n",
       "[240] \"EFNB2_EPHB4\"                             \n",
       "[241] \"ESAM_ESAM\"                               \n",
       "[242] \"JAM2_ITGA3_ITGB1\"                        \n",
       "[243] \"JAM2_ITGAV_ITGB1\"                        \n",
       "[244] \"JAM2_JAM3\"                               \n",
       "[245] \"F11R_F11R\"                               \n",
       "[246] \"JAM2_F11R\"                               \n",
       "[247] \"JAM2_JAM2\"                               \n",
       "[248] \"JAM3_F11R\"                               \n",
       "[249] \"JAM3_JAM3\"                               \n",
       "[250] \"MAG_MAG\"                                 \n",
       "[251] \"MPZL1_MPZL1\"                             \n",
       "[252] \"NCAM1_FGFR1\"                             \n",
       "[253] \"NCAM1_NCAM1\"                             \n",
       "[254] \"NECTIN3_NECTIN2\"                         \n",
       "[255] \"NEGR1_NEGR1\"                             \n",
       "[256] \"DLK1_NOTCH1\"                             \n",
       "[257] \"DLK1_NOTCH2\"                             \n",
       "[258] \"DLK1_NOTCH3\"                             \n",
       "[259] \"DLK1_NOTCH4\"                             \n",
       "[260] \"DLL4_NOTCH1\"                             \n",
       "[261] \"DLL4_NOTCH2\"                             \n",
       "[262] \"DLL4_NOTCH3\"                             \n",
       "[263] \"DLL4_NOTCH4\"                             \n",
       "[264] \"JAG1_NOTCH1\"                             \n",
       "[265] \"JAG1_NOTCH2\"                             \n",
       "[266] \"JAG1_NOTCH3\"                             \n",
       "[267] \"JAG1_NOTCH4\"                             \n",
       "[268] \"NRXN3_NLGN1\"                             \n",
       "[269] \"NRXN3_NLGN2\"                             \n",
       "[270] \"PECAM1_PECAM1\"                           \n",
       "[271] \"PTPRM_PTPRM\"                             \n",
       "[272] \"SEMA4C_PLXNB2\"                           \n",
       "[273] \"SEMA6A_PLXNA2\"                           \n",
       "[274] \"SEMA6A_PLXNA4\"                           \n",
       "[275] \"SEMA7A_ITGB1_ITGA1\"                      \n",
       "[276] \"SEMA7A_PLXNC1\"                           \n",
       "[277] \"ITGA9_ITGB1_VCAM1\"                       \n",
       "[278] \"VSIR_IGSF11\"                             \n",
       "[279] \"CDH3_CDH3\"                               \n",
       "[280] \"GJA5_GJA5\"                               \n",
       "[281] \"GJA1_GJA5\"                               \n",
       "[282] \"GJC1_GJA5\"                               \n",
       "[283] \"GJA1_GJA1\"                               \n",
       "[284] \"GJC1_GJA1\"                               \n",
       "[285] \"GJC1_GJC1\"                               \n",
       "[286] \"PCDHGA1_PCDHGA1\"                         \n",
       "[287] \"SDC2_ADGRA2\"                             \n",
       "[288] \"CD55_ADGRE5\"                             \n",
       "[289] \"LPAR1_ADGRE5\"                            \n",
       "[290] \"THY1_ADGRE5\"                             \n",
       "[291] \"ITGAV_ITGB1_ADGRE5\"                      \n",
       "[292] \"COL4A1_ADGRG6\"                           \n",
       "[293] \"COL4A2_ADGRG6\"                           \n",
       "[294] \"COL4A5_ADGRG6\"                           \n",
       "[295] \"LAMA2_ADGRG6\"                            \n",
       "[296] \"PRNP_ADGRG6\"                             \n",
       "[297] \"FLRT2_ADGRL2\"                            \n",
       "[298] \"TENM2_ADGRL2\"                            \n",
       "[299] \"TENM3_ADGRL2\"                            \n",
       "[300] \"TENM4_ADGRL2\"                            \n",
       "[301] \"FLRT2_ADGRL3\"                            \n",
       "[302] \"TENM2_ADGRL3\"                            \n",
       "[303] \"TENM3_ADGRL3\"                            \n",
       "[304] \"TENM4_ADGRL3\"                            \n",
       "[305] \"FLRT2_FLRT2\"                             \n",
       "[306] \"FLRT2_UNC5C\"                             \n",
       "[307] \"Q9HDB5_CLSTN1\"                           \n",
       "[308] \"Q9HDB5_CLSTN2\"                           \n",
       "[309] \"NRXN3_LRRTM3\"                            \n",
       "[310] \"APP_SORL1\"                               \n",
       "[311] \"CADM1_NECTIN3\"                           \n",
       "[312] \"CD99_PILRB\"                              \n",
       "[313] \"CNTN1_NOTCH1\"                            \n",
       "[314] \"CNTN1_NOTCH2\"                            \n",
       "[315] \"NTN1_DCC\"                                \n",
       "[316] \"NTN1_DSCAM\"                              \n",
       "[317] \"NTN1_UNC5C\"                              \n",
       "[318] \"NTN4_NTRK2\"                              \n",
       "[319] \"BMP6_ACVR1_ACVR2A\"                       \n",
       "[320] \"WNT2B_FZD1_LRP5\"                         \n",
       "[321] \"WNT2B_FZD6_LRP5\"                         \n",
       "[322] \"WNT2B_FZD7_LRP5\"                         \n",
       "[323] \"WNT2B_FZD8_LRP5\"                         \n",
       "[324] \"WNT8A_FZD6_LRP5\"                         \n",
       "[325] \"WNT2B_FZD1_LRP6\"                         \n",
       "[326] \"WNT2B_FZD6_LRP6\"                         \n",
       "[327] \"WNT2B_FZD8_LRP6\"                         \n",
       "[328] \"WNT8A_FZD6_LRP6\"                         \n",
       "[329] \"WNT5B_FZD6\"                              \n",
       "[330] \"NRG3_ERBB4\"                              \n",
       "[331] \"FGF2_FGFR4\"                              \n",
       "[332] \"PDGFA_PDGFRA\"                            \n",
       "[333] \"PDGFA_PDGFRB\"                            \n",
       "[334] \"VEGFA_VEGFR1R2\"                          \n",
       "[335] \"CCL8_ACKR1\"                              \n",
       "[336] \"CXCL12_ACKR3\"                            \n",
       "[337] \"MIF_ACKR3\"                               \n",
       "[338] \"CTF1_LIFR_IL6ST\"                         \n",
       "[339] \"SPP1_CD44\"                               \n",
       "[340] \"SPP1_ITGAV_ITGB1\"                        \n",
       "[341] \"SPP1_ITGAV_ITGB5\"                        \n",
       "[342] \"SPP1_ITGA9_ITGB1\"                        \n",
       "[343] \"SPP1_ITGA5_ITGB1\"                        \n",
       "[344] \"ANGPTL2_TLR4\"                            \n",
       "[345] \"PTN_SDC3\"                                \n",
       "[346] \"SEMA3C_NRP1_PLXNA1\"                      \n",
       "[347] \"SEMA3C_NRP2_PLXNA1\"                      \n",
       "[348] \"SEMA3F_NRP2_PLXNA1\"                      \n",
       "[349] \"SEMA3F_NRP2_PLXNA4\"                      \n",
       "[350] \"GAS6_MERTK\"                              \n",
       "[351] \"SLIT2_ROBO1\"                             \n",
       "[352] \"SLIT2_ROBO2\"                             \n",
       "[353] \"PROS1_MERTK\"                             \n",
       "[354] \"THBS4_ITGA3_ITGB1\"                       \n",
       "[355] \"COL6A6_ITGA9_ITGB1\"                      \n",
       "[356] \"LAMC2_ITGA9_ITGB1\"                       \n",
       "[357] \"LAMC2_ITGAV_ITGB8\"                       \n",
       "[358] \"COL6A6_ITGAV_ITGB8\"                      \n",
       "[359] \"COL6A6_CD44\"                             \n",
       "[360] \"COL6A6_SDC4\"                             \n",
       "[361] \"THBS4_SDC4\"                              \n",
       "[362] \"LAMA2_SV2A\"                              \n",
       "[363] \"LAMA4_SV2A\"                              \n",
       "[364] \"LAMB2_SV2A\"                              \n",
       "[365] \"LAMC1_SV2A\"                              \n",
       "[366] \"LAMC2_SV2A\"                              \n",
       "[367] \"THBS4_CD36\"                              \n",
       "[368] \"THBS4_CD47\"                              \n",
       "[369] \"AGRN_DAG1\"                               \n",
       "[370] \"LAMC2_DAG1\"                              \n",
       "[371] \"PGF2a-AKR1C3_PTGFR\"                      \n",
       "[372] \"PGF2a-CBR1_PTGFR\"                        \n",
       "[373] \"RetinoicAcid-RA-ALDH1A1_RARA_CRABP2\"     \n",
       "[374] \"RetinoicAcid-RA-ALDH1A1_RARA_RXRA_CRABP2\"\n",
       "[375] \"RetinoicAcid-RA-ALDH1A1_RXRA_CRABP2\"     \n",
       "[376] \"RetinoicAcid-RA-ALDH1A2_RARA_CRABP2\"     \n",
       "[377] \"RetinoicAcid-RA-ALDH1A2_RARA_RXRA_CRABP2\"\n",
       "[378] \"RetinoicAcid-RA-ALDH1A2_RXRA_CRABP2\"     \n",
       "[379] \"RetinoicAcid-RA-ALDH1A3_RARA_CRABP2\"     \n",
       "[380] \"RetinoicAcid-RA-ALDH1A3_RARA_RXRA_CRABP2\"\n",
       "[381] \"RetinoicAcid-RA-ALDH1A3_RXRA_CRABP2\"     \n",
       "[382] \"CD34_SELP\"                               \n",
       "[383] \"CNTN1_NRCAM\"                             \n",
       "[384] \"NCAM1_NCAM2\"                             \n",
       "[385] \"NECTIN3_NECTIN1\"                         \n",
       "[386] \"NECTIN3_PVR\"                             \n",
       "[387] \"JAG2_NOTCH1\"                             \n",
       "[388] \"JAG2_NOTCH2\"                             \n",
       "[389] \"JAG2_NOTCH3\"                             \n",
       "[390] \"NRXN1_NLGN1\"                             \n",
       "[391] \"NRXN1_NLGN2\"                             \n",
       "[392] \"NRXN2_NLGN2\"                             \n",
       "[393] \"SELE_CD44\"                               \n",
       "[394] \"SELE_GLG1\"                               \n",
       "[395] \"SEMA4D_PLXNB1\"                           \n",
       "[396] \"SEMA4D_PLXNB2\"                           \n",
       "[397] \"SEMA4G_PLXNB2\"                           \n",
       "[398] \"CDH12_ITGA1_ITGB1\"                       \n",
       "[399] \"COL3A1_ADGRG1\"                           \n",
       "[400] \"TGM2_ADGRG1\"                             \n",
       "[401] \"FLRT2_UNC5B\"                             \n",
       "[402] \"NRXN1_CLSTN1\"                            \n",
       "[403] \"NRXN1_CLSTN2\"                            \n",
       "[404] \"NRXN1_DAG1\"                              \n",
       "[405] \"NRXN2_DAG1\"                              \n",
       "[406] \"LRFN5_PTPRD\"                             \n",
       "[407] \"LRFN5_PTPRF\"                             \n",
       "[408] \"LRFN5_PTPRS\"                             \n",
       "[409] \"LRRC4C_PTPRF\"                            \n",
       "[410] \"NRXN1_LRRTM4\"                            \n",
       "[411] \"NRXN3_LRRTM4\"                            \n",
       "[412] \"LRRTM4_PTPRS\"                            \n",
       "[413] \"NTN4_UNC5B\"                              \n",
       "[414] \"TGFB3_TGFBR1_TGFBR2\"                     \n",
       "[415] \"TGFB3_ACVR1_TGFBR1\"                      \n",
       "[416] \"SPP1_ITGA8_ITGB1\"                        \n",
       "[417] \"ANGPTL4_ITGA5_ITGB1\"                     \n",
       "[418] \"ANGPTL4_CDH11\"                           \n",
       "[419] \"ANGPTL4_SDC2\"                            \n",
       "[420] \"ANGPTL4_SDC4\"                            \n",
       "[421] \"EDN1_EDNRA\"                              \n",
       "[422] \"EDN1_EDNRB\"                              \n",
       "[423] \"PGE2-PTGES3_PTGER2\"                      \n",
       "[424] \"RetinoicAcid-RA-ALDH1A1_RORB\"            \n",
       "[425] \"RetinoicAcid-RA-ALDH1A2_RORB\"            \n",
       "[426] \"RetinoicAcid-RA-ALDH1A3_RORB\"            \n",
       "[427] \"Dehydroepiandrosterone-DHEA-STS_ESR1\"    \n",
       "[428] \"CDH1_CDH1\"                               \n",
       "[429] \"FLRT3_ADGRL2\"                            \n",
       "[430] \"FLRT3_ADGRL3\"                            \n",
       "[431] \"FLRT3_FLRT3\"                             \n",
       "[432] \"FLRT3_UNC5C\"                             \n",
       "[433] \"NRXN2_CLSTN1\"                            \n",
       "[434] \"NRXN2_CLSTN2\"                            \n",
       "[435] \"NRXN1_LRRTM3\"                            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"in all samples\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TGFB1_TGFBR1_TGFBR2'</li><li>'TGFB2_TGFBR1_TGFBR2'</li><li>'TGFB1_ACVR1_TGFBR1'</li><li>'TGFB2_ACVR1_TGFBR1'</li><li>'AREG_EGFR'</li><li>'NRG1_ERBB4'</li><li>'FGF2_FGFR1'</li><li>'FGF7_FGFR1'</li><li>'PDGFC_PDGFRA'</li><li>'PDGFD_PDGFRB'</li><li>'VEGFB_VEGFR1'</li><li>'PGF_VEGFR1'</li><li>'IGF2_IGF1R'</li><li>'IGF2_IGF2R'</li><li>'CCL2_ACKR1'</li><li>'CCL14_ACKR1'</li><li>'MIF_CD74_CD44'</li><li>'TNFSF12_TNFRSF12A'</li><li>'NAMPT_ITGA5_ITGB1'</li><li>'ANGPTL2_ITGA5_ITGB1'</li><li>'ANGPT2_TEK'</li><li>'ANGPT2_ITGA5_ITGB1'</li><li>'MDK_SDC2'</li><li>'MDK_SDC4'</li><li>'MDK_ITGA6_ITGB1'</li><li>'MDK_LRP1'</li><li>'MDK_NCL'</li><li>'PTN_SDC2'</li><li>'PTN_SDC4'</li><li>'PTN_NCL'</li><li>'SEMA3C_NRP1_PLXNA2'</li><li>'SEMA3C_NRP1_PLXNA4'</li><li>'SEMA3C_NRP2_PLXNA2'</li><li>'SEMA3C_NRP2_PLXNA4'</li><li>'SEMA3G_NRP2_PLXNA2'</li><li>'SEMA3G_NRP2_PLXNA4'</li><li>'SEMA3C_NRP1_NRP2'</li><li>'SEMA3C_PLXND1'</li><li>'GAS6_AXL'</li><li>'GAS6_TYRO3'</li><li>'GRN_SORT1'</li><li>'PROS1_AXL'</li><li>'PROS1_TYRO3'</li><li>'PTPRS_NTRK3'</li><li>'IGFBP3_TMEM219'</li><li>'PPIA_BSG'</li><li>'COL1A1_ITGA1_ITGB1'</li><li>'COL1A2_ITGA1_ITGB1'</li><li>'COL4A1_ITGA1_ITGB1'</li><li>'COL4A2_ITGA1_ITGB1'</li><li>'COL6A1_ITGA1_ITGB1'</li><li>'COL6A2_ITGA1_ITGB1'</li><li>'COL6A3_ITGA1_ITGB1'</li><li>'FN1_ITGA3_ITGB1'</li><li>'FN1_ITGA5_ITGB1'</li><li>'LAMA2_ITGA1_ITGB1'</li><li>'LAMA4_ITGA1_ITGB1'</li><li>'LAMA5_ITGA1_ITGB1'</li><li>'LAMB1_ITGA1_ITGB1'</li><li>'LAMB2_ITGA1_ITGB1'</li><li>'LAMC1_ITGA1_ITGB1'</li><li>'LAMA2_ITGA3_ITGB1'</li><li>'LAMA4_ITGA3_ITGB1'</li><li>'LAMA5_ITGA3_ITGB1'</li><li>'LAMB1_ITGA3_ITGB1'</li><li>'LAMB2_ITGA3_ITGB1'</li><li>'LAMC1_ITGA3_ITGB1'</li><li>'THBS1_ITGA3_ITGB1'</li><li>'THBS2_ITGA3_ITGB1'</li><li>'COL1A1_ITGA3_ITGB1'</li><li>'COL1A2_ITGA3_ITGB1'</li><li>'COL4A1_ITGA3_ITGB1'</li><li>'COL4A2_ITGA3_ITGB1'</li><li>'COL6A1_ITGA3_ITGB1'</li><li>'COL6A2_ITGA3_ITGB1'</li><li>'COL6A3_ITGA3_ITGB1'</li><li>'LAMA2_ITGA6_ITGB1'</li><li>'LAMA4_ITGA6_ITGB1'</li><li>'LAMA5_ITGA6_ITGB1'</li><li>'LAMB1_ITGA6_ITGB1'</li><li>'LAMB2_ITGA6_ITGB1'</li><li>'LAMC1_ITGA6_ITGB1'</li><li>'LAMA2_ITGA7_ITGB1'</li><li>'LAMA4_ITGA7_ITGB1'</li><li>'LAMA5_ITGA7_ITGB1'</li><li>'LAMB1_ITGA7_ITGB1'</li><li>'LAMB2_ITGA7_ITGB1'</li><li>'LAMC1_ITGA7_ITGB1'</li><li>'TNXB_ITGA9_ITGB1'</li><li>'COL1A1_ITGA9_ITGB1'</li><li>'COL1A2_ITGA9_ITGB1'</li><li>'COL4A1_ITGA9_ITGB1'</li><li>'COL4A2_ITGA9_ITGB1'</li><li>'COL6A1_ITGA9_ITGB1'</li><li>'COL6A2_ITGA9_ITGB1'</li><li>'COL6A3_ITGA9_ITGB1'</li><li>'LAMA2_ITGA9_ITGB1'</li><li>'LAMA4_ITGA9_ITGB1'</li><li>'LAMA5_ITGA9_ITGB1'</li><li>'LAMB1_ITGA9_ITGB1'</li><li>'LAMB2_ITGA9_ITGB1'</li><li>'LAMC1_ITGA9_ITGB1'</li><li>'FN1_ITGAV_ITGB1'</li><li>'LAMA2_ITGAV_ITGB8'</li><li>'LAMA4_ITGAV_ITGB8'</li><li>'FN1_CD44'</li><li>'COL1A1_CD44'</li><li>'COL1A2_CD44'</li><li>'COL4A1_CD44'</li><li>'COL4A2_CD44'</li><li>'COL6A1_CD44'</li><li>'COL6A2_CD44'</li><li>'COL6A3_CD44'</li><li>'LAMA2_CD44'</li><li>'LAMA4_CD44'</li><li>'LAMA5_CD44'</li><li>'LAMB1_CD44'</li><li>'LAMB2_CD44'</li><li>'LAMC1_CD44'</li><li>'COL1A1_SDC4'</li><li>'COL1A2_SDC4'</li><li>'COL4A1_SDC4'</li><li>'COL4A2_SDC4'</li><li>'COL6A1_SDC4'</li><li>'COL6A2_SDC4'</li><li>'COL6A3_SDC4'</li><li>'FN1_SDC4'</li><li>'TNXB_SDC4'</li><li>'THBS1_SDC4'</li><li>'THBS2_SDC4'</li><li>'THBS1_CD47'</li><li>'THBS2_CD47'</li><li>'HSPG2_DAG1'</li><li>'LAMA2_DAG1'</li><li>'LAMA4_DAG1'</li><li>'LAMA5_DAG1'</li><li>'LAMB1_DAG1'</li><li>'LAMB2_DAG1'</li><li>'LAMC1_DAG1'</li><li>'ANDROSTENEDIONE-CYP17A1_AR'</li><li>'Cholesterol-Cholesterol-DHCR24_RORA'</li><li>'Cholesterol-Cholesterol-DHCR7_RORA'</li><li>'Desmosterol-Desmosterol-DHCR7_NR1H2'</li><li>'Testosterone-Testosterone-HSD17B12_AR'</li><li>'APP_CD74'</li><li>'CADM1_CADM1'</li><li>'CD46_JAG1'</li><li>'CD99_CD99'</li><li>'CD99_CD99L2'</li><li>'CDH4_CDH4'</li><li>'CDH5_CDH5'</li><li>'CLDN11_CLDN11'</li><li>'CSPG4_ITGA2_ITGB1'</li><li>'EFNA1_EPHA3'</li><li>'EFNA5_EPHA3'</li><li>'EFNB2_EPHB4'</li><li>'ESAM_ESAM'</li><li>'JAM2_ITGA3_ITGB1'</li><li>'JAM2_ITGAV_ITGB1'</li><li>'JAM2_JAM3'</li><li>'JAM2_JAM2'</li><li>'JAM3_JAM3'</li><li>'MPZL1_MPZL1'</li><li>'NCAM1_FGFR1'</li><li>'NCAM1_NCAM1'</li><li>'NECTIN3_NECTIN2'</li><li>'NEGR1_NEGR1'</li><li>'DLK1_NOTCH2'</li><li>'DLK1_NOTCH3'</li><li>'JAG1_NOTCH2'</li><li>'JAG1_NOTCH3'</li><li>'NRXN3_NLGN1'</li><li>'NRXN3_NLGN2'</li><li>'PECAM1_PECAM1'</li><li>'PTPRM_PTPRM'</li><li>'SEMA4C_PLXNB2'</li><li>'SEMA6A_PLXNA2'</li><li>'SEMA6A_PLXNA4'</li><li>'ITGA9_ITGB1_VCAM1'</li><li>'CDH3_CDH3'</li><li>'GJA5_GJA5'</li><li>'GJA1_GJA5'</li><li>'GJC1_GJA5'</li><li>'GJA1_GJA1'</li><li>'GJC1_GJA1'</li><li>'GJC1_GJC1'</li><li>'PCDHGA1_PCDHGA1'</li><li>'SDC2_ADGRA2'</li><li>'CD55_ADGRE5'</li><li>'LPAR1_ADGRE5'</li><li>'ITGAV_ITGB1_ADGRE5'</li><li>'FLRT2_ADGRL2'</li><li>'TENM2_ADGRL2'</li><li>'TENM3_ADGRL2'</li><li>'TENM4_ADGRL2'</li><li>'FLRT2_ADGRL3'</li><li>'TENM2_ADGRL3'</li><li>'TENM3_ADGRL3'</li><li>'TENM4_ADGRL3'</li><li>'FLRT2_FLRT2'</li><li>'FLRT2_UNC5C'</li><li>'Q9HDB5_CLSTN1'</li><li>'Q9HDB5_CLSTN2'</li><li>'CADM1_NECTIN3'</li><li>'CNTN1_NOTCH2'</li><li>'NTN4_NTRK2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TGFB1\\_TGFBR1\\_TGFBR2'\n",
       "\\item 'TGFB2\\_TGFBR1\\_TGFBR2'\n",
       "\\item 'TGFB1\\_ACVR1\\_TGFBR1'\n",
       "\\item 'TGFB2\\_ACVR1\\_TGFBR1'\n",
       "\\item 'AREG\\_EGFR'\n",
       "\\item 'NRG1\\_ERBB4'\n",
       "\\item 'FGF2\\_FGFR1'\n",
       "\\item 'FGF7\\_FGFR1'\n",
       "\\item 'PDGFC\\_PDGFRA'\n",
       "\\item 'PDGFD\\_PDGFRB'\n",
       "\\item 'VEGFB\\_VEGFR1'\n",
       "\\item 'PGF\\_VEGFR1'\n",
       "\\item 'IGF2\\_IGF1R'\n",
       "\\item 'IGF2\\_IGF2R'\n",
       "\\item 'CCL2\\_ACKR1'\n",
       "\\item 'CCL14\\_ACKR1'\n",
       "\\item 'MIF\\_CD74\\_CD44'\n",
       "\\item 'TNFSF12\\_TNFRSF12A'\n",
       "\\item 'NAMPT\\_ITGA5\\_ITGB1'\n",
       "\\item 'ANGPTL2\\_ITGA5\\_ITGB1'\n",
       "\\item 'ANGPT2\\_TEK'\n",
       "\\item 'ANGPT2\\_ITGA5\\_ITGB1'\n",
       "\\item 'MDK\\_SDC2'\n",
       "\\item 'MDK\\_SDC4'\n",
       "\\item 'MDK\\_ITGA6\\_ITGB1'\n",
       "\\item 'MDK\\_LRP1'\n",
       "\\item 'MDK\\_NCL'\n",
       "\\item 'PTN\\_SDC2'\n",
       "\\item 'PTN\\_SDC4'\n",
       "\\item 'PTN\\_NCL'\n",
       "\\item 'SEMA3C\\_NRP1\\_PLXNA2'\n",
       "\\item 'SEMA3C\\_NRP1\\_PLXNA4'\n",
       "\\item 'SEMA3C\\_NRP2\\_PLXNA2'\n",
       "\\item 'SEMA3C\\_NRP2\\_PLXNA4'\n",
       "\\item 'SEMA3G\\_NRP2\\_PLXNA2'\n",
       "\\item 'SEMA3G\\_NRP2\\_PLXNA4'\n",
       "\\item 'SEMA3C\\_NRP1\\_NRP2'\n",
       "\\item 'SEMA3C\\_PLXND1'\n",
       "\\item 'GAS6\\_AXL'\n",
       "\\item 'GAS6\\_TYRO3'\n",
       "\\item 'GRN\\_SORT1'\n",
       "\\item 'PROS1\\_AXL'\n",
       "\\item 'PROS1\\_TYRO3'\n",
       "\\item 'PTPRS\\_NTRK3'\n",
       "\\item 'IGFBP3\\_TMEM219'\n",
       "\\item 'PPIA\\_BSG'\n",
       "\\item 'COL1A1\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL1A2\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL4A1\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL4A2\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL6A1\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL6A2\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL6A3\\_ITGA1\\_ITGB1'\n",
       "\\item 'FN1\\_ITGA3\\_ITGB1'\n",
       "\\item 'FN1\\_ITGA5\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA3\\_ITGB1'\n",
       "\\item 'THBS1\\_ITGA3\\_ITGB1'\n",
       "\\item 'THBS2\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL1A1\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL1A2\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL4A1\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL4A2\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL6A1\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL6A2\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL6A3\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA7\\_ITGB1'\n",
       "\\item 'TNXB\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL1A1\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL1A2\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL4A1\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL4A2\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL6A1\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL6A2\\_ITGA9\\_ITGB1'\n",
       "\\item 'COL6A3\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA4\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA5\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMB1\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMB2\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMC1\\_ITGA9\\_ITGB1'\n",
       "\\item 'FN1\\_ITGAV\\_ITGB1'\n",
       "\\item 'LAMA2\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMA4\\_ITGAV\\_ITGB8'\n",
       "\\item 'FN1\\_CD44'\n",
       "\\item 'COL1A1\\_CD44'\n",
       "\\item 'COL1A2\\_CD44'\n",
       "\\item 'COL4A1\\_CD44'\n",
       "\\item 'COL4A2\\_CD44'\n",
       "\\item 'COL6A1\\_CD44'\n",
       "\\item 'COL6A2\\_CD44'\n",
       "\\item 'COL6A3\\_CD44'\n",
       "\\item 'LAMA2\\_CD44'\n",
       "\\item 'LAMA4\\_CD44'\n",
       "\\item 'LAMA5\\_CD44'\n",
       "\\item 'LAMB1\\_CD44'\n",
       "\\item 'LAMB2\\_CD44'\n",
       "\\item 'LAMC1\\_CD44'\n",
       "\\item 'COL1A1\\_SDC4'\n",
       "\\item 'COL1A2\\_SDC4'\n",
       "\\item 'COL4A1\\_SDC4'\n",
       "\\item 'COL4A2\\_SDC4'\n",
       "\\item 'COL6A1\\_SDC4'\n",
       "\\item 'COL6A2\\_SDC4'\n",
       "\\item 'COL6A3\\_SDC4'\n",
       "\\item 'FN1\\_SDC4'\n",
       "\\item 'TNXB\\_SDC4'\n",
       "\\item 'THBS1\\_SDC4'\n",
       "\\item 'THBS2\\_SDC4'\n",
       "\\item 'THBS1\\_CD47'\n",
       "\\item 'THBS2\\_CD47'\n",
       "\\item 'HSPG2\\_DAG1'\n",
       "\\item 'LAMA2\\_DAG1'\n",
       "\\item 'LAMA4\\_DAG1'\n",
       "\\item 'LAMA5\\_DAG1'\n",
       "\\item 'LAMB1\\_DAG1'\n",
       "\\item 'LAMB2\\_DAG1'\n",
       "\\item 'LAMC1\\_DAG1'\n",
       "\\item 'ANDROSTENEDIONE-CYP17A1\\_AR'\n",
       "\\item 'Cholesterol-Cholesterol-DHCR24\\_RORA'\n",
       "\\item 'Cholesterol-Cholesterol-DHCR7\\_RORA'\n",
       "\\item 'Desmosterol-Desmosterol-DHCR7\\_NR1H2'\n",
       "\\item 'Testosterone-Testosterone-HSD17B12\\_AR'\n",
       "\\item 'APP\\_CD74'\n",
       "\\item 'CADM1\\_CADM1'\n",
       "\\item 'CD46\\_JAG1'\n",
       "\\item 'CD99\\_CD99'\n",
       "\\item 'CD99\\_CD99L2'\n",
       "\\item 'CDH4\\_CDH4'\n",
       "\\item 'CDH5\\_CDH5'\n",
       "\\item 'CLDN11\\_CLDN11'\n",
       "\\item 'CSPG4\\_ITGA2\\_ITGB1'\n",
       "\\item 'EFNA1\\_EPHA3'\n",
       "\\item 'EFNA5\\_EPHA3'\n",
       "\\item 'EFNB2\\_EPHB4'\n",
       "\\item 'ESAM\\_ESAM'\n",
       "\\item 'JAM2\\_ITGA3\\_ITGB1'\n",
       "\\item 'JAM2\\_ITGAV\\_ITGB1'\n",
       "\\item 'JAM2\\_JAM3'\n",
       "\\item 'JAM2\\_JAM2'\n",
       "\\item 'JAM3\\_JAM3'\n",
       "\\item 'MPZL1\\_MPZL1'\n",
       "\\item 'NCAM1\\_FGFR1'\n",
       "\\item 'NCAM1\\_NCAM1'\n",
       "\\item 'NECTIN3\\_NECTIN2'\n",
       "\\item 'NEGR1\\_NEGR1'\n",
       "\\item 'DLK1\\_NOTCH2'\n",
       "\\item 'DLK1\\_NOTCH3'\n",
       "\\item 'JAG1\\_NOTCH2'\n",
       "\\item 'JAG1\\_NOTCH3'\n",
       "\\item 'NRXN3\\_NLGN1'\n",
       "\\item 'NRXN3\\_NLGN2'\n",
       "\\item 'PECAM1\\_PECAM1'\n",
       "\\item 'PTPRM\\_PTPRM'\n",
       "\\item 'SEMA4C\\_PLXNB2'\n",
       "\\item 'SEMA6A\\_PLXNA2'\n",
       "\\item 'SEMA6A\\_PLXNA4'\n",
       "\\item 'ITGA9\\_ITGB1\\_VCAM1'\n",
       "\\item 'CDH3\\_CDH3'\n",
       "\\item 'GJA5\\_GJA5'\n",
       "\\item 'GJA1\\_GJA5'\n",
       "\\item 'GJC1\\_GJA5'\n",
       "\\item 'GJA1\\_GJA1'\n",
       "\\item 'GJC1\\_GJA1'\n",
       "\\item 'GJC1\\_GJC1'\n",
       "\\item 'PCDHGA1\\_PCDHGA1'\n",
       "\\item 'SDC2\\_ADGRA2'\n",
       "\\item 'CD55\\_ADGRE5'\n",
       "\\item 'LPAR1\\_ADGRE5'\n",
       "\\item 'ITGAV\\_ITGB1\\_ADGRE5'\n",
       "\\item 'FLRT2\\_ADGRL2'\n",
       "\\item 'TENM2\\_ADGRL2'\n",
       "\\item 'TENM3\\_ADGRL2'\n",
       "\\item 'TENM4\\_ADGRL2'\n",
       "\\item 'FLRT2\\_ADGRL3'\n",
       "\\item 'TENM2\\_ADGRL3'\n",
       "\\item 'TENM3\\_ADGRL3'\n",
       "\\item 'TENM4\\_ADGRL3'\n",
       "\\item 'FLRT2\\_FLRT2'\n",
       "\\item 'FLRT2\\_UNC5C'\n",
       "\\item 'Q9HDB5\\_CLSTN1'\n",
       "\\item 'Q9HDB5\\_CLSTN2'\n",
       "\\item 'CADM1\\_NECTIN3'\n",
       "\\item 'CNTN1\\_NOTCH2'\n",
       "\\item 'NTN4\\_NTRK2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TGFB1_TGFBR1_TGFBR2'\n",
       "2. 'TGFB2_TGFBR1_TGFBR2'\n",
       "3. 'TGFB1_ACVR1_TGFBR1'\n",
       "4. 'TGFB2_ACVR1_TGFBR1'\n",
       "5. 'AREG_EGFR'\n",
       "6. 'NRG1_ERBB4'\n",
       "7. 'FGF2_FGFR1'\n",
       "8. 'FGF7_FGFR1'\n",
       "9. 'PDGFC_PDGFRA'\n",
       "10. 'PDGFD_PDGFRB'\n",
       "11. 'VEGFB_VEGFR1'\n",
       "12. 'PGF_VEGFR1'\n",
       "13. 'IGF2_IGF1R'\n",
       "14. 'IGF2_IGF2R'\n",
       "15. 'CCL2_ACKR1'\n",
       "16. 'CCL14_ACKR1'\n",
       "17. 'MIF_CD74_CD44'\n",
       "18. 'TNFSF12_TNFRSF12A'\n",
       "19. 'NAMPT_ITGA5_ITGB1'\n",
       "20. 'ANGPTL2_ITGA5_ITGB1'\n",
       "21. 'ANGPT2_TEK'\n",
       "22. 'ANGPT2_ITGA5_ITGB1'\n",
       "23. 'MDK_SDC2'\n",
       "24. 'MDK_SDC4'\n",
       "25. 'MDK_ITGA6_ITGB1'\n",
       "26. 'MDK_LRP1'\n",
       "27. 'MDK_NCL'\n",
       "28. 'PTN_SDC2'\n",
       "29. 'PTN_SDC4'\n",
       "30. 'PTN_NCL'\n",
       "31. 'SEMA3C_NRP1_PLXNA2'\n",
       "32. 'SEMA3C_NRP1_PLXNA4'\n",
       "33. 'SEMA3C_NRP2_PLXNA2'\n",
       "34. 'SEMA3C_NRP2_PLXNA4'\n",
       "35. 'SEMA3G_NRP2_PLXNA2'\n",
       "36. 'SEMA3G_NRP2_PLXNA4'\n",
       "37. 'SEMA3C_NRP1_NRP2'\n",
       "38. 'SEMA3C_PLXND1'\n",
       "39. 'GAS6_AXL'\n",
       "40. 'GAS6_TYRO3'\n",
       "41. 'GRN_SORT1'\n",
       "42. 'PROS1_AXL'\n",
       "43. 'PROS1_TYRO3'\n",
       "44. 'PTPRS_NTRK3'\n",
       "45. 'IGFBP3_TMEM219'\n",
       "46. 'PPIA_BSG'\n",
       "47. 'COL1A1_ITGA1_ITGB1'\n",
       "48. 'COL1A2_ITGA1_ITGB1'\n",
       "49. 'COL4A1_ITGA1_ITGB1'\n",
       "50. 'COL4A2_ITGA1_ITGB1'\n",
       "51. 'COL6A1_ITGA1_ITGB1'\n",
       "52. 'COL6A2_ITGA1_ITGB1'\n",
       "53. 'COL6A3_ITGA1_ITGB1'\n",
       "54. 'FN1_ITGA3_ITGB1'\n",
       "55. 'FN1_ITGA5_ITGB1'\n",
       "56. 'LAMA2_ITGA1_ITGB1'\n",
       "57. 'LAMA4_ITGA1_ITGB1'\n",
       "58. 'LAMA5_ITGA1_ITGB1'\n",
       "59. 'LAMB1_ITGA1_ITGB1'\n",
       "60. 'LAMB2_ITGA1_ITGB1'\n",
       "61. 'LAMC1_ITGA1_ITGB1'\n",
       "62. 'LAMA2_ITGA3_ITGB1'\n",
       "63. 'LAMA4_ITGA3_ITGB1'\n",
       "64. 'LAMA5_ITGA3_ITGB1'\n",
       "65. 'LAMB1_ITGA3_ITGB1'\n",
       "66. 'LAMB2_ITGA3_ITGB1'\n",
       "67. 'LAMC1_ITGA3_ITGB1'\n",
       "68. 'THBS1_ITGA3_ITGB1'\n",
       "69. 'THBS2_ITGA3_ITGB1'\n",
       "70. 'COL1A1_ITGA3_ITGB1'\n",
       "71. 'COL1A2_ITGA3_ITGB1'\n",
       "72. 'COL4A1_ITGA3_ITGB1'\n",
       "73. 'COL4A2_ITGA3_ITGB1'\n",
       "74. 'COL6A1_ITGA3_ITGB1'\n",
       "75. 'COL6A2_ITGA3_ITGB1'\n",
       "76. 'COL6A3_ITGA3_ITGB1'\n",
       "77. 'LAMA2_ITGA6_ITGB1'\n",
       "78. 'LAMA4_ITGA6_ITGB1'\n",
       "79. 'LAMA5_ITGA6_ITGB1'\n",
       "80. 'LAMB1_ITGA6_ITGB1'\n",
       "81. 'LAMB2_ITGA6_ITGB1'\n",
       "82. 'LAMC1_ITGA6_ITGB1'\n",
       "83. 'LAMA2_ITGA7_ITGB1'\n",
       "84. 'LAMA4_ITGA7_ITGB1'\n",
       "85. 'LAMA5_ITGA7_ITGB1'\n",
       "86. 'LAMB1_ITGA7_ITGB1'\n",
       "87. 'LAMB2_ITGA7_ITGB1'\n",
       "88. 'LAMC1_ITGA7_ITGB1'\n",
       "89. 'TNXB_ITGA9_ITGB1'\n",
       "90. 'COL1A1_ITGA9_ITGB1'\n",
       "91. 'COL1A2_ITGA9_ITGB1'\n",
       "92. 'COL4A1_ITGA9_ITGB1'\n",
       "93. 'COL4A2_ITGA9_ITGB1'\n",
       "94. 'COL6A1_ITGA9_ITGB1'\n",
       "95. 'COL6A2_ITGA9_ITGB1'\n",
       "96. 'COL6A3_ITGA9_ITGB1'\n",
       "97. 'LAMA2_ITGA9_ITGB1'\n",
       "98. 'LAMA4_ITGA9_ITGB1'\n",
       "99. 'LAMA5_ITGA9_ITGB1'\n",
       "100. 'LAMB1_ITGA9_ITGB1'\n",
       "101. 'LAMB2_ITGA9_ITGB1'\n",
       "102. 'LAMC1_ITGA9_ITGB1'\n",
       "103. 'FN1_ITGAV_ITGB1'\n",
       "104. 'LAMA2_ITGAV_ITGB8'\n",
       "105. 'LAMA4_ITGAV_ITGB8'\n",
       "106. 'FN1_CD44'\n",
       "107. 'COL1A1_CD44'\n",
       "108. 'COL1A2_CD44'\n",
       "109. 'COL4A1_CD44'\n",
       "110. 'COL4A2_CD44'\n",
       "111. 'COL6A1_CD44'\n",
       "112. 'COL6A2_CD44'\n",
       "113. 'COL6A3_CD44'\n",
       "114. 'LAMA2_CD44'\n",
       "115. 'LAMA4_CD44'\n",
       "116. 'LAMA5_CD44'\n",
       "117. 'LAMB1_CD44'\n",
       "118. 'LAMB2_CD44'\n",
       "119. 'LAMC1_CD44'\n",
       "120. 'COL1A1_SDC4'\n",
       "121. 'COL1A2_SDC4'\n",
       "122. 'COL4A1_SDC4'\n",
       "123. 'COL4A2_SDC4'\n",
       "124. 'COL6A1_SDC4'\n",
       "125. 'COL6A2_SDC4'\n",
       "126. 'COL6A3_SDC4'\n",
       "127. 'FN1_SDC4'\n",
       "128. 'TNXB_SDC4'\n",
       "129. 'THBS1_SDC4'\n",
       "130. 'THBS2_SDC4'\n",
       "131. 'THBS1_CD47'\n",
       "132. 'THBS2_CD47'\n",
       "133. 'HSPG2_DAG1'\n",
       "134. 'LAMA2_DAG1'\n",
       "135. 'LAMA4_DAG1'\n",
       "136. 'LAMA5_DAG1'\n",
       "137. 'LAMB1_DAG1'\n",
       "138. 'LAMB2_DAG1'\n",
       "139. 'LAMC1_DAG1'\n",
       "140. 'ANDROSTENEDIONE-CYP17A1_AR'\n",
       "141. 'Cholesterol-Cholesterol-DHCR24_RORA'\n",
       "142. 'Cholesterol-Cholesterol-DHCR7_RORA'\n",
       "143. 'Desmosterol-Desmosterol-DHCR7_NR1H2'\n",
       "144. 'Testosterone-Testosterone-HSD17B12_AR'\n",
       "145. 'APP_CD74'\n",
       "146. 'CADM1_CADM1'\n",
       "147. 'CD46_JAG1'\n",
       "148. 'CD99_CD99'\n",
       "149. 'CD99_CD99L2'\n",
       "150. 'CDH4_CDH4'\n",
       "151. 'CDH5_CDH5'\n",
       "152. 'CLDN11_CLDN11'\n",
       "153. 'CSPG4_ITGA2_ITGB1'\n",
       "154. 'EFNA1_EPHA3'\n",
       "155. 'EFNA5_EPHA3'\n",
       "156. 'EFNB2_EPHB4'\n",
       "157. 'ESAM_ESAM'\n",
       "158. 'JAM2_ITGA3_ITGB1'\n",
       "159. 'JAM2_ITGAV_ITGB1'\n",
       "160. 'JAM2_JAM3'\n",
       "161. 'JAM2_JAM2'\n",
       "162. 'JAM3_JAM3'\n",
       "163. 'MPZL1_MPZL1'\n",
       "164. 'NCAM1_FGFR1'\n",
       "165. 'NCAM1_NCAM1'\n",
       "166. 'NECTIN3_NECTIN2'\n",
       "167. 'NEGR1_NEGR1'\n",
       "168. 'DLK1_NOTCH2'\n",
       "169. 'DLK1_NOTCH3'\n",
       "170. 'JAG1_NOTCH2'\n",
       "171. 'JAG1_NOTCH3'\n",
       "172. 'NRXN3_NLGN1'\n",
       "173. 'NRXN3_NLGN2'\n",
       "174. 'PECAM1_PECAM1'\n",
       "175. 'PTPRM_PTPRM'\n",
       "176. 'SEMA4C_PLXNB2'\n",
       "177. 'SEMA6A_PLXNA2'\n",
       "178. 'SEMA6A_PLXNA4'\n",
       "179. 'ITGA9_ITGB1_VCAM1'\n",
       "180. 'CDH3_CDH3'\n",
       "181. 'GJA5_GJA5'\n",
       "182. 'GJA1_GJA5'\n",
       "183. 'GJC1_GJA5'\n",
       "184. 'GJA1_GJA1'\n",
       "185. 'GJC1_GJA1'\n",
       "186. 'GJC1_GJC1'\n",
       "187. 'PCDHGA1_PCDHGA1'\n",
       "188. 'SDC2_ADGRA2'\n",
       "189. 'CD55_ADGRE5'\n",
       "190. 'LPAR1_ADGRE5'\n",
       "191. 'ITGAV_ITGB1_ADGRE5'\n",
       "192. 'FLRT2_ADGRL2'\n",
       "193. 'TENM2_ADGRL2'\n",
       "194. 'TENM3_ADGRL2'\n",
       "195. 'TENM4_ADGRL2'\n",
       "196. 'FLRT2_ADGRL3'\n",
       "197. 'TENM2_ADGRL3'\n",
       "198. 'TENM3_ADGRL3'\n",
       "199. 'TENM4_ADGRL3'\n",
       "200. 'FLRT2_FLRT2'\n",
       "201. 'FLRT2_UNC5C'\n",
       "202. 'Q9HDB5_CLSTN1'\n",
       "203. 'Q9HDB5_CLSTN2'\n",
       "204. 'CADM1_NECTIN3'\n",
       "205. 'CNTN1_NOTCH2'\n",
       "206. 'NTN4_NTRK2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"TGFB1_TGFBR1_TGFBR2\"                  \n",
       "  [2] \"TGFB2_TGFBR1_TGFBR2\"                  \n",
       "  [3] \"TGFB1_ACVR1_TGFBR1\"                   \n",
       "  [4] \"TGFB2_ACVR1_TGFBR1\"                   \n",
       "  [5] \"AREG_EGFR\"                            \n",
       "  [6] \"NRG1_ERBB4\"                           \n",
       "  [7] \"FGF2_FGFR1\"                           \n",
       "  [8] \"FGF7_FGFR1\"                           \n",
       "  [9] \"PDGFC_PDGFRA\"                         \n",
       " [10] \"PDGFD_PDGFRB\"                         \n",
       " [11] \"VEGFB_VEGFR1\"                         \n",
       " [12] \"PGF_VEGFR1\"                           \n",
       " [13] \"IGF2_IGF1R\"                           \n",
       " [14] \"IGF2_IGF2R\"                           \n",
       " [15] \"CCL2_ACKR1\"                           \n",
       " [16] \"CCL14_ACKR1\"                          \n",
       " [17] \"MIF_CD74_CD44\"                        \n",
       " [18] \"TNFSF12_TNFRSF12A\"                    \n",
       " [19] \"NAMPT_ITGA5_ITGB1\"                    \n",
       " [20] \"ANGPTL2_ITGA5_ITGB1\"                  \n",
       " [21] \"ANGPT2_TEK\"                           \n",
       " [22] \"ANGPT2_ITGA5_ITGB1\"                   \n",
       " [23] \"MDK_SDC2\"                             \n",
       " [24] \"MDK_SDC4\"                             \n",
       " [25] \"MDK_ITGA6_ITGB1\"                      \n",
       " [26] \"MDK_LRP1\"                             \n",
       " [27] \"MDK_NCL\"                              \n",
       " [28] \"PTN_SDC2\"                             \n",
       " [29] \"PTN_SDC4\"                             \n",
       " [30] \"PTN_NCL\"                              \n",
       " [31] \"SEMA3C_NRP1_PLXNA2\"                   \n",
       " [32] \"SEMA3C_NRP1_PLXNA4\"                   \n",
       " [33] \"SEMA3C_NRP2_PLXNA2\"                   \n",
       " [34] \"SEMA3C_NRP2_PLXNA4\"                   \n",
       " [35] \"SEMA3G_NRP2_PLXNA2\"                   \n",
       " [36] \"SEMA3G_NRP2_PLXNA4\"                   \n",
       " [37] \"SEMA3C_NRP1_NRP2\"                     \n",
       " [38] \"SEMA3C_PLXND1\"                        \n",
       " [39] \"GAS6_AXL\"                             \n",
       " [40] \"GAS6_TYRO3\"                           \n",
       " [41] \"GRN_SORT1\"                            \n",
       " [42] \"PROS1_AXL\"                            \n",
       " [43] \"PROS1_TYRO3\"                          \n",
       " [44] \"PTPRS_NTRK3\"                          \n",
       " [45] \"IGFBP3_TMEM219\"                       \n",
       " [46] \"PPIA_BSG\"                             \n",
       " [47] \"COL1A1_ITGA1_ITGB1\"                   \n",
       " [48] \"COL1A2_ITGA1_ITGB1\"                   \n",
       " [49] \"COL4A1_ITGA1_ITGB1\"                   \n",
       " [50] \"COL4A2_ITGA1_ITGB1\"                   \n",
       " [51] \"COL6A1_ITGA1_ITGB1\"                   \n",
       " [52] \"COL6A2_ITGA1_ITGB1\"                   \n",
       " [53] \"COL6A3_ITGA1_ITGB1\"                   \n",
       " [54] \"FN1_ITGA3_ITGB1\"                      \n",
       " [55] \"FN1_ITGA5_ITGB1\"                      \n",
       " [56] \"LAMA2_ITGA1_ITGB1\"                    \n",
       " [57] \"LAMA4_ITGA1_ITGB1\"                    \n",
       " [58] \"LAMA5_ITGA1_ITGB1\"                    \n",
       " [59] \"LAMB1_ITGA1_ITGB1\"                    \n",
       " [60] \"LAMB2_ITGA1_ITGB1\"                    \n",
       " [61] \"LAMC1_ITGA1_ITGB1\"                    \n",
       " [62] \"LAMA2_ITGA3_ITGB1\"                    \n",
       " [63] \"LAMA4_ITGA3_ITGB1\"                    \n",
       " [64] \"LAMA5_ITGA3_ITGB1\"                    \n",
       " [65] \"LAMB1_ITGA3_ITGB1\"                    \n",
       " [66] \"LAMB2_ITGA3_ITGB1\"                    \n",
       " [67] \"LAMC1_ITGA3_ITGB1\"                    \n",
       " [68] \"THBS1_ITGA3_ITGB1\"                    \n",
       " [69] \"THBS2_ITGA3_ITGB1\"                    \n",
       " [70] \"COL1A1_ITGA3_ITGB1\"                   \n",
       " [71] \"COL1A2_ITGA3_ITGB1\"                   \n",
       " [72] \"COL4A1_ITGA3_ITGB1\"                   \n",
       " [73] \"COL4A2_ITGA3_ITGB1\"                   \n",
       " [74] \"COL6A1_ITGA3_ITGB1\"                   \n",
       " [75] \"COL6A2_ITGA3_ITGB1\"                   \n",
       " [76] \"COL6A3_ITGA3_ITGB1\"                   \n",
       " [77] \"LAMA2_ITGA6_ITGB1\"                    \n",
       " [78] \"LAMA4_ITGA6_ITGB1\"                    \n",
       " [79] \"LAMA5_ITGA6_ITGB1\"                    \n",
       " [80] \"LAMB1_ITGA6_ITGB1\"                    \n",
       " [81] \"LAMB2_ITGA6_ITGB1\"                    \n",
       " [82] \"LAMC1_ITGA6_ITGB1\"                    \n",
       " [83] \"LAMA2_ITGA7_ITGB1\"                    \n",
       " [84] \"LAMA4_ITGA7_ITGB1\"                    \n",
       " [85] \"LAMA5_ITGA7_ITGB1\"                    \n",
       " [86] \"LAMB1_ITGA7_ITGB1\"                    \n",
       " [87] \"LAMB2_ITGA7_ITGB1\"                    \n",
       " [88] \"LAMC1_ITGA7_ITGB1\"                    \n",
       " [89] \"TNXB_ITGA9_ITGB1\"                     \n",
       " [90] \"COL1A1_ITGA9_ITGB1\"                   \n",
       " [91] \"COL1A2_ITGA9_ITGB1\"                   \n",
       " [92] \"COL4A1_ITGA9_ITGB1\"                   \n",
       " [93] \"COL4A2_ITGA9_ITGB1\"                   \n",
       " [94] \"COL6A1_ITGA9_ITGB1\"                   \n",
       " [95] \"COL6A2_ITGA9_ITGB1\"                   \n",
       " [96] \"COL6A3_ITGA9_ITGB1\"                   \n",
       " [97] \"LAMA2_ITGA9_ITGB1\"                    \n",
       " [98] \"LAMA4_ITGA9_ITGB1\"                    \n",
       " [99] \"LAMA5_ITGA9_ITGB1\"                    \n",
       "[100] \"LAMB1_ITGA9_ITGB1\"                    \n",
       "[101] \"LAMB2_ITGA9_ITGB1\"                    \n",
       "[102] \"LAMC1_ITGA9_ITGB1\"                    \n",
       "[103] \"FN1_ITGAV_ITGB1\"                      \n",
       "[104] \"LAMA2_ITGAV_ITGB8\"                    \n",
       "[105] \"LAMA4_ITGAV_ITGB8\"                    \n",
       "[106] \"FN1_CD44\"                             \n",
       "[107] \"COL1A1_CD44\"                          \n",
       "[108] \"COL1A2_CD44\"                          \n",
       "[109] \"COL4A1_CD44\"                          \n",
       "[110] \"COL4A2_CD44\"                          \n",
       "[111] \"COL6A1_CD44\"                          \n",
       "[112] \"COL6A2_CD44\"                          \n",
       "[113] \"COL6A3_CD44\"                          \n",
       "[114] \"LAMA2_CD44\"                           \n",
       "[115] \"LAMA4_CD44\"                           \n",
       "[116] \"LAMA5_CD44\"                           \n",
       "[117] \"LAMB1_CD44\"                           \n",
       "[118] \"LAMB2_CD44\"                           \n",
       "[119] \"LAMC1_CD44\"                           \n",
       "[120] \"COL1A1_SDC4\"                          \n",
       "[121] \"COL1A2_SDC4\"                          \n",
       "[122] \"COL4A1_SDC4\"                          \n",
       "[123] \"COL4A2_SDC4\"                          \n",
       "[124] \"COL6A1_SDC4\"                          \n",
       "[125] \"COL6A2_SDC4\"                          \n",
       "[126] \"COL6A3_SDC4\"                          \n",
       "[127] \"FN1_SDC4\"                             \n",
       "[128] \"TNXB_SDC4\"                            \n",
       "[129] \"THBS1_SDC4\"                           \n",
       "[130] \"THBS2_SDC4\"                           \n",
       "[131] \"THBS1_CD47\"                           \n",
       "[132] \"THBS2_CD47\"                           \n",
       "[133] \"HSPG2_DAG1\"                           \n",
       "[134] \"LAMA2_DAG1\"                           \n",
       "[135] \"LAMA4_DAG1\"                           \n",
       "[136] \"LAMA5_DAG1\"                           \n",
       "[137] \"LAMB1_DAG1\"                           \n",
       "[138] \"LAMB2_DAG1\"                           \n",
       "[139] \"LAMC1_DAG1\"                           \n",
       "[140] \"ANDROSTENEDIONE-CYP17A1_AR\"           \n",
       "[141] \"Cholesterol-Cholesterol-DHCR24_RORA\"  \n",
       "[142] \"Cholesterol-Cholesterol-DHCR7_RORA\"   \n",
       "[143] \"Desmosterol-Desmosterol-DHCR7_NR1H2\"  \n",
       "[144] \"Testosterone-Testosterone-HSD17B12_AR\"\n",
       "[145] \"APP_CD74\"                             \n",
       "[146] \"CADM1_CADM1\"                          \n",
       "[147] \"CD46_JAG1\"                            \n",
       "[148] \"CD99_CD99\"                            \n",
       "[149] \"CD99_CD99L2\"                          \n",
       "[150] \"CDH4_CDH4\"                            \n",
       "[151] \"CDH5_CDH5\"                            \n",
       "[152] \"CLDN11_CLDN11\"                        \n",
       "[153] \"CSPG4_ITGA2_ITGB1\"                    \n",
       "[154] \"EFNA1_EPHA3\"                          \n",
       "[155] \"EFNA5_EPHA3\"                          \n",
       "[156] \"EFNB2_EPHB4\"                          \n",
       "[157] \"ESAM_ESAM\"                            \n",
       "[158] \"JAM2_ITGA3_ITGB1\"                     \n",
       "[159] \"JAM2_ITGAV_ITGB1\"                     \n",
       "[160] \"JAM2_JAM3\"                            \n",
       "[161] \"JAM2_JAM2\"                            \n",
       "[162] \"JAM3_JAM3\"                            \n",
       "[163] \"MPZL1_MPZL1\"                          \n",
       "[164] \"NCAM1_FGFR1\"                          \n",
       "[165] \"NCAM1_NCAM1\"                          \n",
       "[166] \"NECTIN3_NECTIN2\"                      \n",
       "[167] \"NEGR1_NEGR1\"                          \n",
       "[168] \"DLK1_NOTCH2\"                          \n",
       "[169] \"DLK1_NOTCH3\"                          \n",
       "[170] \"JAG1_NOTCH2\"                          \n",
       "[171] \"JAG1_NOTCH3\"                          \n",
       "[172] \"NRXN3_NLGN1\"                          \n",
       "[173] \"NRXN3_NLGN2\"                          \n",
       "[174] \"PECAM1_PECAM1\"                        \n",
       "[175] \"PTPRM_PTPRM\"                          \n",
       "[176] \"SEMA4C_PLXNB2\"                        \n",
       "[177] \"SEMA6A_PLXNA2\"                        \n",
       "[178] \"SEMA6A_PLXNA4\"                        \n",
       "[179] \"ITGA9_ITGB1_VCAM1\"                    \n",
       "[180] \"CDH3_CDH3\"                            \n",
       "[181] \"GJA5_GJA5\"                            \n",
       "[182] \"GJA1_GJA5\"                            \n",
       "[183] \"GJC1_GJA5\"                            \n",
       "[184] \"GJA1_GJA1\"                            \n",
       "[185] \"GJC1_GJA1\"                            \n",
       "[186] \"GJC1_GJC1\"                            \n",
       "[187] \"PCDHGA1_PCDHGA1\"                      \n",
       "[188] \"SDC2_ADGRA2\"                          \n",
       "[189] \"CD55_ADGRE5\"                          \n",
       "[190] \"LPAR1_ADGRE5\"                         \n",
       "[191] \"ITGAV_ITGB1_ADGRE5\"                   \n",
       "[192] \"FLRT2_ADGRL2\"                         \n",
       "[193] \"TENM2_ADGRL2\"                         \n",
       "[194] \"TENM3_ADGRL2\"                         \n",
       "[195] \"TENM4_ADGRL2\"                         \n",
       "[196] \"FLRT2_ADGRL3\"                         \n",
       "[197] \"TENM2_ADGRL3\"                         \n",
       "[198] \"TENM3_ADGRL3\"                         \n",
       "[199] \"TENM4_ADGRL3\"                         \n",
       "[200] \"FLRT2_FLRT2\"                          \n",
       "[201] \"FLRT2_UNC5C\"                          \n",
       "[202] \"Q9HDB5_CLSTN1\"                        \n",
       "[203] \"Q9HDB5_CLSTN2\"                        \n",
       "[204] \"CADM1_NECTIN3\"                        \n",
       "[205] \"CNTN1_NOTCH2\"                         \n",
       "[206] \"NTN4_NTRK2\"                           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"fresh and slow\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'BMP6_ACVR1_BMPR2'</li><li>'BMP6_BMPR1A_BMPR2'</li><li>'BMP6_BMPR1B_BMPR2'</li><li>'VEGFC_VEGFR2'</li><li>'IGF1_IGF1R'</li><li>'NTF3_NTRK2'</li><li>'NTF3_NTRK3'</li><li>'SEMA3A_NRP1_PLXNA2'</li><li>'SEMA3A_NRP1_PLXNA4'</li><li>'ADM_CALCRL'</li><li>'LAMA3_ITGA1_ITGB1'</li><li>'LAMA3_ITGA3_ITGB1'</li><li>'LAMA3_ITGA6_ITGB1'</li><li>'LAMA3_ITGA7_ITGB1'</li><li>'LAMA3_ITGA9_ITGB1'</li><li>'LAMA3_CD44'</li><li>'THBS1_CD36'</li><li>'THBS2_CD36'</li><li>'LAMA3_DAG1'</li><li>'Cholesterol-Cholesterol-LIPA_RORA'</li><li>'EFNA1_EPHA4'</li><li>'EFNA5_EPHA4'</li><li>'EFNA5_EPHB2'</li><li>'EFNB1_EPHA4'</li><li>'EFNB1_EPHB4'</li><li>'EFNB2_EPHA4'</li><li>'EFNB2_EPHB2'</li><li>'F11R_F11R'</li><li>'JAM2_F11R'</li><li>'JAM3_F11R'</li><li>'DLK1_NOTCH1'</li><li>'JAG1_NOTCH1'</li><li>'THY1_ADGRE5'</li><li>'CNTN1_NOTCH1'</li><li>'NTN1_DCC'</li><li>'NTN1_DSCAM'</li><li>'NTN1_UNC5C'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'BMP6\\_ACVR1\\_BMPR2'\n",
       "\\item 'BMP6\\_BMPR1A\\_BMPR2'\n",
       "\\item 'BMP6\\_BMPR1B\\_BMPR2'\n",
       "\\item 'VEGFC\\_VEGFR2'\n",
       "\\item 'IGF1\\_IGF1R'\n",
       "\\item 'NTF3\\_NTRK2'\n",
       "\\item 'NTF3\\_NTRK3'\n",
       "\\item 'SEMA3A\\_NRP1\\_PLXNA2'\n",
       "\\item 'SEMA3A\\_NRP1\\_PLXNA4'\n",
       "\\item 'ADM\\_CALCRL'\n",
       "\\item 'LAMA3\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA3\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA3\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA3\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMA3\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA3\\_CD44'\n",
       "\\item 'THBS1\\_CD36'\n",
       "\\item 'THBS2\\_CD36'\n",
       "\\item 'LAMA3\\_DAG1'\n",
       "\\item 'Cholesterol-Cholesterol-LIPA\\_RORA'\n",
       "\\item 'EFNA1\\_EPHA4'\n",
       "\\item 'EFNA5\\_EPHA4'\n",
       "\\item 'EFNA5\\_EPHB2'\n",
       "\\item 'EFNB1\\_EPHA4'\n",
       "\\item 'EFNB1\\_EPHB4'\n",
       "\\item 'EFNB2\\_EPHA4'\n",
       "\\item 'EFNB2\\_EPHB2'\n",
       "\\item 'F11R\\_F11R'\n",
       "\\item 'JAM2\\_F11R'\n",
       "\\item 'JAM3\\_F11R'\n",
       "\\item 'DLK1\\_NOTCH1'\n",
       "\\item 'JAG1\\_NOTCH1'\n",
       "\\item 'THY1\\_ADGRE5'\n",
       "\\item 'CNTN1\\_NOTCH1'\n",
       "\\item 'NTN1\\_DCC'\n",
       "\\item 'NTN1\\_DSCAM'\n",
       "\\item 'NTN1\\_UNC5C'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'BMP6_ACVR1_BMPR2'\n",
       "2. 'BMP6_BMPR1A_BMPR2'\n",
       "3. 'BMP6_BMPR1B_BMPR2'\n",
       "4. 'VEGFC_VEGFR2'\n",
       "5. 'IGF1_IGF1R'\n",
       "6. 'NTF3_NTRK2'\n",
       "7. 'NTF3_NTRK3'\n",
       "8. 'SEMA3A_NRP1_PLXNA2'\n",
       "9. 'SEMA3A_NRP1_PLXNA4'\n",
       "10. 'ADM_CALCRL'\n",
       "11. 'LAMA3_ITGA1_ITGB1'\n",
       "12. 'LAMA3_ITGA3_ITGB1'\n",
       "13. 'LAMA3_ITGA6_ITGB1'\n",
       "14. 'LAMA3_ITGA7_ITGB1'\n",
       "15. 'LAMA3_ITGA9_ITGB1'\n",
       "16. 'LAMA3_CD44'\n",
       "17. 'THBS1_CD36'\n",
       "18. 'THBS2_CD36'\n",
       "19. 'LAMA3_DAG1'\n",
       "20. 'Cholesterol-Cholesterol-LIPA_RORA'\n",
       "21. 'EFNA1_EPHA4'\n",
       "22. 'EFNA5_EPHA4'\n",
       "23. 'EFNA5_EPHB2'\n",
       "24. 'EFNB1_EPHA4'\n",
       "25. 'EFNB1_EPHB4'\n",
       "26. 'EFNB2_EPHA4'\n",
       "27. 'EFNB2_EPHB2'\n",
       "28. 'F11R_F11R'\n",
       "29. 'JAM2_F11R'\n",
       "30. 'JAM3_F11R'\n",
       "31. 'DLK1_NOTCH1'\n",
       "32. 'JAG1_NOTCH1'\n",
       "33. 'THY1_ADGRE5'\n",
       "34. 'CNTN1_NOTCH1'\n",
       "35. 'NTN1_DCC'\n",
       "36. 'NTN1_DSCAM'\n",
       "37. 'NTN1_UNC5C'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"BMP6_ACVR1_BMPR2\"                  \"BMP6_BMPR1A_BMPR2\"                \n",
       " [3] \"BMP6_BMPR1B_BMPR2\"                 \"VEGFC_VEGFR2\"                     \n",
       " [5] \"IGF1_IGF1R\"                        \"NTF3_NTRK2\"                       \n",
       " [7] \"NTF3_NTRK3\"                        \"SEMA3A_NRP1_PLXNA2\"               \n",
       " [9] \"SEMA3A_NRP1_PLXNA4\"                \"ADM_CALCRL\"                       \n",
       "[11] \"LAMA3_ITGA1_ITGB1\"                 \"LAMA3_ITGA3_ITGB1\"                \n",
       "[13] \"LAMA3_ITGA6_ITGB1\"                 \"LAMA3_ITGA7_ITGB1\"                \n",
       "[15] \"LAMA3_ITGA9_ITGB1\"                 \"LAMA3_CD44\"                       \n",
       "[17] \"THBS1_CD36\"                        \"THBS2_CD36\"                       \n",
       "[19] \"LAMA3_DAG1\"                        \"Cholesterol-Cholesterol-LIPA_RORA\"\n",
       "[21] \"EFNA1_EPHA4\"                       \"EFNA5_EPHA4\"                      \n",
       "[23] \"EFNA5_EPHB2\"                       \"EFNB1_EPHA4\"                      \n",
       "[25] \"EFNB1_EPHB4\"                       \"EFNB2_EPHA4\"                      \n",
       "[27] \"EFNB2_EPHB2\"                       \"F11R_F11R\"                        \n",
       "[29] \"JAM2_F11R\"                         \"JAM3_F11R\"                        \n",
       "[31] \"DLK1_NOTCH1\"                       \"JAG1_NOTCH1\"                      \n",
       "[33] \"THY1_ADGRE5\"                       \"CNTN1_NOTCH1\"                     \n",
       "[35] \"NTN1_DCC\"                          \"NTN1_DSCAM\"                       \n",
       "[37] \"NTN1_UNC5C\"                       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"fresh and vitri\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CXCL8_ACKR1'</li><li>'COL4A5_ITGA1_ITGB1'</li><li>'FN1_ITGA8_ITGB1'</li><li>'COL4A5_ITGA3_ITGB1'</li><li>'TNXB_ITGA8_ITGB1'</li><li>'COL4A5_ITGA9_ITGB1'</li><li>'FN1_ITGAV_ITGB8'</li><li>'LAMA5_ITGAV_ITGB8'</li><li>'LAMB1_ITGAV_ITGB8'</li><li>'LAMB2_ITGAV_ITGB8'</li><li>'LAMC1_ITGAV_ITGB8'</li><li>'COL1A1_ITGAV_ITGB8'</li><li>'COL1A2_ITGAV_ITGB8'</li><li>'COL4A1_ITGAV_ITGB8'</li><li>'COL4A2_ITGAV_ITGB8'</li><li>'COL4A5_ITGAV_ITGB8'</li><li>'COL6A1_ITGAV_ITGB8'</li><li>'COL6A2_ITGAV_ITGB8'</li><li>'COL6A3_ITGAV_ITGB8'</li><li>'COL4A5_CD44'</li><li>'COL4A5_SDC4'</li><li>'COL4A1_ADGRG6'</li><li>'COL4A2_ADGRG6'</li><li>'COL4A5_ADGRG6'</li><li>'LAMA2_ADGRG6'</li><li>'PRNP_ADGRG6'</li><li>'NRXN3_LRRTM3'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CXCL8\\_ACKR1'\n",
       "\\item 'COL4A5\\_ITGA1\\_ITGB1'\n",
       "\\item 'FN1\\_ITGA8\\_ITGB1'\n",
       "\\item 'COL4A5\\_ITGA3\\_ITGB1'\n",
       "\\item 'TNXB\\_ITGA8\\_ITGB1'\n",
       "\\item 'COL4A5\\_ITGA9\\_ITGB1'\n",
       "\\item 'FN1\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMA5\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMB1\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMB2\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMC1\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL1A1\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL1A2\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL4A1\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL4A2\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL4A5\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A1\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A2\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A3\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL4A5\\_CD44'\n",
       "\\item 'COL4A5\\_SDC4'\n",
       "\\item 'COL4A1\\_ADGRG6'\n",
       "\\item 'COL4A2\\_ADGRG6'\n",
       "\\item 'COL4A5\\_ADGRG6'\n",
       "\\item 'LAMA2\\_ADGRG6'\n",
       "\\item 'PRNP\\_ADGRG6'\n",
       "\\item 'NRXN3\\_LRRTM3'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CXCL8_ACKR1'\n",
       "2. 'COL4A5_ITGA1_ITGB1'\n",
       "3. 'FN1_ITGA8_ITGB1'\n",
       "4. 'COL4A5_ITGA3_ITGB1'\n",
       "5. 'TNXB_ITGA8_ITGB1'\n",
       "6. 'COL4A5_ITGA9_ITGB1'\n",
       "7. 'FN1_ITGAV_ITGB8'\n",
       "8. 'LAMA5_ITGAV_ITGB8'\n",
       "9. 'LAMB1_ITGAV_ITGB8'\n",
       "10. 'LAMB2_ITGAV_ITGB8'\n",
       "11. 'LAMC1_ITGAV_ITGB8'\n",
       "12. 'COL1A1_ITGAV_ITGB8'\n",
       "13. 'COL1A2_ITGAV_ITGB8'\n",
       "14. 'COL4A1_ITGAV_ITGB8'\n",
       "15. 'COL4A2_ITGAV_ITGB8'\n",
       "16. 'COL4A5_ITGAV_ITGB8'\n",
       "17. 'COL6A1_ITGAV_ITGB8'\n",
       "18. 'COL6A2_ITGAV_ITGB8'\n",
       "19. 'COL6A3_ITGAV_ITGB8'\n",
       "20. 'COL4A5_CD44'\n",
       "21. 'COL4A5_SDC4'\n",
       "22. 'COL4A1_ADGRG6'\n",
       "23. 'COL4A2_ADGRG6'\n",
       "24. 'COL4A5_ADGRG6'\n",
       "25. 'LAMA2_ADGRG6'\n",
       "26. 'PRNP_ADGRG6'\n",
       "27. 'NRXN3_LRRTM3'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"CXCL8_ACKR1\"        \"COL4A5_ITGA1_ITGB1\" \"FN1_ITGA8_ITGB1\"   \n",
       " [4] \"COL4A5_ITGA3_ITGB1\" \"TNXB_ITGA8_ITGB1\"   \"COL4A5_ITGA9_ITGB1\"\n",
       " [7] \"FN1_ITGAV_ITGB8\"    \"LAMA5_ITGAV_ITGB8\"  \"LAMB1_ITGAV_ITGB8\" \n",
       "[10] \"LAMB2_ITGAV_ITGB8\"  \"LAMC1_ITGAV_ITGB8\"  \"COL1A1_ITGAV_ITGB8\"\n",
       "[13] \"COL1A2_ITGAV_ITGB8\" \"COL4A1_ITGAV_ITGB8\" \"COL4A2_ITGAV_ITGB8\"\n",
       "[16] \"COL4A5_ITGAV_ITGB8\" \"COL6A1_ITGAV_ITGB8\" \"COL6A2_ITGAV_ITGB8\"\n",
       "[19] \"COL6A3_ITGAV_ITGB8\" \"COL4A5_CD44\"        \"COL4A5_SDC4\"       \n",
       "[22] \"COL4A1_ADGRG6\"      \"COL4A2_ADGRG6\"      \"COL4A5_ADGRG6\"     \n",
       "[25] \"LAMA2_ADGRG6\"       \"PRNP_ADGRG6\"        \"NRXN3_LRRTM3\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"slow and vitri\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'WNT2B_FZD1_LRP6'</li><li>'WNT2B_FZD8_LRP6'</li><li>'PDGFA_PDGFRA'</li><li>'PDGFA_PDGFRB'</li><li>'SPP1_CD44'</li><li>'SPP1_ITGAV_ITGB1'</li><li>'SPP1_ITGAV_ITGB5'</li><li>'SPP1_ITGA9_ITGB1'</li><li>'SPP1_ITGA5_ITGB1'</li><li>'PTN_SDC3'</li><li>'CD34_SELP'</li><li>'NECTIN3_NECTIN1'</li><li>'NECTIN3_PVR'</li><li>'NRXN1_NLGN1'</li><li>'NRXN1_NLGN2'</li><li>'NRXN2_NLGN2'</li><li>'SELE_CD44'</li><li>'SELE_GLG1'</li><li>'CDH12_ITGA1_ITGB1'</li><li>'COL3A1_ADGRG1'</li><li>'TGM2_ADGRG1'</li><li>'NRXN1_CLSTN1'</li><li>'NRXN1_CLSTN2'</li><li>'NRXN1_DAG1'</li><li>'NRXN2_DAG1'</li><li>'LRRC4C_PTPRF'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'WNT2B\\_FZD1\\_LRP6'\n",
       "\\item 'WNT2B\\_FZD8\\_LRP6'\n",
       "\\item 'PDGFA\\_PDGFRA'\n",
       "\\item 'PDGFA\\_PDGFRB'\n",
       "\\item 'SPP1\\_CD44'\n",
       "\\item 'SPP1\\_ITGAV\\_ITGB1'\n",
       "\\item 'SPP1\\_ITGAV\\_ITGB5'\n",
       "\\item 'SPP1\\_ITGA9\\_ITGB1'\n",
       "\\item 'SPP1\\_ITGA5\\_ITGB1'\n",
       "\\item 'PTN\\_SDC3'\n",
       "\\item 'CD34\\_SELP'\n",
       "\\item 'NECTIN3\\_NECTIN1'\n",
       "\\item 'NECTIN3\\_PVR'\n",
       "\\item 'NRXN1\\_NLGN1'\n",
       "\\item 'NRXN1\\_NLGN2'\n",
       "\\item 'NRXN2\\_NLGN2'\n",
       "\\item 'SELE\\_CD44'\n",
       "\\item 'SELE\\_GLG1'\n",
       "\\item 'CDH12\\_ITGA1\\_ITGB1'\n",
       "\\item 'COL3A1\\_ADGRG1'\n",
       "\\item 'TGM2\\_ADGRG1'\n",
       "\\item 'NRXN1\\_CLSTN1'\n",
       "\\item 'NRXN1\\_CLSTN2'\n",
       "\\item 'NRXN1\\_DAG1'\n",
       "\\item 'NRXN2\\_DAG1'\n",
       "\\item 'LRRC4C\\_PTPRF'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'WNT2B_FZD1_LRP6'\n",
       "2. 'WNT2B_FZD8_LRP6'\n",
       "3. 'PDGFA_PDGFRA'\n",
       "4. 'PDGFA_PDGFRB'\n",
       "5. 'SPP1_CD44'\n",
       "6. 'SPP1_ITGAV_ITGB1'\n",
       "7. 'SPP1_ITGAV_ITGB5'\n",
       "8. 'SPP1_ITGA9_ITGB1'\n",
       "9. 'SPP1_ITGA5_ITGB1'\n",
       "10. 'PTN_SDC3'\n",
       "11. 'CD34_SELP'\n",
       "12. 'NECTIN3_NECTIN1'\n",
       "13. 'NECTIN3_PVR'\n",
       "14. 'NRXN1_NLGN1'\n",
       "15. 'NRXN1_NLGN2'\n",
       "16. 'NRXN2_NLGN2'\n",
       "17. 'SELE_CD44'\n",
       "18. 'SELE_GLG1'\n",
       "19. 'CDH12_ITGA1_ITGB1'\n",
       "20. 'COL3A1_ADGRG1'\n",
       "21. 'TGM2_ADGRG1'\n",
       "22. 'NRXN1_CLSTN1'\n",
       "23. 'NRXN1_CLSTN2'\n",
       "24. 'NRXN1_DAG1'\n",
       "25. 'NRXN2_DAG1'\n",
       "26. 'LRRC4C_PTPRF'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"WNT2B_FZD1_LRP6\"   \"WNT2B_FZD8_LRP6\"   \"PDGFA_PDGFRA\"     \n",
       " [4] \"PDGFA_PDGFRB\"      \"SPP1_CD44\"         \"SPP1_ITGAV_ITGB1\" \n",
       " [7] \"SPP1_ITGAV_ITGB5\"  \"SPP1_ITGA9_ITGB1\"  \"SPP1_ITGA5_ITGB1\" \n",
       "[10] \"PTN_SDC3\"          \"CD34_SELP\"         \"NECTIN3_NECTIN1\"  \n",
       "[13] \"NECTIN3_PVR\"       \"NRXN1_NLGN1\"       \"NRXN1_NLGN2\"      \n",
       "[16] \"NRXN2_NLGN2\"       \"SELE_CD44\"         \"SELE_GLG1\"        \n",
       "[19] \"CDH12_ITGA1_ITGB1\" \"COL3A1_ADGRG1\"     \"TGM2_ADGRG1\"      \n",
       "[22] \"NRXN1_CLSTN1\"      \"NRXN1_CLSTN2\"      \"NRXN1_DAG1\"       \n",
       "[25] \"NRXN2_DAG1\"        \"LRRC4C_PTPRF\"     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"pathways unique to fresh \"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'WNT4_FZD1_LRP5'</li><li>'WNT4_FZD3_LRP5'</li><li>'WNT4_FZD1_LRP6'</li><li>'WNT4_FZD3_LRP6'</li><li>'WNT4_FZD7_LRP6'</li><li>'WNT4_FZD8_LRP6'</li><li>'WNT5B_FZD1'</li><li>'WNT5B_FZD3'</li><li>'WNT5B_FZD7'</li><li>'WNT5B_FZD8'</li><li>'AREG_EGFR_ERBB2'</li><li>'FGF2_FGFR2'</li><li>'FGF7_FGFR2'</li><li>'PDGFB_PDGFRA'</li><li>'PDGFB_PDGFRB'</li><li>'CCL5_ACKR1'</li><li>'CXCL12_CXCR4'</li><li>'MIF_CD74_CXCR4'</li><li>'ANGPTL1_ITGA1_ITGB1'</li><li>'LAMA1_ITGA1_ITGB1'</li><li>'LAMA1_ITGA3_ITGB1'</li><li>'RELN_ITGA3_ITGB1'</li><li>'LAMA1_ITGA6_ITGB1'</li><li>'LAMA1_ITGA7_ITGB1'</li><li>'LAMA1_ITGA9_ITGB1'</li><li>'LAMA1_ITGAV_ITGB8'</li><li>'LAMA3_ITGAV_ITGB8'</li><li>'LAMA1_CD44'</li><li>'LAMA1_DAG1'</li><li>'RELN_VLDLR'</li><li>'27Hydroxycholesterol-27HC-CYP11A1_NR1H4'</li><li>'ANDROSTERONE-HSD17B6_NR1H4'</li><li>'CDH2_CDH2'</li><li>'EFNB1_EPHB1'</li><li>'EFNB1_EPHB2'</li><li>'EFNB2_EPHB1'</li><li>'MAG_MAG'</li><li>'DLK1_NOTCH4'</li><li>'DLL4_NOTCH1'</li><li>'DLL4_NOTCH2'</li><li>'DLL4_NOTCH3'</li><li>'DLL4_NOTCH4'</li><li>'JAG1_NOTCH4'</li><li>'SEMA7A_ITGB1_ITGA1'</li><li>'SEMA7A_PLXNC1'</li><li>'VSIR_IGSF11'</li><li>'APP_SORL1'</li><li>'CD99_PILRB'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'WNT4\\_FZD1\\_LRP5'\n",
       "\\item 'WNT4\\_FZD3\\_LRP5'\n",
       "\\item 'WNT4\\_FZD1\\_LRP6'\n",
       "\\item 'WNT4\\_FZD3\\_LRP6'\n",
       "\\item 'WNT4\\_FZD7\\_LRP6'\n",
       "\\item 'WNT4\\_FZD8\\_LRP6'\n",
       "\\item 'WNT5B\\_FZD1'\n",
       "\\item 'WNT5B\\_FZD3'\n",
       "\\item 'WNT5B\\_FZD7'\n",
       "\\item 'WNT5B\\_FZD8'\n",
       "\\item 'AREG\\_EGFR\\_ERBB2'\n",
       "\\item 'FGF2\\_FGFR2'\n",
       "\\item 'FGF7\\_FGFR2'\n",
       "\\item 'PDGFB\\_PDGFRA'\n",
       "\\item 'PDGFB\\_PDGFRB'\n",
       "\\item 'CCL5\\_ACKR1'\n",
       "\\item 'CXCL12\\_CXCR4'\n",
       "\\item 'MIF\\_CD74\\_CXCR4'\n",
       "\\item 'ANGPTL1\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA1\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA3\\_ITGB1'\n",
       "\\item 'RELN\\_ITGA3\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA6\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA7\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMA1\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMA3\\_ITGAV\\_ITGB8'\n",
       "\\item 'LAMA1\\_CD44'\n",
       "\\item 'LAMA1\\_DAG1'\n",
       "\\item 'RELN\\_VLDLR'\n",
       "\\item '27Hydroxycholesterol-27HC-CYP11A1\\_NR1H4'\n",
       "\\item 'ANDROSTERONE-HSD17B6\\_NR1H4'\n",
       "\\item 'CDH2\\_CDH2'\n",
       "\\item 'EFNB1\\_EPHB1'\n",
       "\\item 'EFNB1\\_EPHB2'\n",
       "\\item 'EFNB2\\_EPHB1'\n",
       "\\item 'MAG\\_MAG'\n",
       "\\item 'DLK1\\_NOTCH4'\n",
       "\\item 'DLL4\\_NOTCH1'\n",
       "\\item 'DLL4\\_NOTCH2'\n",
       "\\item 'DLL4\\_NOTCH3'\n",
       "\\item 'DLL4\\_NOTCH4'\n",
       "\\item 'JAG1\\_NOTCH4'\n",
       "\\item 'SEMA7A\\_ITGB1\\_ITGA1'\n",
       "\\item 'SEMA7A\\_PLXNC1'\n",
       "\\item 'VSIR\\_IGSF11'\n",
       "\\item 'APP\\_SORL1'\n",
       "\\item 'CD99\\_PILRB'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'WNT4_FZD1_LRP5'\n",
       "2. 'WNT4_FZD3_LRP5'\n",
       "3. 'WNT4_FZD1_LRP6'\n",
       "4. 'WNT4_FZD3_LRP6'\n",
       "5. 'WNT4_FZD7_LRP6'\n",
       "6. 'WNT4_FZD8_LRP6'\n",
       "7. 'WNT5B_FZD1'\n",
       "8. 'WNT5B_FZD3'\n",
       "9. 'WNT5B_FZD7'\n",
       "10. 'WNT5B_FZD8'\n",
       "11. 'AREG_EGFR_ERBB2'\n",
       "12. 'FGF2_FGFR2'\n",
       "13. 'FGF7_FGFR2'\n",
       "14. 'PDGFB_PDGFRA'\n",
       "15. 'PDGFB_PDGFRB'\n",
       "16. 'CCL5_ACKR1'\n",
       "17. 'CXCL12_CXCR4'\n",
       "18. 'MIF_CD74_CXCR4'\n",
       "19. 'ANGPTL1_ITGA1_ITGB1'\n",
       "20. 'LAMA1_ITGA1_ITGB1'\n",
       "21. 'LAMA1_ITGA3_ITGB1'\n",
       "22. 'RELN_ITGA3_ITGB1'\n",
       "23. 'LAMA1_ITGA6_ITGB1'\n",
       "24. 'LAMA1_ITGA7_ITGB1'\n",
       "25. 'LAMA1_ITGA9_ITGB1'\n",
       "26. 'LAMA1_ITGAV_ITGB8'\n",
       "27. 'LAMA3_ITGAV_ITGB8'\n",
       "28. 'LAMA1_CD44'\n",
       "29. 'LAMA1_DAG1'\n",
       "30. 'RELN_VLDLR'\n",
       "31. '27Hydroxycholesterol-27HC-CYP11A1_NR1H4'\n",
       "32. 'ANDROSTERONE-HSD17B6_NR1H4'\n",
       "33. 'CDH2_CDH2'\n",
       "34. 'EFNB1_EPHB1'\n",
       "35. 'EFNB1_EPHB2'\n",
       "36. 'EFNB2_EPHB1'\n",
       "37. 'MAG_MAG'\n",
       "38. 'DLK1_NOTCH4'\n",
       "39. 'DLL4_NOTCH1'\n",
       "40. 'DLL4_NOTCH2'\n",
       "41. 'DLL4_NOTCH3'\n",
       "42. 'DLL4_NOTCH4'\n",
       "43. 'JAG1_NOTCH4'\n",
       "44. 'SEMA7A_ITGB1_ITGA1'\n",
       "45. 'SEMA7A_PLXNC1'\n",
       "46. 'VSIR_IGSF11'\n",
       "47. 'APP_SORL1'\n",
       "48. 'CD99_PILRB'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"WNT4_FZD1_LRP5\"                         \n",
       " [2] \"WNT4_FZD3_LRP5\"                         \n",
       " [3] \"WNT4_FZD1_LRP6\"                         \n",
       " [4] \"WNT4_FZD3_LRP6\"                         \n",
       " [5] \"WNT4_FZD7_LRP6\"                         \n",
       " [6] \"WNT4_FZD8_LRP6\"                         \n",
       " [7] \"WNT5B_FZD1\"                             \n",
       " [8] \"WNT5B_FZD3\"                             \n",
       " [9] \"WNT5B_FZD7\"                             \n",
       "[10] \"WNT5B_FZD8\"                             \n",
       "[11] \"AREG_EGFR_ERBB2\"                        \n",
       "[12] \"FGF2_FGFR2\"                             \n",
       "[13] \"FGF7_FGFR2\"                             \n",
       "[14] \"PDGFB_PDGFRA\"                           \n",
       "[15] \"PDGFB_PDGFRB\"                           \n",
       "[16] \"CCL5_ACKR1\"                             \n",
       "[17] \"CXCL12_CXCR4\"                           \n",
       "[18] \"MIF_CD74_CXCR4\"                         \n",
       "[19] \"ANGPTL1_ITGA1_ITGB1\"                    \n",
       "[20] \"LAMA1_ITGA1_ITGB1\"                      \n",
       "[21] \"LAMA1_ITGA3_ITGB1\"                      \n",
       "[22] \"RELN_ITGA3_ITGB1\"                       \n",
       "[23] \"LAMA1_ITGA6_ITGB1\"                      \n",
       "[24] \"LAMA1_ITGA7_ITGB1\"                      \n",
       "[25] \"LAMA1_ITGA9_ITGB1\"                      \n",
       "[26] \"LAMA1_ITGAV_ITGB8\"                      \n",
       "[27] \"LAMA3_ITGAV_ITGB8\"                      \n",
       "[28] \"LAMA1_CD44\"                             \n",
       "[29] \"LAMA1_DAG1\"                             \n",
       "[30] \"RELN_VLDLR\"                             \n",
       "[31] \"27Hydroxycholesterol-27HC-CYP11A1_NR1H4\"\n",
       "[32] \"ANDROSTERONE-HSD17B6_NR1H4\"             \n",
       "[33] \"CDH2_CDH2\"                              \n",
       "[34] \"EFNB1_EPHB1\"                            \n",
       "[35] \"EFNB1_EPHB2\"                            \n",
       "[36] \"EFNB2_EPHB1\"                            \n",
       "[37] \"MAG_MAG\"                                \n",
       "[38] \"DLK1_NOTCH4\"                            \n",
       "[39] \"DLL4_NOTCH1\"                            \n",
       "[40] \"DLL4_NOTCH2\"                            \n",
       "[41] \"DLL4_NOTCH3\"                            \n",
       "[42] \"DLL4_NOTCH4\"                            \n",
       "[43] \"JAG1_NOTCH4\"                            \n",
       "[44] \"SEMA7A_ITGB1_ITGA1\"                     \n",
       "[45] \"SEMA7A_PLXNC1\"                          \n",
       "[46] \"VSIR_IGSF11\"                            \n",
       "[47] \"APP_SORL1\"                              \n",
       "[48] \"CD99_PILRB\"                             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"pathways unique to slow \"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'BMP6_ACVR1_ACVR2A'</li><li>'WNT2B_FZD1_LRP5'</li><li>'WNT2B_FZD6_LRP5'</li><li>'WNT2B_FZD7_LRP5'</li><li>'WNT2B_FZD8_LRP5'</li><li>'WNT8A_FZD6_LRP5'</li><li>'WNT2B_FZD6_LRP6'</li><li>'WNT8A_FZD6_LRP6'</li><li>'WNT5B_FZD6'</li><li>'NRG3_ERBB4'</li><li>'FGF2_FGFR4'</li><li>'VEGFA_VEGFR1R2'</li><li>'CCL8_ACKR1'</li><li>'CXCL12_ACKR3'</li><li>'MIF_ACKR3'</li><li>'CTF1_LIFR_IL6ST'</li><li>'ANGPTL2_TLR4'</li><li>'SEMA3C_NRP1_PLXNA1'</li><li>'SEMA3C_NRP2_PLXNA1'</li><li>'SEMA3F_NRP2_PLXNA1'</li><li>'SEMA3F_NRP2_PLXNA4'</li><li>'GAS6_MERTK'</li><li>'SLIT2_ROBO1'</li><li>'SLIT2_ROBO2'</li><li>'PROS1_MERTK'</li><li>'THBS4_ITGA3_ITGB1'</li><li>'COL6A6_ITGA9_ITGB1'</li><li>'LAMC2_ITGA9_ITGB1'</li><li>'LAMC2_ITGAV_ITGB8'</li><li>'COL6A6_ITGAV_ITGB8'</li><li>'COL6A6_CD44'</li><li>'COL6A6_SDC4'</li><li>'THBS4_SDC4'</li><li>'LAMA2_SV2A'</li><li>'LAMA4_SV2A'</li><li>'LAMB2_SV2A'</li><li>'LAMC1_SV2A'</li><li>'LAMC2_SV2A'</li><li>'THBS4_CD36'</li><li>'THBS4_CD47'</li><li>'AGRN_DAG1'</li><li>'LAMC2_DAG1'</li><li>'PGF2a-AKR1C3_PTGFR'</li><li>'PGF2a-CBR1_PTGFR'</li><li>'RetinoicAcid-RA-ALDH1A1_RARA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A1_RARA_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A1_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A2_RARA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A2_RARA_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A2_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A3_RARA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A3_RARA_RXRA_CRABP2'</li><li>'RetinoicAcid-RA-ALDH1A3_RXRA_CRABP2'</li><li>'CNTN1_NRCAM'</li><li>'NCAM1_NCAM2'</li><li>'JAG2_NOTCH1'</li><li>'JAG2_NOTCH2'</li><li>'JAG2_NOTCH3'</li><li>'SEMA4D_PLXNB1'</li><li>'SEMA4D_PLXNB2'</li><li>'SEMA4G_PLXNB2'</li><li>'FLRT2_UNC5B'</li><li>'LRFN5_PTPRD'</li><li>'LRFN5_PTPRF'</li><li>'LRFN5_PTPRS'</li><li>'NRXN1_LRRTM4'</li><li>'NRXN3_LRRTM4'</li><li>'LRRTM4_PTPRS'</li><li>'NTN4_UNC5B'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'BMP6\\_ACVR1\\_ACVR2A'\n",
       "\\item 'WNT2B\\_FZD1\\_LRP5'\n",
       "\\item 'WNT2B\\_FZD6\\_LRP5'\n",
       "\\item 'WNT2B\\_FZD7\\_LRP5'\n",
       "\\item 'WNT2B\\_FZD8\\_LRP5'\n",
       "\\item 'WNT8A\\_FZD6\\_LRP5'\n",
       "\\item 'WNT2B\\_FZD6\\_LRP6'\n",
       "\\item 'WNT8A\\_FZD6\\_LRP6'\n",
       "\\item 'WNT5B\\_FZD6'\n",
       "\\item 'NRG3\\_ERBB4'\n",
       "\\item 'FGF2\\_FGFR4'\n",
       "\\item 'VEGFA\\_VEGFR1R2'\n",
       "\\item 'CCL8\\_ACKR1'\n",
       "\\item 'CXCL12\\_ACKR3'\n",
       "\\item 'MIF\\_ACKR3'\n",
       "\\item 'CTF1\\_LIFR\\_IL6ST'\n",
       "\\item 'ANGPTL2\\_TLR4'\n",
       "\\item 'SEMA3C\\_NRP1\\_PLXNA1'\n",
       "\\item 'SEMA3C\\_NRP2\\_PLXNA1'\n",
       "\\item 'SEMA3F\\_NRP2\\_PLXNA1'\n",
       "\\item 'SEMA3F\\_NRP2\\_PLXNA4'\n",
       "\\item 'GAS6\\_MERTK'\n",
       "\\item 'SLIT2\\_ROBO1'\n",
       "\\item 'SLIT2\\_ROBO2'\n",
       "\\item 'PROS1\\_MERTK'\n",
       "\\item 'THBS4\\_ITGA3\\_ITGB1'\n",
       "\\item 'COL6A6\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMC2\\_ITGA9\\_ITGB1'\n",
       "\\item 'LAMC2\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A6\\_ITGAV\\_ITGB8'\n",
       "\\item 'COL6A6\\_CD44'\n",
       "\\item 'COL6A6\\_SDC4'\n",
       "\\item 'THBS4\\_SDC4'\n",
       "\\item 'LAMA2\\_SV2A'\n",
       "\\item 'LAMA4\\_SV2A'\n",
       "\\item 'LAMB2\\_SV2A'\n",
       "\\item 'LAMC1\\_SV2A'\n",
       "\\item 'LAMC2\\_SV2A'\n",
       "\\item 'THBS4\\_CD36'\n",
       "\\item 'THBS4\\_CD47'\n",
       "\\item 'AGRN\\_DAG1'\n",
       "\\item 'LAMC2\\_DAG1'\n",
       "\\item 'PGF2a-AKR1C3\\_PTGFR'\n",
       "\\item 'PGF2a-CBR1\\_PTGFR'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A1\\_RARA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A1\\_RARA\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A1\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A2\\_RARA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A2\\_RARA\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A2\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A3\\_RARA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A3\\_RARA\\_RXRA\\_CRABP2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A3\\_RXRA\\_CRABP2'\n",
       "\\item 'CNTN1\\_NRCAM'\n",
       "\\item 'NCAM1\\_NCAM2'\n",
       "\\item 'JAG2\\_NOTCH1'\n",
       "\\item 'JAG2\\_NOTCH2'\n",
       "\\item 'JAG2\\_NOTCH3'\n",
       "\\item 'SEMA4D\\_PLXNB1'\n",
       "\\item 'SEMA4D\\_PLXNB2'\n",
       "\\item 'SEMA4G\\_PLXNB2'\n",
       "\\item 'FLRT2\\_UNC5B'\n",
       "\\item 'LRFN5\\_PTPRD'\n",
       "\\item 'LRFN5\\_PTPRF'\n",
       "\\item 'LRFN5\\_PTPRS'\n",
       "\\item 'NRXN1\\_LRRTM4'\n",
       "\\item 'NRXN3\\_LRRTM4'\n",
       "\\item 'LRRTM4\\_PTPRS'\n",
       "\\item 'NTN4\\_UNC5B'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'BMP6_ACVR1_ACVR2A'\n",
       "2. 'WNT2B_FZD1_LRP5'\n",
       "3. 'WNT2B_FZD6_LRP5'\n",
       "4. 'WNT2B_FZD7_LRP5'\n",
       "5. 'WNT2B_FZD8_LRP5'\n",
       "6. 'WNT8A_FZD6_LRP5'\n",
       "7. 'WNT2B_FZD6_LRP6'\n",
       "8. 'WNT8A_FZD6_LRP6'\n",
       "9. 'WNT5B_FZD6'\n",
       "10. 'NRG3_ERBB4'\n",
       "11. 'FGF2_FGFR4'\n",
       "12. 'VEGFA_VEGFR1R2'\n",
       "13. 'CCL8_ACKR1'\n",
       "14. 'CXCL12_ACKR3'\n",
       "15. 'MIF_ACKR3'\n",
       "16. 'CTF1_LIFR_IL6ST'\n",
       "17. 'ANGPTL2_TLR4'\n",
       "18. 'SEMA3C_NRP1_PLXNA1'\n",
       "19. 'SEMA3C_NRP2_PLXNA1'\n",
       "20. 'SEMA3F_NRP2_PLXNA1'\n",
       "21. 'SEMA3F_NRP2_PLXNA4'\n",
       "22. 'GAS6_MERTK'\n",
       "23. 'SLIT2_ROBO1'\n",
       "24. 'SLIT2_ROBO2'\n",
       "25. 'PROS1_MERTK'\n",
       "26. 'THBS4_ITGA3_ITGB1'\n",
       "27. 'COL6A6_ITGA9_ITGB1'\n",
       "28. 'LAMC2_ITGA9_ITGB1'\n",
       "29. 'LAMC2_ITGAV_ITGB8'\n",
       "30. 'COL6A6_ITGAV_ITGB8'\n",
       "31. 'COL6A6_CD44'\n",
       "32. 'COL6A6_SDC4'\n",
       "33. 'THBS4_SDC4'\n",
       "34. 'LAMA2_SV2A'\n",
       "35. 'LAMA4_SV2A'\n",
       "36. 'LAMB2_SV2A'\n",
       "37. 'LAMC1_SV2A'\n",
       "38. 'LAMC2_SV2A'\n",
       "39. 'THBS4_CD36'\n",
       "40. 'THBS4_CD47'\n",
       "41. 'AGRN_DAG1'\n",
       "42. 'LAMC2_DAG1'\n",
       "43. 'PGF2a-AKR1C3_PTGFR'\n",
       "44. 'PGF2a-CBR1_PTGFR'\n",
       "45. 'RetinoicAcid-RA-ALDH1A1_RARA_CRABP2'\n",
       "46. 'RetinoicAcid-RA-ALDH1A1_RARA_RXRA_CRABP2'\n",
       "47. 'RetinoicAcid-RA-ALDH1A1_RXRA_CRABP2'\n",
       "48. 'RetinoicAcid-RA-ALDH1A2_RARA_CRABP2'\n",
       "49. 'RetinoicAcid-RA-ALDH1A2_RARA_RXRA_CRABP2'\n",
       "50. 'RetinoicAcid-RA-ALDH1A2_RXRA_CRABP2'\n",
       "51. 'RetinoicAcid-RA-ALDH1A3_RARA_CRABP2'\n",
       "52. 'RetinoicAcid-RA-ALDH1A3_RARA_RXRA_CRABP2'\n",
       "53. 'RetinoicAcid-RA-ALDH1A3_RXRA_CRABP2'\n",
       "54. 'CNTN1_NRCAM'\n",
       "55. 'NCAM1_NCAM2'\n",
       "56. 'JAG2_NOTCH1'\n",
       "57. 'JAG2_NOTCH2'\n",
       "58. 'JAG2_NOTCH3'\n",
       "59. 'SEMA4D_PLXNB1'\n",
       "60. 'SEMA4D_PLXNB2'\n",
       "61. 'SEMA4G_PLXNB2'\n",
       "62. 'FLRT2_UNC5B'\n",
       "63. 'LRFN5_PTPRD'\n",
       "64. 'LRFN5_PTPRF'\n",
       "65. 'LRFN5_PTPRS'\n",
       "66. 'NRXN1_LRRTM4'\n",
       "67. 'NRXN3_LRRTM4'\n",
       "68. 'LRRTM4_PTPRS'\n",
       "69. 'NTN4_UNC5B'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"BMP6_ACVR1_ACVR2A\"                       \n",
       " [2] \"WNT2B_FZD1_LRP5\"                         \n",
       " [3] \"WNT2B_FZD6_LRP5\"                         \n",
       " [4] \"WNT2B_FZD7_LRP5\"                         \n",
       " [5] \"WNT2B_FZD8_LRP5\"                         \n",
       " [6] \"WNT8A_FZD6_LRP5\"                         \n",
       " [7] \"WNT2B_FZD6_LRP6\"                         \n",
       " [8] \"WNT8A_FZD6_LRP6\"                         \n",
       " [9] \"WNT5B_FZD6\"                              \n",
       "[10] \"NRG3_ERBB4\"                              \n",
       "[11] \"FGF2_FGFR4\"                              \n",
       "[12] \"VEGFA_VEGFR1R2\"                          \n",
       "[13] \"CCL8_ACKR1\"                              \n",
       "[14] \"CXCL12_ACKR3\"                            \n",
       "[15] \"MIF_ACKR3\"                               \n",
       "[16] \"CTF1_LIFR_IL6ST\"                         \n",
       "[17] \"ANGPTL2_TLR4\"                            \n",
       "[18] \"SEMA3C_NRP1_PLXNA1\"                      \n",
       "[19] \"SEMA3C_NRP2_PLXNA1\"                      \n",
       "[20] \"SEMA3F_NRP2_PLXNA1\"                      \n",
       "[21] \"SEMA3F_NRP2_PLXNA4\"                      \n",
       "[22] \"GAS6_MERTK\"                              \n",
       "[23] \"SLIT2_ROBO1\"                             \n",
       "[24] \"SLIT2_ROBO2\"                             \n",
       "[25] \"PROS1_MERTK\"                             \n",
       "[26] \"THBS4_ITGA3_ITGB1\"                       \n",
       "[27] \"COL6A6_ITGA9_ITGB1\"                      \n",
       "[28] \"LAMC2_ITGA9_ITGB1\"                       \n",
       "[29] \"LAMC2_ITGAV_ITGB8\"                       \n",
       "[30] \"COL6A6_ITGAV_ITGB8\"                      \n",
       "[31] \"COL6A6_CD44\"                             \n",
       "[32] \"COL6A6_SDC4\"                             \n",
       "[33] \"THBS4_SDC4\"                              \n",
       "[34] \"LAMA2_SV2A\"                              \n",
       "[35] \"LAMA4_SV2A\"                              \n",
       "[36] \"LAMB2_SV2A\"                              \n",
       "[37] \"LAMC1_SV2A\"                              \n",
       "[38] \"LAMC2_SV2A\"                              \n",
       "[39] \"THBS4_CD36\"                              \n",
       "[40] \"THBS4_CD47\"                              \n",
       "[41] \"AGRN_DAG1\"                               \n",
       "[42] \"LAMC2_DAG1\"                              \n",
       "[43] \"PGF2a-AKR1C3_PTGFR\"                      \n",
       "[44] \"PGF2a-CBR1_PTGFR\"                        \n",
       "[45] \"RetinoicAcid-RA-ALDH1A1_RARA_CRABP2\"     \n",
       "[46] \"RetinoicAcid-RA-ALDH1A1_RARA_RXRA_CRABP2\"\n",
       "[47] \"RetinoicAcid-RA-ALDH1A1_RXRA_CRABP2\"     \n",
       "[48] \"RetinoicAcid-RA-ALDH1A2_RARA_CRABP2\"     \n",
       "[49] \"RetinoicAcid-RA-ALDH1A2_RARA_RXRA_CRABP2\"\n",
       "[50] \"RetinoicAcid-RA-ALDH1A2_RXRA_CRABP2\"     \n",
       "[51] \"RetinoicAcid-RA-ALDH1A3_RARA_CRABP2\"     \n",
       "[52] \"RetinoicAcid-RA-ALDH1A3_RARA_RXRA_CRABP2\"\n",
       "[53] \"RetinoicAcid-RA-ALDH1A3_RXRA_CRABP2\"     \n",
       "[54] \"CNTN1_NRCAM\"                             \n",
       "[55] \"NCAM1_NCAM2\"                             \n",
       "[56] \"JAG2_NOTCH1\"                             \n",
       "[57] \"JAG2_NOTCH2\"                             \n",
       "[58] \"JAG2_NOTCH3\"                             \n",
       "[59] \"SEMA4D_PLXNB1\"                           \n",
       "[60] \"SEMA4D_PLXNB2\"                           \n",
       "[61] \"SEMA4G_PLXNB2\"                           \n",
       "[62] \"FLRT2_UNC5B\"                             \n",
       "[63] \"LRFN5_PTPRD\"                             \n",
       "[64] \"LRFN5_PTPRF\"                             \n",
       "[65] \"LRFN5_PTPRS\"                             \n",
       "[66] \"NRXN1_LRRTM4\"                            \n",
       "[67] \"NRXN3_LRRTM4\"                            \n",
       "[68] \"LRRTM4_PTPRS\"                            \n",
       "[69] \"NTN4_UNC5B\"                              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"pathways unique to vitri\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TGFB3_TGFBR1_TGFBR2'</li><li>'TGFB3_ACVR1_TGFBR1'</li><li>'SPP1_ITGA8_ITGB1'</li><li>'ANGPTL4_ITGA5_ITGB1'</li><li>'ANGPTL4_CDH11'</li><li>'ANGPTL4_SDC2'</li><li>'ANGPTL4_SDC4'</li><li>'EDN1_EDNRA'</li><li>'EDN1_EDNRB'</li><li>'PGE2-PTGES3_PTGER2'</li><li>'RetinoicAcid-RA-ALDH1A1_RORB'</li><li>'RetinoicAcid-RA-ALDH1A2_RORB'</li><li>'RetinoicAcid-RA-ALDH1A3_RORB'</li><li>'Dehydroepiandrosterone-DHEA-STS_ESR1'</li><li>'CDH1_CDH1'</li><li>'FLRT3_ADGRL2'</li><li>'FLRT3_ADGRL3'</li><li>'FLRT3_FLRT3'</li><li>'FLRT3_UNC5C'</li><li>'NRXN2_CLSTN1'</li><li>'NRXN2_CLSTN2'</li><li>'NRXN1_LRRTM3'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TGFB3\\_TGFBR1\\_TGFBR2'\n",
       "\\item 'TGFB3\\_ACVR1\\_TGFBR1'\n",
       "\\item 'SPP1\\_ITGA8\\_ITGB1'\n",
       "\\item 'ANGPTL4\\_ITGA5\\_ITGB1'\n",
       "\\item 'ANGPTL4\\_CDH11'\n",
       "\\item 'ANGPTL4\\_SDC2'\n",
       "\\item 'ANGPTL4\\_SDC4'\n",
       "\\item 'EDN1\\_EDNRA'\n",
       "\\item 'EDN1\\_EDNRB'\n",
       "\\item 'PGE2-PTGES3\\_PTGER2'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A1\\_RORB'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A2\\_RORB'\n",
       "\\item 'RetinoicAcid-RA-ALDH1A3\\_RORB'\n",
       "\\item 'Dehydroepiandrosterone-DHEA-STS\\_ESR1'\n",
       "\\item 'CDH1\\_CDH1'\n",
       "\\item 'FLRT3\\_ADGRL2'\n",
       "\\item 'FLRT3\\_ADGRL3'\n",
       "\\item 'FLRT3\\_FLRT3'\n",
       "\\item 'FLRT3\\_UNC5C'\n",
       "\\item 'NRXN2\\_CLSTN1'\n",
       "\\item 'NRXN2\\_CLSTN2'\n",
       "\\item 'NRXN1\\_LRRTM3'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TGFB3_TGFBR1_TGFBR2'\n",
       "2. 'TGFB3_ACVR1_TGFBR1'\n",
       "3. 'SPP1_ITGA8_ITGB1'\n",
       "4. 'ANGPTL4_ITGA5_ITGB1'\n",
       "5. 'ANGPTL4_CDH11'\n",
       "6. 'ANGPTL4_SDC2'\n",
       "7. 'ANGPTL4_SDC4'\n",
       "8. 'EDN1_EDNRA'\n",
       "9. 'EDN1_EDNRB'\n",
       "10. 'PGE2-PTGES3_PTGER2'\n",
       "11. 'RetinoicAcid-RA-ALDH1A1_RORB'\n",
       "12. 'RetinoicAcid-RA-ALDH1A2_RORB'\n",
       "13. 'RetinoicAcid-RA-ALDH1A3_RORB'\n",
       "14. 'Dehydroepiandrosterone-DHEA-STS_ESR1'\n",
       "15. 'CDH1_CDH1'\n",
       "16. 'FLRT3_ADGRL2'\n",
       "17. 'FLRT3_ADGRL3'\n",
       "18. 'FLRT3_FLRT3'\n",
       "19. 'FLRT3_UNC5C'\n",
       "20. 'NRXN2_CLSTN1'\n",
       "21. 'NRXN2_CLSTN2'\n",
       "22. 'NRXN1_LRRTM3'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"TGFB3_TGFBR1_TGFBR2\"                 \n",
       " [2] \"TGFB3_ACVR1_TGFBR1\"                  \n",
       " [3] \"SPP1_ITGA8_ITGB1\"                    \n",
       " [4] \"ANGPTL4_ITGA5_ITGB1\"                 \n",
       " [5] \"ANGPTL4_CDH11\"                       \n",
       " [6] \"ANGPTL4_SDC2\"                        \n",
       " [7] \"ANGPTL4_SDC4\"                        \n",
       " [8] \"EDN1_EDNRA\"                          \n",
       " [9] \"EDN1_EDNRB\"                          \n",
       "[10] \"PGE2-PTGES3_PTGER2\"                  \n",
       "[11] \"RetinoicAcid-RA-ALDH1A1_RORB\"        \n",
       "[12] \"RetinoicAcid-RA-ALDH1A2_RORB\"        \n",
       "[13] \"RetinoicAcid-RA-ALDH1A3_RORB\"        \n",
       "[14] \"Dehydroepiandrosterone-DHEA-STS_ESR1\"\n",
       "[15] \"CDH1_CDH1\"                           \n",
       "[16] \"FLRT3_ADGRL2\"                        \n",
       "[17] \"FLRT3_ADGRL3\"                        \n",
       "[18] \"FLRT3_FLRT3\"                         \n",
       "[19] \"FLRT3_UNC5C\"                         \n",
       "[20] \"NRXN2_CLSTN1\"                        \n",
       "[21] \"NRXN2_CLSTN2\"                        \n",
       "[22] \"NRXN1_LRRTM3\"                        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('union of all samples')\n",
    "union(union(unique(df$fresh$interaction_name), unique(df$slow$interaction_name)), unique(df$vitri$interaction_name))\n",
    "print('in all samples')\n",
    "all = intersect(intersect(unique(df$fresh$interaction_name), unique(df$slow$interaction_name)), unique(df$vitri$interaction_name))\n",
    "all\n",
    "print('fresh and slow')\n",
    "setdiff(intersect(unique(df$fresh$interaction_name), unique(df$slow$interaction_name)), all)\n",
    "print('fresh and vitri')\n",
    "setdiff(intersect(unique(df$fresh$interaction_name), unique(df$vitri$interaction_name)), all)\n",
    "print('slow and vitri')\n",
    "setdiff(intersect(unique(df$slow$interaction_name), unique(df$vitri$interaction_name)), all)\n",
    "\n",
    "\n",
    "# pathways unique to each slice\n",
    "print('pathways unique to fresh ')\n",
    "setdiff(unique(df$fresh$interaction_name), union(unique(df$slow$interaction_name), unique(df$vitri$interaction_name)))\n",
    "print('pathways unique to slow ')\n",
    "setdiff(unique(df$slow$interaction_name), union(unique(df$fresh$interaction_name), unique(df$vitri$interaction_name)))\n",
    "print('pathways unique to vitri')\n",
    "setdiff(unique(df$vitri$interaction_name), union(unique(df$fresh$interaction_name), unique(df$slow$interaction_name)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Form different comparison sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The CellChat object will be lifted up using the cell labels B_Plasma(IGKC), B_Plasma(IGLC2), B_Plasma(IGLL5), ENDO_SM_0, ENDO_SM_1, ENDO_SM_2, ENDO_SM_3, ENDO_SM_4, ENDO_SM_5, ENDO_SM_6, ERY_ENDO(HBB), FIB(COL1A1), O(PDCD5), ST_LIP(GSTA1), ST_LIP(INSL3), ST_SM(ADAMTS4), ST_SM(MT1A), ST(ARL15), ST(BEX2), ST(BNC2), ST(EXT1), ST(FNDC3B), ST(KRT19), ST(MAGI1), ST(MAML2), ST(PRKG1), ST(RAD51B), ST(RBFOX1), ST(RPL41), ST(SECTM1), ST(SLCO5A1), ST(ZFPM2)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Update slots object@net, object@netP, object@idents in a single dataset... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The CellChat object will be lifted up using the cell labels B_Plasma(IGKC), B_Plasma(IGLC2), B_Plasma(IGLL5), ENDO_SM_0, ENDO_SM_1, ENDO_SM_2, ENDO_SM_3, ENDO_SM_4, ENDO_SM_5, ENDO_SM_6, ERY_ENDO(HBB), FIB(COL1A1), O(PDCD5), ST_LIP(GSTA1), ST_LIP(INSL3), ST_SM(ADAMTS4), ST_SM(MT1A), ST(ARL15), ST(BEX2), ST(BNC2), ST(EXT1), ST(FNDC3B), ST(KRT19), ST(MAGI1), ST(MAML2), ST(PRKG1), ST(RAD51B), ST(RBFOX1), ST(RPL41), ST(SECTM1), ST(SLCO5A1), ST(ZFPM2)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Update slots object@net, object@netP, object@idents in a single dataset... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The CellChat object will be lifted up using the cell labels B_Plasma(IGKC), B_Plasma(IGLC2), B_Plasma(IGLL5), ENDO_SM_0, ENDO_SM_1, ENDO_SM_2, ENDO_SM_3, ENDO_SM_4, ENDO_SM_5, ENDO_SM_6, ERY_ENDO(HBB), FIB(COL1A1), O(PDCD5), ST_LIP(GSTA1), ST_LIP(INSL3), ST_SM(ADAMTS4), ST_SM(MT1A), ST(ARL15), ST(BEX2), ST(BNC2), ST(EXT1), ST(FNDC3B), ST(KRT19), ST(MAGI1), ST(MAML2), ST(PRKG1), ST(RAD51B), ST(RBFOX1), ST(RPL41), ST(SECTM1), ST(SLCO5A1), ST(ZFPM2)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Update slots object@net, object@netP, object@idents in a single dataset... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The CellChat object will be lifted up using the cell labels B_Plasma(IGKC), B_Plasma(IGLC2), B_Plasma(IGLL5), ENDO_SM_0, ENDO_SM_1, ENDO_SM_2, ENDO_SM_3, ENDO_SM_4, ENDO_SM_5, ENDO_SM_6, ERY_ENDO(HBB), FIB(COL1A1), O(PDCD5), ST_LIP(GSTA1), ST_LIP(INSL3), ST_SM(ADAMTS4), ST_SM(MT1A), ST(ARL15), ST(BEX2), ST(BNC2), ST(EXT1), ST(FNDC3B), ST(KRT19), ST(MAGI1), ST(MAML2), ST(PRKG1), ST(RAD51B), ST(RBFOX1), ST(RPL41), ST(SECTM1), ST(SLCO5A1), ST(ZFPM2)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Update slots object@net, object@netP, object@idents in a single dataset... \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Merge the following slots: 'data.signaling','images','net', 'netP','meta', 'idents', 'var.features' , 'DB', and 'LR'.\n",
      "\n",
      "Merge the following slots: 'data.signaling','images','net', 'netP','meta', 'idents', 'var.features' , 'DB', and 'LR'.\n",
      "\n",
      "Merge the following slots: 'data.signaling','images','net', 'netP','meta', 'idents', 'var.features' , 'DB', and 'LR'.\n",
      "\n",
      "Merge the following slots: 'data.signaling','images','net', 'netP','meta', 'idents', 'var.features' , 'DB', and 'LR'.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "group.new = union(levels(cellchat.slow@idents), levels(cellchat.vitri@idents))\n",
    "cellchat.sl <- liftCellChat(cellchat.slow, group.new)\n",
    "cellchat.vi <- liftCellChat(cellchat.vitri, group.new)\n",
    "\n",
    "frozen.list <- list(vitri= cellchat.vi, slow = cellchat.sl)\n",
    "frozen.pathways <- intersect(cellchat.vi@netP$pathways, cellchat.sl@netP$pathways)\n",
    "\n",
    "group.new = levels(cellchat.fresh@idents)\n",
    "cellchat.vitri<- liftCellChat(cellchat.vitri, group.new)\n",
    "cellchat.slow <- liftCellChat(cellchat.slow, group.new)\n",
    "\n",
    "freshslow.list <- list(fresh = cellchat.fresh, slow = cellchat.slow)\n",
    "freshslow.pathways <- intersect(cellchat.fresh@netP$pathways, cellchat.slow@netP$pathways)\n",
    "\n",
    "freshvitri.list <- list(fresh = cellchat.fresh, vitri= cellchat.vitri)\n",
    "freshvitri.pathways <- intersect(cellchat.fresh@netP$pathways, cellchat.vitri@netP$pathways)\n",
    "\n",
    "object.list <- list(fresh = cellchat.fresh, slow = cellchat.slow, vitri= cellchat.vitri)\n",
    "\n",
    "cellchat.frozen <- mergeCellChat(frozen.list, add.names = names(frozen.list))\n",
    "cellchat.freshvitri<- mergeCellChat(freshvitri.list, add.names = names(freshvitri.list))\n",
    "cellchat.freshslow <- mergeCellChat(freshslow.list, add.names = names(freshslow.list))\n",
    "cellchat <- mergeCellChat(object.list, add.names = names(object.list))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ec2506c",
   "metadata": {},
   "source": [
    "** Change here and run all below to generate comparison figures between chosen conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cch = cellchat.frozen\n",
    "pathways = frozen.pathways\n",
    "cch.list = frozen.list\n",
    "samples = paste0(names(cch.list), collapse = '.')\n",
    "fig.dir = paste0(base.fig.dir, '/', samples, '/')\n",
    "dir.create(fig.dir, showWarnings = FALSE)\n",
    "\n",
    "df.net <- subsetCommunication(cch)\n",
    "vert.rec = which(unique(cch@idents[[1]]) %in% celltype_OI)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79190f05",
   "metadata": {},
   "source": [
    "# Part I: Predict general principles of cell-cell communication"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f43c80f4",
   "metadata": {},
   "source": [
    "CellChat starts with the big picture to predict general principles of cell-cell communication. When comparing cell-cell communication among multiple biological conditions, it can answer the following biological questions:\n",
    "\n",
    "* Whether the cell-cell communication is enhanced or not\n",
    "\n",
    "* The interaction between which cell types is significantly changed\n",
    "\n",
    "* How the major sources and targets change from one condition to another"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2792ac6",
   "metadata": {},
   "source": [
    "## Compare the total number of interactions and interaction strength"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67572b29",
   "metadata": {},
   "source": [
    "To answer on question on whether the cell-cell communication is enhanced or not, CellChat compares the the total number of interactions and interaction strength of the inferred cell-cell communication networks from different biological conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setwd(fig.dir)\n",
    "pdf(paste0('ttl_interactions.pdf'), width = 10, height = 10)\n",
    "gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2,3))\n",
    "gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2,3), measure = \"weight\")\n",
    "gg1 + gg2\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f3b257f",
   "metadata": {},
   "source": [
    "## Compare the number of interactions and interaction strength among different cell populations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11a35f82",
   "metadata": {},
   "source": [
    "To identify the interaction between which cell populations showing significant changes, CellChat compares the number of interactions and interaction strength among different cell populations."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f72a4685",
   "metadata": {},
   "source": [
    "### Differential number of interactions or interaction strength among different cell populations\n",
    "The differential number of interactions or interaction strength in the cell-cell communication network between two datasets can be visualized using circle plot, where $\\color{red}{\\text{red}}$ (or $\\color{blue}{\\text{blue}}$) colored edges represent $\\color{red}{\\text{increased}}$ (or $\\color{blue}{\\text{decreased}}$) signaling in the second dataset compared to the first one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c44e3467",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Vitri  Slow \n",
       "46364 45796 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table(cch@meta$batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e615df8a",
   "metadata": {
    "Rmd_chunk_options": "fig.width=12,fig.height = 6, fig.wide = TRUE, fig.align = \"center\"",
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setwd(fig.dir)\n",
    "f = paste0(samples, '.diffInteraction_circle.pdf')\n",
    "pdf(f, width = 10, height = 10)\n",
    "par(mfrow = c(1,length(cch.list)))\n",
    "g1 = netVisual_diffInteraction(cch, weight.scale = T, vertex.label.cex=0.7)\n",
    "g2 = netVisual_diffInteraction(cch, weight.scale = T, measure = \"weight\", vertex.label.cex=0.7)\n",
    "dev.off()\n",
    "# ggsave(filename = f, width = 10, height = 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d847241",
   "metadata": {},
   "source": [
    "We can also show differential number of interactions or interaction strength in a greater details using a heatmap. The top colored bar plot represents the sum of column of values displayed in the heatmap (incoming signaling). The right colored bar plot represents the sum of row of values (outgoing signaling). In the colorbar, $\\color{red}{\\text{red}}$ (or $\\color{blue}{\\text{blue}}$) represents $\\color{red}{\\text{increased}}$ (or $\\color{blue}{\\text{decreased}}$) signaling in the second dataset compared to the first one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eb8dc65d",
   "metadata": {
    "Rmd_chunk_options": "fig.width=10,fig.height = 5, fig.wide = TRUE, fig.align = \"center\"",
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Do heatmap based on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "f = paste0(samples, '.diffInteraction_heatmap_cnts.pdf')\n",
    "pdf(f, width = 10, height = 10)\n",
    "netVisual_heatmap(cch)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Do heatmap based on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "f = paste0(samples, '.diffInteraction_heatmap_weights.pdf')\n",
    "pdf(f, width = 10, height = 10)\n",
    "netVisual_heatmap(cch, measure = \"weight\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52733653",
   "metadata": {},
   "source": [
    "## Compare the major sources and targets in 2D space"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e80dc5e5",
   "metadata": {},
   "source": [
    "Comparing the outgoing and incoming interaction strength in 2D space allows ready identification of the cell populations with significant changes in sending or receiving signals between different datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "62781d40",
   "metadata": {
    "Rmd_chunk_options": "fig.width=9,fig.height = 4, fig.wide = TRUE, fig.align = \"center\"",
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num.link <- sapply(frozen.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})\n",
    "weight.MinMax <- c(min(num.link), max(num.link)) # control the dot size in the different datasets\n",
    "\n",
    "pdf(paste0(samples, '.diffInteraction_scatter.pdf'), width = 10, height = 5)\n",
    "gg <- list()\n",
    "for (i in 1:length(cch.list)) {\n",
    "  gg[[i]] <- netAnalysis_signalingRole_scatter(cch.list[[i]], title = names(cch.list)[i], weight.MinMax = weight.MinMax, )\n",
    "}\n",
    "patchwork::wrap_plots(plots = gg)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "96a0dc9f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ## Identify signaling changes associated with one cell group\n",
    "# for (celltype in levels(cch@idents[[1]])) {\n",
    "#     f = paste0(samples, '.', celltype, '.diffInteraction_scatter.pdf')\n",
    "#     tryCatch({\n",
    "#         pdf(f, width = 10, height = 5)\n",
    "#         gg1 = netAnalysis_signalingChanges_scatter(cch, idents.use = celltype)\n",
    "#         gg2 = netAnalysis_signalingChanges_scatter(cch, idents.use = celltype)\n",
    "#         patchwork::wrap_plots(plots = c(gg1,gg2))\n",
    "#         dev.off()\n",
    "#         # ggsave(filename = f, plot=gg1+gg2, width = 10, height = 5)\n",
    "#     }, error = function(e) {\n",
    "#         print(e)\n",
    "#     })\n",
    "# }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "563c193a",
   "metadata": {},
   "source": [
    "# Part II: Identify altered signaling with distinct network architecture and interaction strength"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "526b7267",
   "metadata": {},
   "source": [
    "CellChat then can identify signaling networks with larger (or less) difference, signaling groups, and the conserved and context-specific signaling pathways based on their cell-cell communication networks among multiple biological conditions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2edc8134",
   "metadata": {},
   "source": [
    "## Identify signaling networks with larger (or less) difference as well as signaling groups based on their functional/structure similarity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eef4a6e",
   "metadata": {},
   "source": [
    "CellChat performs joint manifold learning and classification of the inferred communication networks based on their functional and topological similarity. NB: Such analysis is applicable to more than two datasets.\n",
    "\n",
    "**Functional similarity**: High degree of functional similarity indicates major senders and receivers are similar, and it can be interpreted as the two signaling pathways or two ligand-receptor pairs exhibit similar and/or redundant roles. **NB**: Functional similarity analysis is not applicable to multiple datsets with different cell type composition.\n",
    "\n",
    "**Structural similarity**: A structural similarity was used to compare their signaling network structure, without considering the similarity of senders and receivers. **NB**: Structural similarity analysis is applicable to multiple datsets with the same cell type composition or the vastly different cell type composition.\n",
    "\n",
    "Here we can run the manifold and classification learning analysis based on the functional similarity because the two datasets have the the same cell type composition."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "832d1a33",
   "metadata": {},
   "source": [
    "## Identify and visualize the conserved and context-specific signaling pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f02a008",
   "metadata": {},
   "source": [
    "By comparing the information flow/interaction strengh of each signaling pathway, we can identify signaling pathways, (i) turn off, (ii) decrease, (iii) turn on or (iv) increase, by change their information flow at one condition as compared to another condition."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7128d5c6",
   "metadata": {},
   "source": [
    "### Compare the overall information flow of each signaling pathway\n",
    "We can identify the conserved and context-specific signaling pathways by simply comparing the information flow for each signaling pathway, which is defined by the sum of communication probability among all pairs of cell groups in the inferred network (i.e., the total weights in the network).\n",
    "\n",
    "This bar graph can be plotted in a stacked mode or not. Significant signaling pathways were ranked based on differences in the overall information flow within the inferred networks between slice1 and slice2 skin. The top signaling pathways colored red are enriched in slice1 skin, and these colored green were enriched in the slice2 skin."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1aea2d6e",
   "metadata": {
    "Rmd_chunk_options": "fig.width=9,fig.height = 4, fig.wide = TRUE, fig.align = \"center\"",
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setwd(fig.dir)\n",
    "pdf(paste0(samples, '.ranknet.pdf'), width = 10, height = 10)\n",
    "# rankNet(cch, mode = \"comparison\", stacked = T, do.stat = TRUE)\n",
    "rankNet(cch, mode = \"comparison\", stacked = F, do.stat = TRUE)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "860237ff",
   "metadata": {},
   "source": [
    "### Compare outgoing (or incoming) signaling associated with each cell population\n",
    "The above analysis summarize the information from the outgoing and incoming signaling together. We can also compare the outgoing (or incoming) signaling pattern between two datasets, allowing to identify signaling pathways/ligand-receptors that exhibit different signaling patterns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4ef90478",
   "metadata": {
    "Rmd_chunk_options": "fig.width=11, fig.height = 8, fig.align = \"center\"",
    "jupyter": {
     "source_hidden": true
    },
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.14.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(ComplexHeatmap)\n",
    "setwd(fig.dir)\n",
    "if (length(cch.list)==3) {\n",
    "    pathway.union <- union(union(cch.list[[1]]@netP$pathways, cch.list[[2]]@netP$pathways), cch.list[[3]]@netP$pathways)\n",
    "} else if (length(cch.list)==2) {\n",
    "    pathway.union <- union(cch.list[[1]]@netP$pathways, cch.list[[2]]@netP$pathways)\n",
    "}\n",
    "\n",
    "title = names(cch.list)[[1]]\n",
    "pdf(paste0(samples, '.signalingRole_heatmap_outgoing_', title,'.pdf'), width = 14, height = 14)\n",
    "netAnalysis_signalingRole_heatmap(cch.list[[title]], pattern = \"outgoing\", signaling = pathway.union, title = title, width = 14, height = 14)\n",
    "dev.off()\n",
    "pdf(paste0(samples, '.signalingRole_heatmap_incoming_', title,'.pdf'), width = 14, height = 14)\n",
    "netAnalysis_signalingRole_heatmap(cch.list[[title]], pattern = \"incoming\", signaling = pathway.union, title = title, width = 14, height = 14, color.heatmap = \"GnBu\")\n",
    "dev.off()\n",
    "pdf(paste0(samples, '.signalingRole_heatmap_all', title,'.pdf'), width = 14, height = 14)\n",
    "netAnalysis_signalingRole_heatmap(cch.list[[title]], pattern = \"all\", signaling = pathway.union, title = title, width = 14, height = 14, color.heatmap = \"OrRd\")\n",
    "dev.off()\n",
    "title = names(cch.list)[[2]]\n",
    "pdf(paste0(samples, '.signalingRole_heatmap_outgoing_', title,'.pdf'), width = 14, height = 14)\n",
    "netAnalysis_signalingRole_heatmap(cch.list[[title]], pattern = \"outgoing\", signaling = pathway.union, title = title, width = 14, height = 14)\n",
    "dev.off()\n",
    "pdf(paste0(samples, '.signalingRole_heatmap_incoming_', title,'.pdf'), width = 14, height = 14)\n",
    "netAnalysis_signalingRole_heatmap(cch.list[[title]], pattern = \"incoming\", signaling = pathway.union, title = title, width = 14, height = 14, color.heatmap = \"GnBu\")\n",
    "dev.off()\n",
    "pdf(paste0(samples, '.signalingRole_heatmap_all', title,'.pdf'), width = 14, height = 14)\n",
    "netAnalysis_signalingRole_heatmap(cch.list[[title]], pattern = \"all\", signaling = pathway.union, title = title, width = 14, height = 14, color.heatmap = \"OrRd\")\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ace8fc7d",
   "metadata": {},
   "source": [
    "# Part III: Identify the upgulated and down-regulated signaling ligand-receptor pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9df98a39",
   "metadata": {},
   "source": [
    "## Identify dysfunctional signaling by comparing the communication probabities"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eabda45e",
   "metadata": {},
   "source": [
    "We can compare the communication probabilities mediated by ligand-receptor pairs from some cell groups to other cell groups. This can be done by setting `comparison` in the function `netVisual_bubble`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5490bcbe",
   "metadata": {},
   "source": [
    "Moreover, we can identify the upgulated (increased) and down-regulated (decreased) signaling ligand-receptor pairs in one dataset compared to the other dataset. This can be done by specifying `max.dataset` and `min.dataset` in the function `netVisual_bubble`. The increased signaling means these signaling have higher communication probability (strength) in one dataset compared to the other dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "all cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b2e0da61",
   "metadata": {
    "Rmd_chunk_options": "fig.width=12, fig.height = 5, fig.align = \"center\"",
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in `$<-.data.frame`(`*tmp*`, \"source.target\", value = character(0)): replacement has 0 rows, data has 1>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in `$<-.data.frame`(`*tmp*`, \"source.target\", value = character(0)): replacement has 0 rows, data has 1>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "setwd(fig.dir)\n",
    "all_celltypes = unique(cch@idents[[1]])\n",
    "setwd(fig.dir)\n",
    "for (i in 1:length(all_celltypes)) {  # length(all_celltypes)\n",
    "    f1 = paste0(samples, '.signalingRole_heatmap_as_source_', all_celltypes[i],'_increased.pdf')\n",
    "    f2 = paste0(samples, '.signalingRole_heatmap_as_target_', all_celltypes[i],'_increased.pdf')\n",
    "    tryCatch({\n",
    "        g = netVisual_bubble(cch, sources.use = all_celltypes[i], comparison = c(1, 2), max.dataset = 2, \n",
    "            title.name = paste0(\"Increased signaling in \",names(cch.list)[2]), angle.x = 45, remove.isolate = T, color.text.use = F)\n",
    "        ggsave(filename = f1, plot=g, width = 10, height = 10)\n",
    "        \n",
    "        g = netVisual_bubble(cch, targets.use = all_celltypes[i],  comparison = c(1, 2), max.dataset = 2, \n",
    "            title.name = paste0(\"Increased signaling in \",names(cch.list)[2]), angle.x = 45, remove.isolate = T, color.text.use = F)\n",
    "        ggsave(filename = f2, plot=g, width = 10, height = 10)\n",
    "    }, error = function(e) {\n",
    "        print(e)\n",
    "    })\n",
    "}\n",
    "\n",
    "setwd(fig.dir)\n",
    "all_celltypes = unique(cch@idents[[1]])\n",
    "setwd(fig.dir)\n",
    "for (i in 1:length(all_celltypes)) {  # length(all_celltypes)\n",
    "    f1 = paste0(samples, '.signalingRole_heatmap_as_source_', all_celltypes[i],'_decreased.pdf')\n",
    "    f2 = paste0(samples, '.signalingRole_heatmap_as_target_', all_celltypes[i],'_decreased.pdf')\n",
    "    tryCatch({\n",
    "        g = netVisual_bubble(cch, sources.use = all_celltypes[i], comparison = c(1, 2), max.dataset = 1, \n",
    "            title.name = paste0(\"Increased signaling in \",names(cch.list)[2]), angle.x = 45, remove.isolate = T, color.text.use = F)\n",
    "        ggsave(filename = f1, plot=g, width = 10, height = 10)\n",
    "        \n",
    "        g = netVisual_bubble(cch, targets.use = all_celltypes[i],  comparison = c(1, 2), max.dataset = 1, \n",
    "            title.name = paste0(\"Increased signaling in \",names(cch.list)[2]), angle.x = 45, remove.isolate = T, color.text.use = F)\n",
    "        ggsave(filename = f2, plot=g, width = 10, height = 10)\n",
    "    }, error = function(e) {\n",
    "        print(e)\n",
    "    })\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "all available pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'COLLAGEN'</li><li>'MK'</li><li>'LAMININ'</li><li>'APP'</li><li>'FLRT'</li><li>'FN1'</li><li>'Cholesterol'</li><li>'CLDN'</li><li>'CD99'</li><li>'ADGRL'</li><li>'CypA'</li><li>'IGF'</li><li>'IGFBP'</li><li>'NOTCH'</li><li>'PTPRM'</li><li>'GAP'</li><li>'PDGF'</li><li>'THBS'</li><li>'PTN'</li><li>'GAS'</li><li>'NCAM'</li><li>'ADGRA'</li><li>'JAM'</li><li>'VCAM'</li><li>'BMP'</li><li>'FGF'</li><li>'MIF'</li><li>'NRXN'</li><li>'TENASCIN'</li><li>'CD46'</li><li>'CNTN'</li><li>'SEMA3'</li><li>'HSPG'</li><li>'ESAM'</li><li>'CDH5'</li><li>'VEGF'</li><li>'PTPR'</li><li>'Androstenedione'</li><li>'ADGRG'</li><li>'CDH'</li><li>'MPZ'</li><li>'ADGRE'</li><li>'PECAM1'</li><li>'TWEAK'</li><li>'NEGR'</li><li>'Testosterone'</li><li>'SEMA4'</li><li>'GRN'</li><li>'SEMA6'</li><li>'Desmosterol'</li><li>'NECTIN'</li><li>'EPHA'</li><li>'CXCL'</li><li>'Netrin'</li><li>'PCDH'</li><li>'PROS'</li><li>'UNC5'</li><li>'ANGPT'</li><li>'ANGPTL'</li><li>'CCL'</li><li>'EPHB'</li><li>'RELN'</li><li>'VISFATIN'</li><li>'27HC'</li><li>'SEMA7'</li><li>'CALCR'</li><li>'NT'</li><li>'CADM'</li><li>'ncWNT'</li><li>'TGFb'</li><li>'CSPG4'</li><li>'MAG'</li><li>'Androsterone'</li><li>'EGF'</li><li>'WNT'</li><li>'VISTA'</li><li>'NRG'</li><li>'RA'</li><li>'SELE'</li><li>'SPP1'</li><li>'Prostaglandin'</li><li>'AGRN'</li><li>'SLIT'</li><li>'CD34'</li><li>'LIFR'</li><li>'EDN'</li><li>'DHEA'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'COLLAGEN'\n",
       "\\item 'MK'\n",
       "\\item 'LAMININ'\n",
       "\\item 'APP'\n",
       "\\item 'FLRT'\n",
       "\\item 'FN1'\n",
       "\\item 'Cholesterol'\n",
       "\\item 'CLDN'\n",
       "\\item 'CD99'\n",
       "\\item 'ADGRL'\n",
       "\\item 'CypA'\n",
       "\\item 'IGF'\n",
       "\\item 'IGFBP'\n",
       "\\item 'NOTCH'\n",
       "\\item 'PTPRM'\n",
       "\\item 'GAP'\n",
       "\\item 'PDGF'\n",
       "\\item 'THBS'\n",
       "\\item 'PTN'\n",
       "\\item 'GAS'\n",
       "\\item 'NCAM'\n",
       "\\item 'ADGRA'\n",
       "\\item 'JAM'\n",
       "\\item 'VCAM'\n",
       "\\item 'BMP'\n",
       "\\item 'FGF'\n",
       "\\item 'MIF'\n",
       "\\item 'NRXN'\n",
       "\\item 'TENASCIN'\n",
       "\\item 'CD46'\n",
       "\\item 'CNTN'\n",
       "\\item 'SEMA3'\n",
       "\\item 'HSPG'\n",
       "\\item 'ESAM'\n",
       "\\item 'CDH5'\n",
       "\\item 'VEGF'\n",
       "\\item 'PTPR'\n",
       "\\item 'Androstenedione'\n",
       "\\item 'ADGRG'\n",
       "\\item 'CDH'\n",
       "\\item 'MPZ'\n",
       "\\item 'ADGRE'\n",
       "\\item 'PECAM1'\n",
       "\\item 'TWEAK'\n",
       "\\item 'NEGR'\n",
       "\\item 'Testosterone'\n",
       "\\item 'SEMA4'\n",
       "\\item 'GRN'\n",
       "\\item 'SEMA6'\n",
       "\\item 'Desmosterol'\n",
       "\\item 'NECTIN'\n",
       "\\item 'EPHA'\n",
       "\\item 'CXCL'\n",
       "\\item 'Netrin'\n",
       "\\item 'PCDH'\n",
       "\\item 'PROS'\n",
       "\\item 'UNC5'\n",
       "\\item 'ANGPT'\n",
       "\\item 'ANGPTL'\n",
       "\\item 'CCL'\n",
       "\\item 'EPHB'\n",
       "\\item 'RELN'\n",
       "\\item 'VISFATIN'\n",
       "\\item '27HC'\n",
       "\\item 'SEMA7'\n",
       "\\item 'CALCR'\n",
       "\\item 'NT'\n",
       "\\item 'CADM'\n",
       "\\item 'ncWNT'\n",
       "\\item 'TGFb'\n",
       "\\item 'CSPG4'\n",
       "\\item 'MAG'\n",
       "\\item 'Androsterone'\n",
       "\\item 'EGF'\n",
       "\\item 'WNT'\n",
       "\\item 'VISTA'\n",
       "\\item 'NRG'\n",
       "\\item 'RA'\n",
       "\\item 'SELE'\n",
       "\\item 'SPP1'\n",
       "\\item 'Prostaglandin'\n",
       "\\item 'AGRN'\n",
       "\\item 'SLIT'\n",
       "\\item 'CD34'\n",
       "\\item 'LIFR'\n",
       "\\item 'EDN'\n",
       "\\item 'DHEA'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'COLLAGEN'\n",
       "2. 'MK'\n",
       "3. 'LAMININ'\n",
       "4. 'APP'\n",
       "5. 'FLRT'\n",
       "6. 'FN1'\n",
       "7. 'Cholesterol'\n",
       "8. 'CLDN'\n",
       "9. 'CD99'\n",
       "10. 'ADGRL'\n",
       "11. 'CypA'\n",
       "12. 'IGF'\n",
       "13. 'IGFBP'\n",
       "14. 'NOTCH'\n",
       "15. 'PTPRM'\n",
       "16. 'GAP'\n",
       "17. 'PDGF'\n",
       "18. 'THBS'\n",
       "19. 'PTN'\n",
       "20. 'GAS'\n",
       "21. 'NCAM'\n",
       "22. 'ADGRA'\n",
       "23. 'JAM'\n",
       "24. 'VCAM'\n",
       "25. 'BMP'\n",
       "26. 'FGF'\n",
       "27. 'MIF'\n",
       "28. 'NRXN'\n",
       "29. 'TENASCIN'\n",
       "30. 'CD46'\n",
       "31. 'CNTN'\n",
       "32. 'SEMA3'\n",
       "33. 'HSPG'\n",
       "34. 'ESAM'\n",
       "35. 'CDH5'\n",
       "36. 'VEGF'\n",
       "37. 'PTPR'\n",
       "38. 'Androstenedione'\n",
       "39. 'ADGRG'\n",
       "40. 'CDH'\n",
       "41. 'MPZ'\n",
       "42. 'ADGRE'\n",
       "43. 'PECAM1'\n",
       "44. 'TWEAK'\n",
       "45. 'NEGR'\n",
       "46. 'Testosterone'\n",
       "47. 'SEMA4'\n",
       "48. 'GRN'\n",
       "49. 'SEMA6'\n",
       "50. 'Desmosterol'\n",
       "51. 'NECTIN'\n",
       "52. 'EPHA'\n",
       "53. 'CXCL'\n",
       "54. 'Netrin'\n",
       "55. 'PCDH'\n",
       "56. 'PROS'\n",
       "57. 'UNC5'\n",
       "58. 'ANGPT'\n",
       "59. 'ANGPTL'\n",
       "60. 'CCL'\n",
       "61. 'EPHB'\n",
       "62. 'RELN'\n",
       "63. 'VISFATIN'\n",
       "64. '27HC'\n",
       "65. 'SEMA7'\n",
       "66. 'CALCR'\n",
       "67. 'NT'\n",
       "68. 'CADM'\n",
       "69. 'ncWNT'\n",
       "70. 'TGFb'\n",
       "71. 'CSPG4'\n",
       "72. 'MAG'\n",
       "73. 'Androsterone'\n",
       "74. 'EGF'\n",
       "75. 'WNT'\n",
       "76. 'VISTA'\n",
       "77. 'NRG'\n",
       "78. 'RA'\n",
       "79. 'SELE'\n",
       "80. 'SPP1'\n",
       "81. 'Prostaglandin'\n",
       "82. 'AGRN'\n",
       "83. 'SLIT'\n",
       "84. 'CD34'\n",
       "85. 'LIFR'\n",
       "86. 'EDN'\n",
       "87. 'DHEA'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"COLLAGEN\"        \"MK\"              \"LAMININ\"         \"APP\"            \n",
       " [5] \"FLRT\"            \"FN1\"             \"Cholesterol\"     \"CLDN\"           \n",
       " [9] \"CD99\"            \"ADGRL\"           \"CypA\"            \"IGF\"            \n",
       "[13] \"IGFBP\"           \"NOTCH\"           \"PTPRM\"           \"GAP\"            \n",
       "[17] \"PDGF\"            \"THBS\"            \"PTN\"             \"GAS\"            \n",
       "[21] \"NCAM\"            \"ADGRA\"           \"JAM\"             \"VCAM\"           \n",
       "[25] \"BMP\"             \"FGF\"             \"MIF\"             \"NRXN\"           \n",
       "[29] \"TENASCIN\"        \"CD46\"            \"CNTN\"            \"SEMA3\"          \n",
       "[33] \"HSPG\"            \"ESAM\"            \"CDH5\"            \"VEGF\"           \n",
       "[37] \"PTPR\"            \"Androstenedione\" \"ADGRG\"           \"CDH\"            \n",
       "[41] \"MPZ\"             \"ADGRE\"           \"PECAM1\"          \"TWEAK\"          \n",
       "[45] \"NEGR\"            \"Testosterone\"    \"SEMA4\"           \"GRN\"            \n",
       "[49] \"SEMA6\"           \"Desmosterol\"     \"NECTIN\"          \"EPHA\"           \n",
       "[53] \"CXCL\"            \"Netrin\"          \"PCDH\"            \"PROS\"           \n",
       "[57] \"UNC5\"            \"ANGPT\"           \"ANGPTL\"          \"CCL\"            \n",
       "[61] \"EPHB\"            \"RELN\"            \"VISFATIN\"        \"27HC\"           \n",
       "[65] \"SEMA7\"           \"CALCR\"           \"NT\"              \"CADM\"           \n",
       "[69] \"ncWNT\"           \"TGFb\"            \"CSPG4\"           \"MAG\"            \n",
       "[73] \"Androsterone\"    \"EGF\"             \"WNT\"             \"VISTA\"          \n",
       "[77] \"NRG\"             \"RA\"              \"SELE\"            \"SPP1\"           \n",
       "[81] \"Prostaglandin\"   \"AGRN\"            \"SLIT\"            \"CD34\"           \n",
       "[85] \"LIFR\"            \"EDN\"             \"DHEA\"           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(unlist(sapply(cellchat@netP, function(x) {x$pathways})))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in subsetCommunication_internal(net, LR, cells.level, slot.name = slot.name,     sources.use = sources.use, targets.use = targets.use, signaling = signaling,     pairLR.use = pairLR.use, thresh = thresh, datasets = datasets,     ligand.pvalues = ligand.pvalues, ligand.logFC = ligand.logFC,     ligand.pct.1 = ligand.pct.1, ligand.pct.2 = ligand.pct.2,     receptor.pvalues = receptor.pvalues, receptor.logFC = receptor.logFC,     receptor.pct.1 = receptor.pct.1, receptor.pct.2 = receptor.pct.2): No significant signaling interactions are inferred based on the input!>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in subsetCommunication_internal(net, LR, cells.level, slot.name = slot.name,     sources.use = sources.use, targets.use = targets.use, signaling = signaling,     pairLR.use = pairLR.use, thresh = thresh, datasets = datasets,     ligand.pvalues = ligand.pvalues, ligand.logFC = ligand.logFC,     ligand.pct.1 = ligand.pct.1, ligand.pct.2 = ligand.pct.2,     receptor.pvalues = receptor.pvalues, receptor.logFC = receptor.logFC,     receptor.pct.1 = receptor.pct.1, receptor.pct.2 = receptor.pct.2): No significant signaling interactions are inferred based on the input!>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in subsetCommunication_internal(net, LR, cells.level, slot.name = slot.name,     sources.use = sources.use, targets.use = targets.use, signaling = signaling,     pairLR.use = pairLR.use, thresh = thresh, datasets = datasets,     ligand.pvalues = ligand.pvalues, ligand.logFC = ligand.logFC,     ligand.pct.1 = ligand.pct.1, ligand.pct.2 = ligand.pct.2,     receptor.pvalues = receptor.pvalues, receptor.logFC = receptor.logFC,     receptor.pct.1 = receptor.pct.1, receptor.pct.2 = receptor.pct.2): No significant signaling interactions are inferred based on the input!>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in subsetCommunication_internal(net, LR, cells.level, slot.name = slot.name,     sources.use = sources.use, targets.use = targets.use, signaling = signaling,     pairLR.use = pairLR.use, thresh = thresh, datasets = datasets,     ligand.pvalues = ligand.pvalues, ligand.logFC = ligand.logFC,     ligand.pct.1 = ligand.pct.1, ligand.pct.2 = ligand.pct.2,     receptor.pvalues = receptor.pvalues, receptor.logFC = receptor.logFC,     receptor.pct.1 = receptor.pct.1, receptor.pct.2 = receptor.pct.2): No significant signaling interactions are inferred based on the input!>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in subsetCommunication_internal(net, LR, cells.level, slot.name = slot.name,     sources.use = sources.use, targets.use = targets.use, signaling = signaling,     pairLR.use = pairLR.use, thresh = thresh, datasets = datasets,     ligand.pvalues = ligand.pvalues, ligand.logFC = ligand.logFC,     ligand.pct.1 = ligand.pct.1, ligand.pct.2 = ligand.pct.2,     receptor.pvalues = receptor.pvalues, receptor.logFC = receptor.logFC,     receptor.pct.1 = receptor.pct.1, receptor.pct.2 = receptor.pct.2): No significant signaling interactions are inferred based on the input!>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in subsetCommunication_internal(net, LR, cells.level, slot.name = slot.name,     sources.use = sources.use, targets.use = targets.use, signaling = signaling,     pairLR.use = pairLR.use, thresh = thresh, datasets = datasets,     ligand.pvalues = ligand.pvalues, ligand.logFC = ligand.logFC,     ligand.pct.1 = ligand.pct.1, ligand.pct.2 = ligand.pct.2,     receptor.pvalues = receptor.pvalues, receptor.logFC = receptor.logFC,     receptor.pct.1 = receptor.pct.1, receptor.pct.2 = receptor.pct.2): No significant signaling interactions are inferred based on the input!>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in subsetCommunication_internal(net, LR, cells.level, slot.name = slot.name,     sources.use = sources.use, targets.use = targets.use, signaling = signaling,     pairLR.use = pairLR.use, thresh = thresh, datasets = datasets,     ligand.pvalues = ligand.pvalues, ligand.logFC = ligand.logFC,     ligand.pct.1 = ligand.pct.1, ligand.pct.2 = ligand.pct.2,     receptor.pvalues = receptor.pvalues, receptor.logFC = receptor.logFC,     receptor.pct.1 = receptor.pct.1, receptor.pct.2 = receptor.pct.2): No significant signaling interactions are inferred based on the input!>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in subsetCommunication_internal(net, LR, cells.level, slot.name = slot.name,     sources.use = sources.use, targets.use = targets.use, signaling = signaling,     pairLR.use = pairLR.use, thresh = thresh, datasets = datasets,     ligand.pvalues = ligand.pvalues, ligand.logFC = ligand.logFC,     ligand.pct.1 = ligand.pct.1, ligand.pct.2 = ligand.pct.2,     receptor.pvalues = receptor.pvalues, receptor.logFC = receptor.logFC,     receptor.pct.1 = receptor.pct.1, receptor.pct.2 = receptor.pct.2): No significant signaling interactions are inferred based on the input!>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in subsetCommunication_internal(net, LR, cells.level, slot.name = slot.name,     sources.use = sources.use, targets.use = targets.use, signaling = signaling,     pairLR.use = pairLR.use, thresh = thresh, datasets = datasets,     ligand.pvalues = ligand.pvalues, ligand.logFC = ligand.logFC,     ligand.pct.1 = ligand.pct.1, ligand.pct.2 = ligand.pct.2,     receptor.pvalues = receptor.pvalues, receptor.logFC = receptor.logFC,     receptor.pct.1 = receptor.pct.1, receptor.pct.2 = receptor.pct.2): No significant signaling interactions are inferred based on the input!>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "setwd(fig.dir)\n",
    "all_pathways = unique(unlist(sapply(cch@netP, function(x) {x$pathways})))\n",
    "\n",
    "for (i in 1:length(all_pathways)) {  # length(all_celltypes)\n",
    "    f1 = paste0(samples, '.signalingRole_heatmap_as_source_', all_pathways[i],'_increased.pdf')\n",
    "    tryCatch({\n",
    "        g = netVisual_bubble(cch, signaling = all_pathways[i], comparison = c(1, 2), max.dataset = 2, \n",
    "            title.name = paste0(\"Increased signaling in \",names(cch.list)[2]), angle.x = 45, remove.isolate = T, color.text.use = F)\n",
    "        ggsave(filename = f1, plot=g, width = 10, height = 10)        \n",
    "    }, error = function(e) {\n",
    "        print(e)\n",
    "    })\n",
    "}\n",
    "for (i in 1:length(all_celltypes)) {  # length(all_celltypes)\n",
    "    f1 = paste0(samples, '.signalingRole_heatmap_as_source_', all_pathways[i],'_decreased.pdf')\n",
    "    tryCatch({\n",
    "        g = netVisual_bubble(cch, signaling = all_pathways[i], comparison = c(1, 2), max.dataset = 1, \n",
    "            title.name = paste0(\"Increased signaling in \",names(cch.list)[2]), angle.x = 45, remove.isolate = T, color.text.use = F)\n",
    "        ggsave(filename = f1, plot=g, width = 10, height = 10)\n",
    "        \n",
    "    }, error = function(e) {\n",
    "        print(e)\n",
    "    })\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63847389",
   "metadata": {},
   "source": [
    "NB: The ligand-receptor pairs shown in the bubble plot can be accessed via `signaling.slice2Increased = gg1$data`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e7fe0d2",
   "metadata": {},
   "source": [
    "## Identify differential signaling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a6827f9",
   "metadata": {},
   "source": [
    "The above method for identifying the upgulated and down-regulated signaling is perfomed by comparing the communication probability between two datasets for each L-R pair and each pair of cell groups. Alternative, we can identify the upgulated and down-regulated signaling ligand-receptor pairs based on the differential gene expression analysis. Specifically, we perform differential expression analysis between two biological conditions (i.e., slice1 and slice2) for each cell group, and then obtain the upgulated and down-regulated signaling based on the fold change of ligands in the sender cells and receptors in the receiver cells. Such analysis can be done as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d09326cc",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Use the joint cell labels from the merged CellChat object\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CCL2'</li><li>'CCL14'</li><li>'MIF'</li><li>'NAMPT'</li><li>'MDK'</li><li>'FN1'</li><li>'COL1A1'</li><li>'COL1A2'</li><li>'COL4A1'</li><li>'COL4A2'</li><li>'COL6A1'</li><li>'COL6A2'</li><li>'COL6A3'</li><li>'LAMB1'</li><li>'APP'</li><li>'CDH5'</li><li>'CLDN11'</li><li>'ESAM'</li><li>'JAM3'</li><li>'DLK1'</li><li>'JAG1'</li><li>'PECAM1'</li><li>'GJA5'</li><li>'GJA1'</li><li>'DHCR24'</li><li>'ITGA9'</li><li>'ITGB1'</li><li>'ACKR1'</li><li>'SDC2'</li><li>'LRP1'</li><li>'NCL'</li><li>'CD44'</li><li>'RORA'</li><li>'CD74'</li><li>'CDH5'</li><li>'CLDN11'</li><li>'ESAM'</li><li>'JAM3'</li><li>'NOTCH3'</li><li>'PECAM1'</li><li>'VCAM1'</li><li>'GJA5'</li><li>'GJA1'</li><li>'ITGA5'</li><li>'ITGA9'</li><li>'ITGAV'</li><li>'ITGB1'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CCL2'\n",
       "\\item 'CCL14'\n",
       "\\item 'MIF'\n",
       "\\item 'NAMPT'\n",
       "\\item 'MDK'\n",
       "\\item 'FN1'\n",
       "\\item 'COL1A1'\n",
       "\\item 'COL1A2'\n",
       "\\item 'COL4A1'\n",
       "\\item 'COL4A2'\n",
       "\\item 'COL6A1'\n",
       "\\item 'COL6A2'\n",
       "\\item 'COL6A3'\n",
       "\\item 'LAMB1'\n",
       "\\item 'APP'\n",
       "\\item 'CDH5'\n",
       "\\item 'CLDN11'\n",
       "\\item 'ESAM'\n",
       "\\item 'JAM3'\n",
       "\\item 'DLK1'\n",
       "\\item 'JAG1'\n",
       "\\item 'PECAM1'\n",
       "\\item 'GJA5'\n",
       "\\item 'GJA1'\n",
       "\\item 'DHCR24'\n",
       "\\item 'ITGA9'\n",
       "\\item 'ITGB1'\n",
       "\\item 'ACKR1'\n",
       "\\item 'SDC2'\n",
       "\\item 'LRP1'\n",
       "\\item 'NCL'\n",
       "\\item 'CD44'\n",
       "\\item 'RORA'\n",
       "\\item 'CD74'\n",
       "\\item 'CDH5'\n",
       "\\item 'CLDN11'\n",
       "\\item 'ESAM'\n",
       "\\item 'JAM3'\n",
       "\\item 'NOTCH3'\n",
       "\\item 'PECAM1'\n",
       "\\item 'VCAM1'\n",
       "\\item 'GJA5'\n",
       "\\item 'GJA1'\n",
       "\\item 'ITGA5'\n",
       "\\item 'ITGA9'\n",
       "\\item 'ITGAV'\n",
       "\\item 'ITGB1'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CCL2'\n",
       "2. 'CCL14'\n",
       "3. 'MIF'\n",
       "4. 'NAMPT'\n",
       "5. 'MDK'\n",
       "6. 'FN1'\n",
       "7. 'COL1A1'\n",
       "8. 'COL1A2'\n",
       "9. 'COL4A1'\n",
       "10. 'COL4A2'\n",
       "11. 'COL6A1'\n",
       "12. 'COL6A2'\n",
       "13. 'COL6A3'\n",
       "14. 'LAMB1'\n",
       "15. 'APP'\n",
       "16. 'CDH5'\n",
       "17. 'CLDN11'\n",
       "18. 'ESAM'\n",
       "19. 'JAM3'\n",
       "20. 'DLK1'\n",
       "21. 'JAG1'\n",
       "22. 'PECAM1'\n",
       "23. 'GJA5'\n",
       "24. 'GJA1'\n",
       "25. 'DHCR24'\n",
       "26. 'ITGA9'\n",
       "27. 'ITGB1'\n",
       "28. 'ACKR1'\n",
       "29. 'SDC2'\n",
       "30. 'LRP1'\n",
       "31. 'NCL'\n",
       "32. 'CD44'\n",
       "33. 'RORA'\n",
       "34. 'CD74'\n",
       "35. 'CDH5'\n",
       "36. 'CLDN11'\n",
       "37. 'ESAM'\n",
       "38. 'JAM3'\n",
       "39. 'NOTCH3'\n",
       "40. 'PECAM1'\n",
       "41. 'VCAM1'\n",
       "42. 'GJA5'\n",
       "43. 'GJA1'\n",
       "44. 'ITGA5'\n",
       "45. 'ITGA9'\n",
       "46. 'ITGAV'\n",
       "47. 'ITGB1'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"CCL2\"   \"CCL14\"  \"MIF\"    \"NAMPT\"  \"MDK\"    \"FN1\"    \"COL1A1\" \"COL1A2\"\n",
       " [9] \"COL4A1\" \"COL4A2\" \"COL6A1\" \"COL6A2\" \"COL6A3\" \"LAMB1\"  \"APP\"    \"CDH5\"  \n",
       "[17] \"CLDN11\" \"ESAM\"   \"JAM3\"   \"DLK1\"   \"JAG1\"   \"PECAM1\" \"GJA5\"   \"GJA1\"  \n",
       "[25] \"DHCR24\" \"ITGA9\"  \"ITGB1\"  \"ACKR1\"  \"SDC2\"   \"LRP1\"   \"NCL\"    \"CD44\"  \n",
       "[33] \"RORA\"   \"CD74\"   \"CDH5\"   \"CLDN11\" \"ESAM\"   \"JAM3\"   \"NOTCH3\" \"PECAM1\"\n",
       "[41] \"VCAM1\"  \"GJA5\"   \"GJA1\"   \"ITGA5\"  \"ITGA9\"  \"ITGAV\"  \"ITGB1\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "netMappingDEG = function (object, features.name, thresh = 0.05) \n",
    "{\n",
    "    features.name <- paste0(features.name, \".info\")\n",
    "    if (!(features.name %in% names(object@var.features))) {\n",
    "        stop(\"The input features.name does not exist in `names(object@var.features)`. Please first run `identifyOverExpressedGenes`! \")\n",
    "    }\n",
    "    DEG <- object@var.features[[features.name]]\n",
    "    geneInfo <- object@DB$geneInfo\n",
    "    complex_input <- object@DB$complex\n",
    "    df.net <- subsetCommunication(object, thresh = thresh)\n",
    "    if (is.list(df.net)) {\n",
    "        net <- data.frame()\n",
    "        for (ii in 1:length(df.net)) {\n",
    "            df.net[[ii]]$datasets <- names(df.net)[ii]\n",
    "            net <- rbind(net, df.net[[ii]])\n",
    "        }\n",
    "    }\n",
    "    else {\n",
    "        net <- df.net\n",
    "    }\n",
    "    net$source.ligand <- paste0(net$source, \".\", net$ligand)\n",
    "    net$target.receptor <- paste0(net$target, \".\", net$receptor)\n",
    "    DEG$clusters.features <- paste0(DEG$clusters, \".\", DEG$features)\n",
    "    net <- cbind(net, data.frame(ligand.pvalues = NA, ligand.logFC = NA, \n",
    "        ligand.pct.1 = NA, ligand.pct.2 = NA, receptor.pvalues = NA, \n",
    "        receptor.logFC = NA, receptor.pct.1 = NA, receptor.pct.2 = NA))\n",
    "    idx1.ligand <- net$ligand %in% geneInfo$Symbol\n",
    "    idx2.ligand <- which((net$ligand %in% geneInfo$Symbol) == \n",
    "        \"FALSE\")\n",
    "    idx.pos <- match(net$source.ligand, DEG$clusters.features)\n",
    "    idx1.source.ligand <- which(!is.na(idx.pos))\n",
    "    idx1.clusters.features <- idx.pos[!is.na(idx.pos)]\n",
    "    idx2.source.ligand <- which(idx1.ligand & !(net$source.ligand %in% \n",
    "        DEG$clusters.features))\n",
    "    net[idx1.source.ligand, c(\"ligand.pvalues\", \"ligand.logFC\", \n",
    "        \"ligand.pct.1\", \"ligand.pct.2\")] <- DEG[idx1.clusters.features, \n",
    "        c(\"pvalues\", \"logFC\", \"pct.1\", \"pct.2\")]\n",
    "    if (length(idx2.ligand) > 0) {\n",
    "        net.temp.all <- data.frame()\n",
    "        for (i in 1:length(idx2.ligand)) {\n",
    "            complex <- net$ligand[idx2.ligand[i]]\n",
    "            complexsubunits <- dplyr::select(complex_input[match(complex, \n",
    "                rownames(complex_input), nomatch = 0), ], starts_with(\"subunit\"))\n",
    "            complexsubunitsV <- unlist(complexsubunits)\n",
    "            complexsubunitsV <- unique(complexsubunitsV[complexsubunitsV != \n",
    "                \"\"])\n",
    "            source.ligand.complex <- paste0(net$source[idx2.ligand[i]], \n",
    "                \".\", complexsubunitsV)\n",
    "            idx.pos <- match(source.ligand.complex, DEG$clusters.features)\n",
    "            idx1.clusters.features <- idx.pos[!is.na(idx.pos)]\n",
    "            if (length(idx1.clusters.features) > 0) {\n",
    "                net.temp <- DEG[idx1.clusters.features, c(\"pvalues\", \n",
    "                  \"logFC\", \"pct.1\", \"pct.2\")]\n",
    "                net.temp <- colMeans(net.temp, na.rm = TRUE)\n",
    "                net.temp <- as.data.frame(t(net.temp))\n",
    "                colnames(net.temp) <- c(\"ligand.pvalues\", \"ligand.logFC\", \n",
    "                  \"ligand.pct.1\", \"ligand.pct.2\")\n",
    "            }\n",
    "            else {\n",
    "                net.temp <- data.frame(ligand.pvalues = NA, ligand.logFC = NA, \n",
    "                  ligand.pct.1 = NA, ligand.pct.2 = NA)\n",
    "            }\n",
    "            net.temp.all <- rbind(net.temp.all, net.temp)\n",
    "        }\n",
    "        net[idx2.ligand, c(\"ligand.pvalues\", \"ligand.logFC\", \n",
    "            \"ligand.pct.1\", \"ligand.pct.2\")] <- net.temp.all\n",
    "    }\n",
    "    idx1.receptor <- net$receptor %in% geneInfo$Symbol\n",
    "    idx2.receptor <- which((net$receptor %in% geneInfo$Symbol) == \n",
    "        \"FALSE\")\n",
    "    idx.pos <- match(net$target.receptor, DEG$clusters.features)\n",
    "    idx1.target.receptor <- which(!is.na(idx.pos))\n",
    "    idx1.clusters.features <- idx.pos[!is.na(idx.pos)]\n",
    "    net[idx1.target.receptor, c(\"receptor.pvalues\", \"receptor.logFC\", \n",
    "        \"receptor.pct.1\", \"receptor.pct.2\")] <- DEG[idx1.clusters.features, \n",
    "        c(\"pvalues\", \"logFC\", \"pct.1\", \"pct.2\")]\n",
    "    if (length(idx2.receptor) > 0) {\n",
    "        net.temp.all <- data.frame()\n",
    "        for (i in 1:length(idx2.receptor)) {\n",
    "            complex <- net$receptor[idx2.receptor[i]]\n",
    "            complexsubunits <- dplyr::select(complex_input[match(complex, \n",
    "                rownames(complex_input), nomatch = 0), ], starts_with(\"subunit\"))\n",
    "            complexsubunitsV <- unlist(complexsubunits)\n",
    "            complexsubunitsV <- unique(complexsubunitsV[complexsubunitsV != \n",
    "                \"\"])\n",
    "            target.receptor.complex <- paste0(net$target[idx2.receptor[i]], \n",
    "                \".\", complexsubunitsV)\n",
    "            idx.pos <- match(target.receptor.complex, DEG$clusters.features)\n",
    "            idx1.clusters.features <- idx.pos[!is.na(idx.pos)]\n",
    "            if (length(idx1.clusters.features) > 0) {\n",
    "                net.temp <- DEG[idx1.clusters.features, c(\"pvalues\", \n",
    "                  \"logFC\", \"pct.1\", \"pct.2\")]\n",
    "                net.temp <- colMeans(net.temp, na.rm = TRUE)\n",
    "                net.temp <- as.data.frame(t(net.temp))\n",
    "                colnames(net.temp) <- c(\"receptor.pvalues\", \"receptor.logFC\", \n",
    "                  \"receptor.pct.1\", \"receptor.pct.2\")\n",
    "            }\n",
    "            else {\n",
    "                net.temp <- data.frame(receptor.pvalues = NA, \n",
    "                  receptor.logFC = NA, receptor.pct.1 = NA, receptor.pct.2 = NA)\n",
    "            }\n",
    "            net.temp.all <- rbind(net.temp.all, net.temp)\n",
    "        }\n",
    "        net[idx2.receptor, c(\"receptor.pvalues\", \"receptor.logFC\", \n",
    "            \"receptor.pct.1\", \"receptor.pct.2\")] <- net.temp.all\n",
    "    }\n",
    "    return(net)\n",
    "}\n",
    "# define a positive dataset, i.e., the dataset with positive fold change against the other dataset\n",
    "pos.dataset = names(cch.list)[2]\n",
    "# define a char name used for storing the results of differential expression analysis\n",
    "features.name = paste0(pos.dataset)\n",
    "# perform differential expression analysis\n",
    "# Of note, compared to CellChat version < v2, CellChat v2 now performs an ultra-fast Wilcoxon test using the presto package, which gives smaller values of logFC.\n",
    "#  Thus we here set a smaller value of thresh.fc compared to the original one (thresh.fc = 0.1). Users can also provide a vector and dataframe of customized DEGs by modifying the cellchat@var.features$slice2 and cellchat@var.features$slice2.info.\n",
    "cch <- identifyOverExpressedGenes(cch, pos.dataset = pos.dataset, \n",
    "            features.name = features.name, only.pos = FALSE, thresh.pc = 0.1, thresh.fc = 0.03, thresh.p = 0.8)\n",
    "# map the results of differential expression analysis onto the inferred cell-cell communications to easily manage/subset the ligand-receptor pairs of interest\n",
    "net <- netMappingDEG(cch, features.name = features.name)\n",
    "write.table(net, file = paste0(fig.dir, 'net.txt'), sep = '\\t', quote = F)\n",
    "\n",
    "\n",
    "# extract the ligand-receptor pairs with upregulated ligands in slice2\n",
    "tryCatch({\n",
    "    net.up <- subsetCommunication(cch, net = net, datasets = pos.dataset, ligand.logFC = 0.1, thresh=0.05, receptor.logFC=0.1)\n",
    "    write.table(net.up, file = paste0(fig.dir, 'net.up.txt'), sep = '\\t', quote = F)\n",
    "}, error = function(e) {\n",
    "    print(e)\n",
    "    net.up = NULL\n",
    "})\n",
    "\n",
    "\n",
    "# tryCatch({\n",
    "#     # extract the ligand-receptor pairs with upregulated ligands and upregulated recetptors in slice1, i.e.,downregulated in slice2\n",
    "#     net.down <- subsetCommunication(cch, net = net, datasets = pos.dataset, ligand.logFC = -0.1, thresh=0.05, receptor.logFC=-0.1)\n",
    "#     write.table(net.down, file = paste0(fig.dir, 'net.down.txt'), sep = '\\t', quote = F)\n",
    "# }, error = function(e) {\n",
    "#     print(e)\n",
    "#     net.down = NULL\n",
    "# })\n",
    "\n",
    "# head(net.down[net.down$pathway_name == 'COLLAGEN',], 10)  # -0.14987451\t\n",
    "# head(net.up[net.up$pathway_name == 'COLLAGEN',], 10)  # 0.14747501\t\n",
    "# head(net.up[net.up$target == 'endothelial_perivascular(IGFBP7)',], 10)\t\n",
    "# head(net.up[net.up$source == 'endothelial_perivascular(IGFBP7)',], 10)\t\n",
    "# head(net.down[net.down$target == 'endothelial_perivascular(IGFBP7)',], 10)\t\n",
    "# head(net.down[net.down$source == 'endothelial_perivascular(IGFBP7)',], 10)\n",
    "\n",
    "if (!is.null(net.up)) {\n",
    "    gene.up <- extractGeneSubsetFromPair(net.up, cch)\n",
    "    write.table(gene.up, file = paste0(fig.dir, 'gene.up.txt'), sep = '\\t', quote = F)\n",
    "}\n",
    "gene.up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "79ba86d5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(RAD51B)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(FNDC3B)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(BEX2)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST_LIP(GSTA1)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(KRT19)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST_SM(ADAMTS4)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(SECTM1)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST_LIP(INSL3)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(RPL41)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] B_Plasma(IGLL5)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in `$<-.data.frame`(`*tmp*`, \"source.target\", value = character(0)): replacement has 0 rows, data has 1>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in seq.default(0.5 + length(dataset.name[comparison]), length(group.names0) *     length(dataset.name[comparison]), by = length(dataset.name[comparison])): wrong sign in 'by' argument>\n",
      "[1] B_Plasma(IGKC)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ENDO_SM_0\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(RBFOX1)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(ZFPM2)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST_SM(MT1A)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(SLCO5A1)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] B_Plasma(IGLC2)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] O(PDCD5)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] FIB(COL1A1)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ENDO_SM_3\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ENDO_SM_4\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(BNC2)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(EXT1)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(MAML2)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(ARL15)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(PRKG1)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ENDO_SM_5\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ENDO_SM_6\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ENDO_SM_1\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ENDO_SM_2\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ST(MAGI1)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] ERY_ENDO(HBB)\n",
      "32 Levels: B_Plasma(IGKC) B_Plasma(IGLC2) B_Plasma(IGLL5) ... ST(ZFPM2)\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n",
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# We then visualize the upgulated and down-regulated signaling ligand-receptor pairs using bubble plot or chord diagram.\n",
    "netVisual_bubble <- function (object, sources.use = NULL, targets.use = NULL, signaling = NULL, \n",
    "    pairLR.use = NULL, sort.by.source = FALSE, sort.by.target = FALSE, \n",
    "    sort.by.source.priority = TRUE, color.heatmap = c(\"Spectral\", \n",
    "        \"viridis\"), n.colors = 10, direction = -1, thresh = 0.05, \n",
    "    comparison = NULL, group = NULL, remove.isolate = FALSE, \n",
    "    max.dataset = NULL, min.dataset = NULL, min.quantile = 0, \n",
    "    max.quantile = 1, line.on = TRUE, line.size = 0.2, color.text.use = TRUE, \n",
    "    color.text = NULL, title.name = NULL, font.size = 10, font.size.title = 10, \n",
    "    show.legend = TRUE, grid.on = TRUE, color.grid = \"grey90\", \n",
    "    angle.x = 90, vjust.x = NULL, hjust.x = NULL, return.data = FALSE) \n",
    "{\n",
    "    color.heatmap <- match.arg(color.heatmap)\n",
    "    if (is.list(object@net[[1]])) {\n",
    "        message(\"Comparing communications on a merged object \\n\")\n",
    "    }\n",
    "    else {\n",
    "        message(\"Comparing communications on a single object \\n\")\n",
    "    }\n",
    "    if (is.null(vjust.x) | is.null(hjust.x)) {\n",
    "        angle = c(0, 45, 90)\n",
    "        hjust = c(0, 1, 1)\n",
    "        vjust = c(0, 1, 0.5)\n",
    "        vjust.x = vjust[angle == angle.x]\n",
    "        hjust.x = hjust[angle == angle.x]\n",
    "    }\n",
    "    if (length(color.heatmap) == 1) {\n",
    "        color.use <- tryCatch({\n",
    "            RColorBrewer::brewer.pal(n = n.colors, name = color.heatmap)\n",
    "        }, error = function(e) {\n",
    "            (scales::viridis_pal(option = color.heatmap, direction = -1))(n.colors)\n",
    "        })\n",
    "    }\n",
    "    else {\n",
    "        color.use <- color.heatmap\n",
    "    }\n",
    "    if (direction == -1) {\n",
    "        color.use <- rev(color.use)\n",
    "    }\n",
    "    if (!is.null(pairLR.use)) {\n",
    "        if (!is.data.frame(pairLR.use)) {\n",
    "            stop(\"pairLR.use should be a data frame with a signle column named either 'interaction_name' or 'pathway_name' \")\n",
    "        }\n",
    "        else if (\"pathway_name\" %in% colnames(pairLR.use)) {\n",
    "            pairLR.use$pathway_name <- as.character(pairLR.use$pathway_name)\n",
    "        }\n",
    "        else if (\"interaction_name\" %in% colnames(pairLR.use)) {\n",
    "            pairLR.use$interaction_name <- as.character(pairLR.use$interaction_name)\n",
    "        }\n",
    "    }\n",
    "    if (is.null(comparison)) {\n",
    "        cells.level <- levels(object@idents)\n",
    "        if (is.numeric(sources.use)) {\n",
    "            sources.use <- cells.level[sources.use]\n",
    "        }\n",
    "        if (is.numeric(targets.use)) {\n",
    "            targets.use <- cells.level[targets.use]\n",
    "        }\n",
    "        df.net <- subsetCommunication(object, slot.name = \"net\", \n",
    "            sources.use = sources.use, targets.use = targets.use, \n",
    "            signaling = signaling, pairLR.use = pairLR.use, thresh = thresh)\n",
    "        df.net$source.target <- paste(df.net$source, df.net$target, \n",
    "            sep = \" -> \")\n",
    "        source.target <- paste(rep(sources.use, each = length(targets.use)), \n",
    "            targets.use, sep = \" -> \")\n",
    "        source.target.isolate <- setdiff(source.target, unique(df.net$source.target))\n",
    "        if (length(source.target.isolate) > 0) {\n",
    "            df.net.isolate <- as.data.frame(matrix(NA, nrow = length(source.target.isolate), \n",
    "                ncol = ncol(df.net)))\n",
    "            colnames(df.net.isolate) <- colnames(df.net)\n",
    "            df.net.isolate$source.target <- source.target.isolate\n",
    "            # print('1')\n",
    "            # print(unique(df.net.isolate$source.target))\n",
    "            df.net.isolate$interaction_name_2 <- df.net$interaction_name_2[1]\n",
    "            df.net.isolate$pval <- 1\n",
    "            a <- stringr::str_split(df.net.isolate$source.target, \n",
    "                \" -> \", simplify = T)\n",
    "            df.net.isolate$source <- as.character(a[, 1])\n",
    "            df.net.isolate$target <- as.character(a[, 2])\n",
    "            df.net <- rbind(df.net, df.net.isolate)\n",
    "        }\n",
    "        df.net$pval[df.net$pval > 0.05] = 1\n",
    "        df.net$pval[df.net$pval > 0.01 & df.net$pval <= 0.05] = 2\n",
    "        df.net$pval[df.net$pval <= 0.01] = 3\n",
    "        df.net$prob[df.net$prob == 0] <- NA\n",
    "        df.net$prob.original <- df.net$prob\n",
    "        df.net$prob <- -1/log(df.net$prob)\n",
    "        idx1 <- which(is.infinite(df.net$prob) | df.net$prob < \n",
    "            0)\n",
    "        if (sum(idx1) > 0) {\n",
    "            values.assign <- seq(max(df.net$prob, na.rm = T) * \n",
    "                1.1, max(df.net$prob, na.rm = T) * 1.5, length.out = length(idx1))\n",
    "            position <- sort(prob.original[idx1], index.return = TRUE)$ix\n",
    "            df.net$prob[idx1] <- values.assign[match(1:length(idx1), \n",
    "                position)]\n",
    "        }\n",
    "        df.net$source <- factor(df.net$source, levels = cells.level[cells.level %in% \n",
    "            unique(df.net$source)])\n",
    "        df.net$target <- factor(df.net$target, levels = cells.level[cells.level %in% \n",
    "            unique(df.net$target)])\n",
    "        group.names <- paste(rep(levels(df.net$source), each = length(levels(df.net$target))), \n",
    "            levels(df.net$target), sep = \" -> \")\n",
    "        df.net$interaction_name_2 <- as.character(df.net$interaction_name_2)\n",
    "        df.net <- with(df.net, df.net[order(interaction_name_2), \n",
    "            ])\n",
    "        df.net$interaction_name_2 <- factor(df.net$interaction_name_2, \n",
    "            levels = unique(df.net$interaction_name_2))\n",
    "        cells.order <- group.names\n",
    "        df.net$source.target <- factor(df.net$source.target, \n",
    "            levels = cells.order)\n",
    "        # print('2')\n",
    "        # print(unique(df.net$source.target))\n",
    "        df <- df.net\n",
    "    }\n",
    "    else {\n",
    "        dataset.name <- names(object@net)\n",
    "        df.net.all <- subsetCommunication(object, slot.name = \"net\", \n",
    "            sources.use = sources.use, targets.use = targets.use, \n",
    "            signaling = signaling, pairLR.use = pairLR.use, thresh = thresh)\n",
    "        df.all <- data.frame()\n",
    "        for (ii in 1:length(comparison)) {\n",
    "            cells.level <- levels(object@idents[[comparison[ii]]])\n",
    "            if (is.numeric(sources.use)) {\n",
    "                sources.use <- cells.level[sources.use]\n",
    "            }\n",
    "            if (is.numeric(targets.use)) {\n",
    "                targets.use <- cells.level[targets.use]\n",
    "            }\n",
    "            df.net <- df.net.all[[comparison[ii]]]\n",
    "            df.net$interaction_name_2 <- as.character(df.net$interaction_name_2)\n",
    "            df.net$source.target <- paste(df.net$source, df.net$target, \n",
    "                sep = \" -> \")\n",
    "            # print('3')\n",
    "            # print(unique(df.net$source.target))\n",
    "            source.target <- paste(rep(sources.use, each = length(targets.use)), \n",
    "                targets.use, sep = \" -> \")\n",
    "            source.target.isolate <- setdiff(source.target, unique(df.net$source.target))\n",
    "            if (length(source.target.isolate) > 0) {\n",
    "                df.net.isolate <- as.data.frame(matrix(NA, nrow = length(source.target.isolate), \n",
    "                  ncol = ncol(df.net)))\n",
    "                colnames(df.net.isolate) <- colnames(df.net)\n",
    "                df.net.isolate$source.target <- source.target.isolate\n",
    "                # print('4')\n",
    "                # print(unique(df.net.isolate$source.target))\n",
    "                df.net.isolate$interaction_name_2 <- df.net$interaction_name_2[1]\n",
    "                df.net.isolate$pval <- 1\n",
    "                a <- stringr::str_split(df.net.isolate$source.target, \n",
    "                  \" -> \", simplify = T)\n",
    "                df.net.isolate$source <- as.character(a[, 1])\n",
    "                df.net.isolate$target <- as.character(a[, 2])\n",
    "                df.net <- rbind(df.net, df.net.isolate)\n",
    "            }\n",
    "            df.net$source <- factor(df.net$source, levels = cells.level[cells.level %in% \n",
    "                unique(df.net$source)])\n",
    "            df.net$target <- factor(df.net$target, levels = cells.level[cells.level %in% \n",
    "                unique(df.net$target)])\n",
    "            group.names <- paste(rep(levels(df.net$source), each = length(levels(df.net$target))), \n",
    "                levels(df.net$target), sep = \" -> \")\n",
    "            group.names0 <- group.names\n",
    "            group.names <- paste0(group.names0, \" (\", dataset.name[comparison[ii]], \n",
    "                \")\")\n",
    "            if (nrow(df.net) > 0) {\n",
    "                df.net$pval[df.net$pval > 0.05] = 1\n",
    "                df.net$pval[df.net$pval > 0.01 & df.net$pval <= \n",
    "                  0.05] = 2\n",
    "                df.net$pval[df.net$pval <= 0.01] = 3\n",
    "                df.net$prob[df.net$prob == 0] <- NA\n",
    "                df.net$prob.original <- df.net$prob\n",
    "                df.net$prob <- -1/log(df.net$prob)\n",
    "            }\n",
    "            else {\n",
    "                df.net <- as.data.frame(matrix(NA, nrow = length(group.names), \n",
    "                  ncol = 5))\n",
    "                colnames(df.net) <- c(\"interaction_name_2\", \"source.target\", \n",
    "                  \"prob\", \"pval\", \"prob.original\")\n",
    "                df.net$source.target <- group.names0\n",
    "                # print('5')\n",
    "                # print(unique(df.net$source.target))\n",
    "            }\n",
    "            df.net$group.names <- as.character(df.net$source.target)\n",
    "            df.net$source.target <- paste0(df.net$source.target, \" (\", dataset.name[comparison[ii]], \")\")\n",
    "            # print('6')\n",
    "            # print(unique(df.net$source.target))\n",
    "            df.net$dataset <- dataset.name[comparison[ii]]\n",
    "            df.all <- rbind(df.all, df.net)\n",
    "        }\n",
    "        if (nrow(df.all) == 0) {\n",
    "            stop(\"No interactions are detected. Please consider changing the cell groups for analysis. \")\n",
    "        }\n",
    "        idx1 <- which(is.infinite(df.all$prob) | df.all$prob < \n",
    "            0)\n",
    "        if (sum(idx1) > 0) {\n",
    "            values.assign <- seq(max(df.all$prob, na.rm = T) * \n",
    "                1.1, max(df.all$prob, na.rm = T) * 1.5, length.out = length(idx1))\n",
    "            position <- sort(df.all$prob.original[idx1], index.return = TRUE)$ix\n",
    "            df.all$prob[idx1] <- values.assign[match(1:length(idx1), \n",
    "                position)]\n",
    "        }\n",
    "        df.all$interaction_name_2[is.na(df.all$interaction_name_2)] <- df.all$interaction_name_2[!is.na(df.all$interaction_name_2)][1]\n",
    "        df <- df.all\n",
    "        df <- with(df, df[order(interaction_name_2), ])\n",
    "        df$interaction_name_2 <- factor(df$interaction_name_2, \n",
    "            levels = unique(df$interaction_name_2))\n",
    "        cells.order <- c()\n",
    "        dataset.name.order <- c()\n",
    "        for (i in 1:length(group.names0)) {\n",
    "            for (j in 1:length(comparison)) {\n",
    "                cells.order <- c(cells.order, paste0(group.names0[i], \n",
    "                  \" (\", dataset.name[comparison[j]], \")\"))\n",
    "                dataset.name.order <- c(dataset.name.order, dataset.name[comparison[j]])\n",
    "            }\n",
    "        }\n",
    "        df$source.target <- factor(df$source.target, levels = cells.order)\n",
    "        # print('7')\n",
    "        # print(unique(df$source.target))\n",
    "    }\n",
    "    min.cutoff <- quantile(df$prob, min.quantile, na.rm = T)\n",
    "    max.cutoff <- quantile(df$prob, max.quantile, na.rm = T)\n",
    "    df$prob[df$prob < min.cutoff] <- min.cutoff\n",
    "    df$prob[df$prob > max.cutoff] <- max.cutoff\n",
    "    if (remove.isolate) {\n",
    "        df <- df[!is.na(df$prob), ]\n",
    "        line.on <- FALSE\n",
    "    }\n",
    "    if (!is.null(max.dataset)) {\n",
    "        signaling <- as.character(unique(df$interaction_name_2))\n",
    "        for (i in signaling) {\n",
    "            df.i <- df[df$interaction_name_2 == i, , drop = FALSE]\n",
    "            cell <- as.character(unique(df.i$group.names))\n",
    "            for (j in cell) {\n",
    "                df.i.j <- df.i[df.i$group.names == j, , drop = FALSE]\n",
    "                values <- df.i.j$prob\n",
    "                idx.max <- which(values == max(values, na.rm = T))\n",
    "                idx.min <- which(values == min(values, na.rm = T))\n",
    "                dataset.na <- c(df.i.j$dataset[is.na(values)], \n",
    "                  setdiff(dataset.name[comparison], df.i.j$dataset))\n",
    "                if (length(idx.max) > 0) {\n",
    "                  if (!(df.i.j$dataset[idx.max] %in% dataset.name[max.dataset])) {\n",
    "                    df.i.j$prob <- NA\n",
    "                  }\n",
    "                  else if ((idx.max != idx.min) & !is.null(min.dataset)) {\n",
    "                    if (!(df.i.j$dataset[idx.min] %in% dataset.name[min.dataset])) {\n",
    "                      df.i.j$prob <- NA\n",
    "                    }\n",
    "                    else if (length(dataset.na) > 0 & sum(!(dataset.name[min.dataset] %in% \n",
    "                      dataset.na)) > 0) {\n",
    "                      df.i.j$prob <- NA\n",
    "                    }\n",
    "                  }\n",
    "                }\n",
    "                df.i[df.i$group.names == j, \"prob\"] <- df.i.j$prob\n",
    "            }\n",
    "            df[df$interaction_name_2 == i, \"prob\"] <- df.i$prob\n",
    "        }\n",
    "    }\n",
    "    if (remove.isolate) {\n",
    "        df <- df[!is.na(df$prob), ]\n",
    "        line.on <- FALSE\n",
    "    }\n",
    "    if (nrow(df) == 0) {\n",
    "        stop(\"No interactions are detected. Please consider changing the cell groups for analysis. \")\n",
    "    }\n",
    "    if (!is.null(pairLR.use)) {\n",
    "        interaction_name_2.order <- intersect(object@DB$interaction[pairLR.use$interaction_name, \n",
    "            ]$interaction_name_2, unique(df$interaction_name_2))\n",
    "        df$interaction_name_2 <- factor(df$interaction_name_2, \n",
    "            levels = interaction_name_2.order)\n",
    "    }\n",
    "    df$source.target = droplevels(df$source.target, exclude = setdiff(levels(df$source.target), \n",
    "        unique(df$source.target)))\n",
    "    # print('8')\n",
    "    # print(unique(df$source.target))\n",
    "    if (sort.by.target & !sort.by.source) {\n",
    "        if (!is.null(targets.use)) {\n",
    "            df$target <- factor(df$target, levels = intersect(targets.use, \n",
    "                df$target))\n",
    "            df <- with(df, df[order(target, source), ])\n",
    "            source.target.order <- unique(as.character(df$source.target))\n",
    "            df$source.target <- factor(df$source.target, levels = source.target.order)\n",
    "        }\n",
    "    }\n",
    "    if (sort.by.source & !sort.by.target) {\n",
    "        if (!is.null(sources.use)) {\n",
    "            df$source <- factor(df$source, levels = intersect(sources.use, \n",
    "                df$source))\n",
    "            df <- with(df, df[order(source, target), ])\n",
    "            source.target.order <- unique(as.character(df$source.target))\n",
    "            df$source.target <- factor(df$source.target, levels = source.target.order)\n",
    "        }\n",
    "    }\n",
    "    if (sort.by.source & sort.by.target) {\n",
    "        if (!is.null(sources.use)) {\n",
    "            df$source <- factor(df$source, levels = intersect(sources.use, \n",
    "                df$source))\n",
    "            if (!is.null(targets.use)) {\n",
    "                df$target <- factor(df$target, levels = intersect(targets.use, \n",
    "                  df$target))\n",
    "            }\n",
    "            if (sort.by.source.priority) {\n",
    "                df <- with(df, df[order(source, target), ])\n",
    "            }\n",
    "            else {\n",
    "                df <- with(df, df[order(target, source), ])\n",
    "            }\n",
    "            source.target.order <- unique(as.character(df$source.target))\n",
    "            df$source.target <- factor(df$source.target, levels = source.target.order)\n",
    "        }\n",
    "    }\n",
    "    df$source.target = as.character(df$source.target)\n",
    "    # print(typeof(df$source.target))    \n",
    "    # print(unique(df$source.target)[4])\n",
    "    # print(typeof(unique(df$source.target)[4]))\n",
    "    # print(df[is.na(df$source.target),])\n",
    "    df = df[!is.na(df$source.target),]\n",
    "    df$source.target = factor(df$source.target)\n",
    "    g <- ggplot(df, aes(x = interaction_name_2, y =  source.target, \n",
    "        color = prob, size = pval)) + geom_point(pch = 16) + \n",
    "        theme_linedraw() + theme(panel.grid.major = element_blank()) + \n",
    "        theme(axis.text.x = element_text(angle = angle.x, hjust = hjust.x, \n",
    "            vjust = vjust.x), axis.title.x = element_blank(), \n",
    "            axis.title.y = element_blank()) + scale_x_discrete(position = \"bottom\")\n",
    "    values <- c(1, 2, 3)\n",
    "    names(values) <- c(\"p > 0.05\", \"0.01 < p < 0.05\", \"p < 0.01\")\n",
    "    g <- g + scale_radius(range = c(min(df$pval), max(df$pval)), \n",
    "        breaks = sort(unique(df$pval)), labels = names(values)[values %in% \n",
    "            sort(unique(df$pval))], name = \"p-value\")\n",
    "    if (min(df$prob, na.rm = T) != max(df$prob, na.rm = T)) {\n",
    "        g <- g + scale_colour_gradientn(colors = colorRampPalette(color.use)(99), \n",
    "            na.value = \"white\", limits = c(quantile(df$prob, \n",
    "                0, na.rm = T), quantile(df$prob, 1, na.rm = T)), \n",
    "            breaks = c(quantile(df$prob, 0, na.rm = T), quantile(df$prob, \n",
    "                1, na.rm = T)), labels = c(\"min\", \"max\")) + guides(color = guide_colourbar(barwidth = 0.5, \n",
    "            title = \"Commun. Prob.\"))\n",
    "    }\n",
    "    else {\n",
    "        g <- g + scale_colour_gradientn(colors = colorRampPalette(color.use)(99), \n",
    "            na.value = \"white\") + guides(color = guide_colourbar(barwidth = 0.5, \n",
    "            title = \"Commun. Prob.\"))\n",
    "    }\n",
    "    g <- g + theme(text = element_text(size = font.size), plot.title = element_text(size = font.size.title)) + \n",
    "        theme(legend.title = element_text(size = 8), legend.text = element_text(size = 6))\n",
    "    if (grid.on) {\n",
    "        if (length(unique(df$source.target)) > 1) {\n",
    "            g <- g + geom_vline(xintercept = seq(1.5, length(unique(df$source.target)) - \n",
    "                0.5, 1), lwd = 0.1, colour = color.grid)\n",
    "        }\n",
    "        if (length(unique(df$interaction_name_2)) > 1) {\n",
    "            g <- g + geom_hline(yintercept = seq(1.5, length(unique(df$interaction_name_2)) - \n",
    "                0.5, 1), lwd = 0.1, colour = color.grid)\n",
    "        }\n",
    "    }\n",
    "    if (!is.null(title.name)) {\n",
    "        g <- g + ggtitle(title.name) + theme(plot.title = element_text(hjust = 0.5))\n",
    "    }\n",
    "    if (!is.null(comparison)) {\n",
    "        if (line.on) {\n",
    "            xintercept = seq(0.5 + length(dataset.name[comparison]), \n",
    "                length(group.names0) * length(dataset.name[comparison]), \n",
    "                by = length(dataset.name[comparison]))\n",
    "            g <- g + geom_vline(xintercept = xintercept, linetype = \"dashed\", \n",
    "                color = \"grey60\", size = line.size)\n",
    "        }\n",
    "        if (color.text.use) {\n",
    "            if (is.null(group)) {\n",
    "                group <- 1:length(comparison)\n",
    "                names(group) <- dataset.name[comparison]\n",
    "            }\n",
    "            if (is.null(color.text)) {\n",
    "                color <- ggPalette(length(unique(group)))\n",
    "            }\n",
    "            else {\n",
    "                color <- color.text\n",
    "            }\n",
    "            names(color) <- names(group[!duplicated(group)])\n",
    "            color <- color[group]\n",
    "            dataset.name.order <- levels(df$source.target)\n",
    "            dataset.name.order <- stringr::str_match(dataset.name.order, \n",
    "                \"\\\\(.*\\\\)\")\n",
    "            dataset.name.order <- stringr::str_sub(dataset.name.order, \n",
    "                2, stringr::str_length(dataset.name.order) - \n",
    "                  1)\n",
    "            xtick.color <- color[dataset.name.order]\n",
    "            g <- g + theme(axis.text.x = element_text(colour = xtick.color))\n",
    "        }\n",
    "    }\n",
    "    if (!show.legend) {\n",
    "        g <- g + theme(legend.position = \"none\")\n",
    "    }\n",
    "    if (return.data) {\n",
    "        return(list(communication = df, gg.obj = g))\n",
    "    }\n",
    "    else {\n",
    "        return(g)\n",
    "    }\n",
    "}\n",
    "all_celltypes = unique(cch@idents[[1]])\n",
    "setwd(fig.dir)\n",
    "for (i in 1:length(all_celltypes)) {  # length(all_celltypes)\n",
    "    print(all_celltypes[i])\n",
    "    f1 = paste0(samples, '_deSig_as_source_', all_celltypes[i], '_up.pdf')\n",
    "    # f2 = paste0(samples, '_deSig_as_source_', all_celltypes[i], '_down.pdf')\n",
    "    f3 = paste0(samples, '_deSig_as_target_', all_celltypes[i], '_up.pdf')\n",
    "    # f4 = paste0(samples, '_deSig_as_target_', all_celltypes[i], '_down.pdf')\n",
    "    print('up')\n",
    "    tryCatch({\n",
    "        pairLR.use.up = net.up[, \"interaction_name\", drop = F]\n",
    "        par(mar = c(1, 1, 1, 1))\n",
    "        gg1 <- netVisual_bubble(cch, pairLR.use = pairLR.use.up, sources.use = all_celltypes[i], comparison = c(1, 2), angle.x=45, color.text.use = F, \n",
    "            title.name = paste0(\"Up-regulated signaling in \", names(cch.list)[2]))\n",
    "        ggsave(filename = f1, plot=gg1, width = 10, height = 10)\n",
    "    }, error = function(e) {\n",
    "        print(e)\n",
    "    })\n",
    "    tryCatch({\n",
    "        par(mar = c(1, 1, 1, 1))\n",
    "        gg3 <- netVisual_bubble(cch, pairLR.use = pairLR.use.up, targets.use = all_celltypes[i], comparison = c(1, 2), angle.x=45, color.text.use = F,\n",
    "            title.name = paste0(\"Up-regulated signaling in \", names(cch.list)[2]))\n",
    "        ggsave(filename = f3, plot=gg1, width = 10, height = 10)\n",
    "    }, error = function(e) {\n",
    "        print(e)\n",
    "    })\n",
    "    # tryCatch({\n",
    "    #     print('down')\n",
    "    #     pairLR.use.down = net.down[, \"interaction_name\", drop = F]\n",
    "    #     gg2 <- netVisual_bubble(cch, pairLR.use = pairLR.use.down, sources.use = all_celltypes[i], comparison = c(1, 2), angle.x=45, color.text.use = F, \n",
    "    #         title.name = paste0(\"Down-regulated signaling in \", names(cch.list)[2]))\n",
    "    #     ggsave(filename = f2, plot=gg2, width = 10, height = 10)\n",
    "    # }, error = function(e) {\n",
    "    #     print(e)\n",
    "    # })\n",
    "    # tryCatch({\n",
    "    #     gg4 <- netVisual_bubble(cch, pairLR.use = pairLR.use.down, targets.use = all_celltypes[i], comparison = c(1, 2), angle.x=45, color.text.use = F,\n",
    "    #         title.name = paste0(\"Down-regulated signaling in \", names(cch.list)[2]))\n",
    "    #     ggsave(filename = f4, plot=gg2, width = 10, height = 10)\n",
    "    # }, error = function(e) {\n",
    "    #     print(e)\n",
    "    # })\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "020be55d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"CCL\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"MIF\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"VISFATIN\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"MK\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"FN1\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"COLLAGEN\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LAMININ\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Cholesterol\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"APP\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"CDH5\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"CLDN\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ESAM\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"JAM\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"NOTCH\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"PECAM1\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"VCAM\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"GAP\"\n",
      "[1] \"up\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Comparing communications on a merged object \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "all_celltypes = unique(c(net.up$pathway_name))\n",
    "setwd(fig.dir)\n",
    "for (i in 1:length(all_celltypes)) {  # length(all_celltypes)\n",
    "    print(all_celltypes[i])\n",
    "    f1 = paste0(samples, '_deSig_', all_celltypes[i], '_up.pdf')\n",
    "    f2 = paste0(samples, '_deSig_', all_celltypes[i], '_down.pdf')\n",
    "    print('up')\n",
    "    if (length(net.up)>=1) {\n",
    "        tryCatch({\n",
    "            pairLR.use.up = subset(net.up, pathway_name == all_celltypes[i])[, \"interaction_name\", drop = F]\n",
    "            gg1 <- netVisual_bubble(cch, pairLR.use = pairLR.use.up, comparison = c(1, 2), angle.x=45, color.text.use = F, \n",
    "                title.name = paste0(\"Up-regulated signaling in \", names(cch.list)[2]))\n",
    "            ggsave(filename = f1, plot=gg1, width = 17, height = 8)\n",
    "        }, error = function(e) {\n",
    "            print(e)\n",
    "        })\n",
    "    }\n",
    "\n",
    "    # if (length(net.down)>=1) {\n",
    "    #     tryCatch({\n",
    "    #         print('down')\n",
    "    #         pairLR.use.down = subset(net.down, pathway_name == all_celltypes[i])[, \"interaction_name\", drop = F]\n",
    "    #         gg2 <- netVisual_bubble(cch, pairLR.use = pairLR.use.down, comparison = c(1, 2), angle.x=45, color.text.use = F, \n",
    "    #             title.name = paste0(\"Down-regulated signaling in \", names(cch.list)[2]))\n",
    "    #         ggsave(filename = f2, plot=gg2, width = 17, height = 8)\n",
    "    #     }, error = function(e) {\n",
    "    #         print(e)\n",
    "    #     })\n",
    "    # }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fc38ad7",
   "metadata": {},
   "source": [
    "Visualize the enriched ligands, signaling,or ligand-receptor pairs in one condition compared to another condition using wordcloud"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ee73c909",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in loadNamespace(x): there is no package called ‘wordcloud’\n",
     "output_type": "error",
     "traceback": [
      "Error in loadNamespace(x): there is no package called ‘wordcloud’\nTraceback:\n",
      "1. computeEnrichmentScore(net.up, species = \"human\")",
      "2. loadNamespace(x)   # at line 49-52 of file <text>",
      "3. withRestarts(stop(cond), retry_loadNamespace = function() NULL)",
      "4. withOneRestart(expr, restarts[[1L]])",
      "5. doWithOneRestart(return(expr), restart)"
     ]
    }
   ],
   "source": [
    "computeEnrichmentScore <- function (df, measure = c(\"ligand\", \"signaling\", \"LR-pair\"), \n",
    "    species = c(\"mouse\", \"human\"), color.use = NULL, color.name = \"Dark2\", \n",
    "    n.color = 8, scale = c(4, 0.8), min.freq = 0, max.words = 200, \n",
    "    random.order = FALSE, rot.per = 0, return.data = FALSE, seed = 1, \n",
    "    ...) \n",
    "{\n",
    "    measure <- match.arg(measure)\n",
    "    species <- match.arg(species)\n",
    "    LRpairs <- as.character(unique(df$interaction_name))\n",
    "    ES <- vector(length = length(LRpairs))\n",
    "    for (i in 1:length(LRpairs)) {\n",
    "        df.i <- subset(df, interaction_name == LRpairs[i])\n",
    "        if (length(which(rowSums(is.na(df.i)) > 0)) > 0) {\n",
    "            df.i <- df.i[-which(rowSums(is.na(df.i)) > 0), , \n",
    "                drop = FALSE]\n",
    "        }\n",
    "        ES[i] = mean(abs(df.i$ligand.logFC) * abs(df.i$receptor.logFC) * \n",
    "            abs(df.i$ligand.pct.2 - df.i$ligand.pct.1) * abs(df.i$receptor.pct.2 - \n",
    "            df.i$receptor.pct.1))\n",
    "    }\n",
    "    if (species == \"mouse\") {\n",
    "        CellChatDB <- CellChatDB.mouse\n",
    "    }\n",
    "    else if (species == \"human\") {\n",
    "        CellChatDB <- CellChatDB.human\n",
    "    }\n",
    "    df.es <- CellChatDB$interaction[LRpairs, c(\"ligand\", \"receptor\", \n",
    "        \"pathway_name\")]\n",
    "    df.es$score <- ES\n",
    "    df.es.ensemble <- df.es %>% group_by(ligand) %>% summarize(total = sum(score))\n",
    "    set.seed(seed)\n",
    "    if (is.null(color.use)) {\n",
    "        color.use <- RColorBrewer::brewer.pal(n.color, color.name)\n",
    "    }\n",
    "    freq <- df.es.ensemble$total\n",
    "    # Check for missing values in freq vector\n",
    "    if (any(is.na(freq))) {\n",
    "    freq <- freq[!is.na(freq)]\n",
    "    }\n",
    "\n",
    "    # Compute the minimum frequency based on non-missing values\n",
    "    min.freq <- ifelse(min.freq > max(freq), 0, min.freq)\n",
    "\n",
    "    # Call the wordcloud function with the updated arguments\n",
    "    # wordcloud::wordcloud(words = df.es.ensemble$ligand, freq = freq, \n",
    "    #                     min.freq = min.freq, max.words = max.words, \n",
    "    #                     scale = scale, random.order = random.order, \n",
    "    #                     rot.per = rot.per, colors = color.use, ...)\n",
    "    wordcloud::wordcloud(words = df.es.ensemble$ligand, freq = freq, \n",
    "        min.freq = min.freq, max.words = max.words, scale = scale, \n",
    "        random.order = random.order, rot.per = rot.per, colors = color.use, \n",
    "        ...)\n",
    "    if (return.data) {\n",
    "        return(df.es.ensemble)\n",
    "    }\n",
    "}\n",
    "# visualize the enriched ligands in the first condition\n",
    "# pdf(paste0(samples, '.enrichment_down.pdf'), width = 10, height = 10)\n",
    "# computeEnrichmentScore(net.down, species = 'human')\n",
    "# dev.off()\n",
    "\n",
    "# visualize the enriched ligands in the second condition\n",
    "pdf(paste0(samples, '.enrichment_up.pdf'), width = 10, height = 10)\n",
    "computeEnrichmentScore(net.up, species = 'human')\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f1191ff",
   "metadata": {},
   "source": [
    "# Part IV: Visually compare cell-cell communication using Hierarchy plot, Circle plot or Chord diagram"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c66291f9",
   "metadata": {},
   "source": [
    "Similar to the CellChat analysis of individual dataset, we can visualize the cell-cell communication network using Hierarchy plot, Circle plot or Chord diagram.\n",
    "\n",
    "**Edge color/weight, node color/size/shape**: In all visualization plots, edge colors are consistent with the sources as sender, and edge weights are proportional to the interaction strength. Thicker edge line indicates a stronger signal. In the **Hierarchy plot and Circle plot**, circle sizes are proportional to the number of cells in each cell group. In the hierarchy plot, solid and open circles represent source and target, respectively. In the **Chord diagram**, the inner thinner bar colors represent the targets that receive signal from the corresponding outer bar. The inner bar size is proportional to the signal strength received by the targets. Such inner bar is helpful for interpreting the complex chord diagram. Note that there exist some inner bars without any chord for some cell groups, please just igore it because this is an issue that has not been addressed by [circlize](https://github.com/jokergoo/circlize) package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "There are 100 unique colors in the vector `col` and 100 unique values\n",
      "in `matrix`. `Heatmap()` will treat it as an exact discrete one-to-one\n",
      "mapping. If this is not what you want, slightly change the number of\n",
      "colors, e.g. by adding one more color or removing a color.\n",
      "\n",
      "No legend element is put in the last 1 row under `nrow = 100`, maybe\n",
      "you should set `by_row = FALSE`? Reset `nrow` to 99.\n",
      "\n",
      "No legend element is put in the last 1 row under `nrow = 100`, maybe\n",
      "you should set `by_row = FALSE`? Reset `nrow` to 99.\n",
      "\n",
      "No legend element is put in the last 1 row under `nrow = 100`, maybe\n",
      "you should set `by_row = FALSE`? Reset `nrow` to 99.\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n",
      "Do heatmap based on a single object \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "setwd(fig.dir)\n",
    "for (pathways.show in pathways) {\n",
    "    f1 = paste0(samples, '_circleplot_', pathways.show,'.pdf')\n",
    "    f2 = paste0(samples, '_regularheatmap_', pathways.show,'.pdf')\n",
    "    tryCatch({\n",
    "        weight.max <- getMaxWeight(frozen.list, slot.name = c(\"netP\"), attribute = pathways.show) # control the edge weights across different datasets\n",
    "        \n",
    "        pdf(f1, width = 10, height = 10)\n",
    "        par(mfrow = c(1,2), xpd=TRUE)\n",
    "        for (i in 1:length(cch.list)) {\n",
    "            netVisual_aggregate(cch.list[[i]], signaling = pathways.show, layout = \"circle\", edge.weight.max = weight.max[1], edge.width.max = 10, signaling.name = paste(pathways.show, names(cch.list)[i]))\n",
    "        }\n",
    "        dev.off()\n",
    "\n",
    "        pdf(f2, width = 10, height = 10)\n",
    "        par(mfrow = c(1,2), xpd=TRUE)\n",
    "        for (i in 1:length(cch.list)) {\n",
    "            ht = netVisual_heatmap(cch.list[[i]], signaling = pathways.show, color.heatmap = \"Reds\", title.name = paste(pathways.show, \"signaling \",names(cch.list)[i]))\n",
    "            ComplexHeatmap::draw(ht, ht_gap = unit(0.5, \"cm\"))\n",
    "        }\n",
    "        dev.off()\n",
    "    }, error = function(e) {\n",
    "        print(e)\n",
    "    })\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71ec836b",
   "metadata": {},
   "source": [
    "# Part V: Compare the signaling gene expression distribution between different datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73b76900",
   "metadata": {},
   "source": [
    "We can plot the gene expression distribution of signaling genes related to L-R pairs or signaling pathway using a Seurat wrapper function `plotGeneExpression`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "51c12d2c",
   "metadata": {
    "Rmd_chunk_options": "fig.width=7,fig.height = 4, fig.wide = TRUE, fig.align = \"center\"",
    "kernel": "R",
    "tags": [
     "report_cell"
    ],
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n",
      "<simpleError in Idents(object): could not find function \"Idents\">\n"
     ]
    }
   ],
   "source": [
    "cch@meta$datasets = factor(cch@meta$datasets, levels = names(cch.list)) # set factor level\n",
    "setwd(fig.dir)\n",
    "for (pathways.show in pathways) {\n",
    "    f1 = paste0(samples, '_violin_', pathways.show,'.pdf')\n",
    "    # f2 = paste0(samples, '_regularheatmap_', pathways.show,'.pdf')\n",
    "    tryCatch({\n",
    "        weight.max <- getMaxWeight(frozen.list, slot.name = c(\"netP\"), attribute = pathways.show) # control the edge weights across different datasets\n",
    "        # pdf(f1, width = 10, height = 10)\n",
    "        g = plotGeneExpression(cch, signaling = pathways.show, split.by = \"datasets\", colors.ggplot = T)\n",
    "        # dev.off()\n",
    "        ggsave(filename = f1, plot=g, width = 10, height = 10)\n",
    "    }, error = function(e) {\n",
    "        print(e)\n",
    "    })\n",
    "}\n"
   ]
  }
 ],
 "metadata": {
  "Rmd_chunk_options": {
   "author": "Suoqin Jin",
   "date": "`r format(Sys.time(), '%d %B, %Y')`",
   "mainfont": "Arial",
   "output": {
    "html_document": {
     "theme": "united",
     "toc": true
    }
   },
   "title": "Comparison analysis of multiple datasets using CellChat",
   "vignette": "\n%\\VignetteIndexEntry{Comparison analysis of multiple datasets using CellChat}\n%\\VignetteEngine{knitr::rmarkdown}\n%\\VignetteEncoding{UTF-8}\n"
  },
  "kernelspec": {
   "display_name": "$name",
   "language": "R",
   "name": "dcats"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ],
    [
     "R",
     "ir",
     "",
     ""
    ],
    [
     "Markdown",
     "markdown",
     "",
     ""
    ]
   ]
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
